<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Automated genomic fine-mapping • echolocatoR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Automated genomic fine-mapping">
<meta property="og:description" content="Automated statistical and functional fine-mapping 
  with extensive access to genome-wide datasets.  ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">echolocatoR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/echolocatoR.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/QTLs.html">QTLs</a>
    </li>
    <li>
      <a href="articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="articles/finemapping_portal.html">echolocatoR Fine-mapping Portal</a>
    </li>
    <li>
      <a href="articles/plot_locus.html">Plot locus</a>
    </li>
    <li>
      <a href="articles/summarise.html">Summarise</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RajLabMSSM/echolocatoR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<img src="https://github.com/RajLabMSSM/echolocatoR/raw/master/inst/hex/hex.png" height="300"><br><br><a href="https://github.com/RajLabMSSM/echolocatoR" class="external-link"><img src="https://img.shields.io/badge/devel%20version-2.0.3-black.svg"></a> <a href="https://github.com/RajLabMSSM/echolocatoR/actions" class="external-link"><img src="https://github.com/RajLabMSSM/echolocatoR/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a> <a href="https://github.com/RajLabMSSM/echolocatoR/commits/master" class="external-link"><img src="https://img.shields.io/github/last-commit/RajLabMSSM/echolocatoR.svg"></a> <a href="https://app.codecov.io/gh/RajLabMSSM/echolocatoR" class="external-link"><img src="https://app.codecov.io/gh/RajLabMSSM/echolocatoR/branch/master/graph/badge.svg"></a> <a href="https://cran.r-project.org/web/licenses/GPL-3" class="external-link"><img src="https://img.shields.io/badge/license-GPL--3-blue.svg" alt="License: GPL-3"></a> <a href="https://doi.org/10.1093/bioinformatics/btab658" class="external-link"><img src="https://img.shields.io/badge/doi-10.1093/bioinformatics/btab658-blue.svg"></a> ¶
<h5>
¶ Author: <i>Brian M. Schilder</i> ¶
</h5>
<h5>
¶ README updated: <i>Nov-07-2022</i> ¶
</h5>
<div class="section level2">
<h2 id="echolocator-automated-statistical-and-functional-fine-mapping">
<code>echolocatoR</code>: Automated statistical and functional fine-mapping<a class="anchor" aria-label="anchor" href="#echolocator-automated-statistical-and-functional-fine-mapping"></a>
</h2>
<p>with extensive access to genome-wide datasets.</p>
<div class="section level3">
<h3 id="the-echoverse">The <em>echoverse</em>
<a class="anchor" aria-label="anchor" href="#the-echoverse"></a>
</h3>
<p><code>echolocatoR</code> is part of the <a href="https://github.com/topics/echoverse" class="external-link"><strong><em>echoverse</em></strong></a>, a suite of R packages designed to facilitate different steps in genetic fine-mapping.</p>
<p><code>echolocatoR</code> calls each of these other packages (i.e. “modules”) internally to create a unified pipeline. However, you can also use each module independently to create your own custom workflows.</p>
<div class="section level4">
<h4 id="echoverse-dependency-graph">
<strong><em>echoverse</em></strong> dependency graph<a class="anchor" aria-label="anchor" href="#echoverse-dependency-graph"></a>
</h4>
<p><img src="./images/echoverse.png" height="400px" style="border-radius: 20px;"></p>
<blockquote>
<p>Made with <a href="https://github.com/RajLabMSSM/echodeps" class="external-link"><code>echodeps</code></a>, yet another <strong><em>echoverse</em></strong> module. See <a href="https://rajlabmssm.github.io/Fine_Mapping/echolocatoR.dep_graph.html" class="external-link">here for the interactive version</a> with package descriptions and links to each GitHub repo.</p>
</blockquote>
</div>
</div>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>If you use <code>echolocatoR</code>, or any of the <strong>echoverse</strong> modules, please cite:</p>
<blockquote>
<p>Brian M Schilder, Jack Humphrey, Towfique Raj (2021) echolocatoR: an automated end-to-end statistical and functional genomic fine-mapping pipeline, <em>Bioinformatics</em>; btab658, <a href="https://doi.org/10.1093/bioinformatics/btab658" class="external-link uri">https://doi.org/10.1093/bioinformatics/btab658</a></p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://remotes.r-lib.org" class="external-link">"remotes"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"RajLabMSSM/echolocatoR"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RajLabMSSM/echolocatoR" class="external-link">echolocatoR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="installation-troubleshooting">Installation troubleshooting<a class="anchor" aria-label="anchor" href="#installation-troubleshooting"></a>
</h4>
<details><ul>
<li>Because <code>echolocatoR</code> now relies on many subpackages that rely on one another, sometimes errors can occur when R tries to update one R package before updating its <em>echoverse</em> dependencies (and thus is unable to find new functions). As <em>echoverse</em> stabilizes over time, this should happen less frequently. However, in the meantime the solution is to simply rerun <code>remotes::install_github("RajLabMSSM/echolocatoR")</code> until all subpackages are fully updates.</li>
<li>System dependencies can sometimes cause issues when using different packages. I’ve tried to account for as many of these as possible automatically within the code, but using the <strong>Docker/Singularity</strong> provided below can further mitigate these issues.</li>
<li>The R package <code>XML</code> (which some <em>echoverse</em> subpackages depend on) has some additional system dependencies that must be installed beforehand. If <code>XML</code> does not install automatically, try installing <code>lbxml</code> on your system using <code>brew install libxml2</code> (MacOS), <code>sudo apt-get install libxml2</code> (Linux) or <code>conda install r-xml</code> if you are running <code>echolocatoR</code> from within a conda environment.</li>
</ul></details>
</div>
<div class="section level3">
<h3 id="optional-dockersingularity">[Optional] <a href="https://rajlabmssm.github.io/echolocatoR/articles/docker" class="external-link">Docker/Singularity</a>
<a class="anchor" aria-label="anchor" href="#optional-dockersingularity"></a>
</h3>
<p><code>echolocatoR</code> now has its own dedicated Docker/Singularity container! This greatly reduces issues related to system dependency conflicts and provides a containerized interface for Rstudio through your web browser. See <a href="https://rajlabmssm.github.io/echolocatoR/articles/docker" class="external-link">here for installation instructions</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<div class="section level3">
<h3 id="website">
<a href="https://rajlabmssm.github.io/echolocatoR" class="external-link">Website</a><a class="anchor" aria-label="anchor" href="#website"></a>
</h3>
</div>
<div class="section level3">
<h3 id="get-started">
<a href="https://rajlabmssm.github.io/echolocatoR/articles/echolocatoR" class="external-link">Get started</a><a class="anchor" aria-label="anchor" href="#get-started"></a>
</h3>
</div>
<div class="section level3">
<h3 id="bugsrequests">
<a href="https://github.com/RajLabMSSM/echolocatoR/issues" class="external-link">Bugs/requests</a><a class="anchor" aria-label="anchor" href="#bugsrequests"></a>
</h3>
<p>Please report any bugs/requests on <a href="https://github.com/RajLabMSSM/echolocatoR/issues" class="external-link">GitHub Issues</a>.</p>
<p><a href="https://github.com/RajLabMSSM/echolocatoR/pulls" class="external-link">Contributions</a> are welcome!</p>
</div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Fine-mapping methods are a powerful means of identifying causal variants underlying a given phenotype, but are underutilized due to the technical challenges of implementation. <code>echolocatoR</code> is an R package that automates end-to-end genomics fine-mapping, annotation, and plotting in order to identify the most probable causal variants associated with a given phenotype.</p>
<p>It requires minimal input from users (a GWAS or QTL summary statistics file), and includes a suite of statistical and functional fine-mapping tools. It also includes extensive access to datasets (linkage disequilibrium panels, epigenomic and genome-wide annotations, QTL).</p>
<p>The elimination of data gathering and preprocessing steps enables rapid fine-mapping of many loci in any phenotype, complete with locus-specific publication-ready figure generation. All results are merged into a single per-SNP summary file for additional downstream analysis and results sharing. Therefore <code>echolocatoR</code> drastically reduces the barriers to identifying causal variants by making the entire fine-mapping pipeline rapid, robust and scalable.</p>
<p><img src="./images/echolocatoR_Fig1.png" style="border-radius: 10px;"></p>
</div>
<div class="section level2">
<h2 id="literature">Literature<a class="anchor" aria-label="anchor" href="#literature"></a>
</h2>
<div class="section level3">
<h3 id="for-applications-of-echolocator-in-the-literature-please-see">For applications of <code>echolocatoR</code> in the literature, please see:<a class="anchor" aria-label="anchor" href="#for-applications-of-echolocator-in-the-literature-please-see"></a>
</h3>
<blockquote>
<ol style="list-style-type: decimal">
<li>E Navarro, E Udine, K de Paiva Lopes, M Parks, G Riboldi, BM Schilder…T Raj (2020) Dysregulation of mitochondrial and proteo-lysosomal genes in Parkinson’s disease myeloid cells. Nature Genetics. <a href="https://doi.org/10.1101/2020.07.20.212407" class="external-link uri">https://doi.org/10.1101/2020.07.20.212407</a>
</li>
<li>BM Schilder, T Raj (2021) Fine-Mapping of Parkinson’s Disease Susceptibility Loci Identifies Putative Causal Variants. Human Molecular Genetics, ddab294, <a href="https://doi.org/10.1093/hmg/ddab294" class="external-link uri">https://doi.org/10.1093/hmg/ddab294</a><br>
</li>
<li>K de Paiva Lopes, G JL Snijders, J Humphrey, A Allan, M Sneeboer, E Navarro, BM Schilder…T Raj (2022) Genetic analysis of the human microglial transcriptome across brain regions, aging and disease pathologies. Nature Genetics, <a href="https://doi.org/10.1038/s41588-021-00976-y" class="external-link uri">https://doi.org/10.1038/s41588-021-00976-y</a>
</li>
</ol>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="echolocator-v10-vs-v20">
<code>echolocatoR</code> v1.0 vs. v2.0<a class="anchor" aria-label="anchor" href="#echolocator-v10-vs-v20"></a>
</h2>
<p>There have been a series of major updates between <code>echolocatoR</code> v1.0 and v2.0. Here are some of the most notable ones (see <strong>Details</strong>):</p>
<details><ul>
<li>
<strong><em>echoverse</em> subpackages</strong>: <code>echolocatoR</code> has been broken into separate subpackages, making it much easier to edit/debug each step of the full <code>finemap_loci</code> pipeline, and improving robustness throughout. It also provides greater flexibility for users to construct their own custom pipelines from these modules.</li>
<li>
<strong><code>GITHUB_TOKEN</code></strong>: GitHub now requires users to create Personal Authentication Tokens (PAT) to avoid download limits. This is essential for installing <code>echolocatoR</code> as many resources from GitHub need to be downloaded. See <a href="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token" class="external-link">here</a> for further instructions. = <code><a href="https://rdrr.io/pkg/echodata/man/construct_colmap.html" class="external-link">echodata::construct_colmap()</a></code>: Previously, users were required to input key column name mappings as separate arguments to <code><a href="reference/finemap_loci.html">echolocatoR::finemap_loci</a></code>. This functionality has been deprecated and replaced with a single argument, <code>colmap=</code>. This allows users to save the <code><a href="https://rdrr.io/pkg/echodata/man/construct_colmap.html" class="external-link">construct_colmap()</a></code> output as a single variable and reuse it later without having to write out each mapping argument again (and helps reduce an already crowded list of arguments).</li>
<li>
<strong><code>MungeSumstats</code></strong>: <code>finemap_loci</code> now accepts the output of <a href="https://github.com/neurogenomics/MungeSumstats" class="external-link"><code>MungeSumstats::format_sumstats</code>/<code>import_sumstats</code></a> as-is (without requiring <code>colmap=</code>, so long as <code>munged=TRUE</code>). Standardizing your GWAS/QTL summary stats this way greatly reduces (or eliminates) the time taken to do manual formatting.</li>
<li>
<strong><code><a href="reference/finemap_loci.html">echolocatoR::finemap_loci</a></code> arguments</strong>: Several arguments have been deprecated or had their names changed to be harmonized across all the subpackages and use a unified naming convention. See <code><a href="reference/finemap_loci.html">?echolocatoR::finemap_loci</a></code> for details.</li>
<li>
<strong><code>echoconda</code></strong>: The <em>echoverse</em> subpackage <code>echoconda</code> now handles all conda environment creation/use internally and automatically, without the need for users to create the conda environment themselves as a separate step. Also, the default conda env <code>echoR</code> has been replaced by <code>echoR_mini</code>, which reduces the number of dependencies to just the bare minimum (thus greatly speeding up build time and reducing potential version conflicts).</li>
<li>
<strong><code>FINEMAP</code></strong>: More outputs from the tool <code>FINEMAP</code> are now recorded in the <code>echolocatoR</code> results (see <code><a href="https://rdrr.io/pkg/echofinemap/man/FINEMAP.html" class="external-link">?echofinemap::FINEMAP</a></code> or <a href="https://github.com/RajLabMSSM/echofinemap/issues/7" class="external-link">this Issue</a> for details). Also, a common dependency conflict between <code>FINEMAP</code>&gt;=1.4 and MacOS has been resolved (see <a href="https://github.com/RajLabMSSM/echofinemap/issues/9" class="external-link">this Issue</a> for details.</li>
<li>
<strong><code>echodata</code></strong>: All example data and data transformation functions have been moved to the <em>echoverse</em> subpackage <a href="https://github.com/RajLabMSSM/echodata" class="external-link"><code>echodata</code></a>.</li>
<li>
<strong><code>LD_reference=</code></strong>: In addition to the <em>UKB</em>, <em>1KGphase1/3</em> LD reference panels, <code><a href="reference/finemap_loci.html">finemap_loci()</a></code> can now take custom LD panels by supplying <code>finemap_loci(LD_reference=)</code> with a list of paths to VCF files (.vcf / vcf.gz / vcf.bgz) or pre-computed LD matrices with RSIDs as the row/col names (.rda / .rds / .csv / .tsv. / .txt / .csv.gz / tsv.gz / txt.gz).</li>
<li>
<strong>Expanded fine-mapping methods</strong>: “ABF”, “COJO_conditional”, “COJO_joint” “COJO_stepwise”,“FINEMAP”,“PAINTOR” (including multi-GWAS and multi-ancestry fine-mapping),“POLYFUN_FINEMAP” ,“POLYFUN_SUSIE”,“SUSIE”</li>
<li>
<strong><code>FINEMAP</code> fixed</strong>: There were a number of issues with <code>FINEMAP</code> due to differing output formats across different versions, system dependency conflicts, and the fact that it can produce multiple Credible Sets. All of these have been fixed and the latest version of <code>FINEMAP</code> can be run on all OS platforms.<br>
</li>
<li>
<strong>Debug mode</strong>: Within <code><a href="reference/finemap_loci.html">finemap_loci()</a></code> I use a <code><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch()</a></code> when iterating across loci so that if one locus fails, the rest can continue. However this prevents using traceback feature in R, making debugging hard. Thus I now enabled debugging mode via a new argument: <code>use_tryCatch=FALSE</code>.</li>
</ul></details>
</div>
<div class="section level2">
<h2 id="output-descriptions">Output descriptions<a class="anchor" aria-label="anchor" href="#output-descriptions"></a>
</h2>
<p>By default, <code><a href="reference/finemap_loci.html">echolocatoR::finemap_loci()</a></code> returns a nested list containing grouped by locus names (e.g. <code>$BST1</code>, <code>$MEX3C</code>). The results of each locus contain the following elements:</p>
<details><ul>
<li>
<code>finemap_dat</code>: Fine-mapping results from all selected methods merged with the original summary statistics (i.e. <strong>Multi-finemap results</strong>).</li>
<li>
<code>locus_plot</code>: A nested list containing one or more zoomed views of locus plots.<br>
</li>
<li>
<code>LD_matrix</code>: The post-processed LD matrix used for fine-mapping.</li>
<li>
<code>LD_plot</code>: An LD plot (if used).</li>
<li>
<code>locus_dir</code>: Locus directory results are saved in.</li>
<li>
<code>arguments</code>: A record of the arguments supplied to <code>finemap_loci</code>.</li>
</ul>
<p>In addition, the following object summarizes the results from the locus-specific elements:<br>
- <code>merged_dat</code>: A merged <code>data.table</code> with all fine-mapping results from all loci.</p>
<div class="section level3">
<h3 id="multi-finemap-results-files">Multi-finemap results files<a class="anchor" aria-label="anchor" href="#multi-finemap-results-files"></a>
</h3>
<p>The main output of <code>echolocatoR</code> are the multi-finemap files (for example, <code><a href="https://rdrr.io/pkg/echodata/man/BST1.html" class="external-link">echodata::BST1</a></code>). They are stored in the locus-specific <em>Multi-finemap</em> subfolders.</p>
<div class="section level4">
<h4 id="column-descriptions">Column descriptions<a class="anchor" aria-label="anchor" href="#column-descriptions"></a>
</h4>
<ul>
<li>
<strong>Standardized GWAS/QTL summary statistics</strong>: e.g. <code>SNP</code>,<code>CHR</code>,<code>POS</code>,<code>Effect</code>,<code>StdErr</code>. See <code>?finemap_loci()</code> for descriptions of each.<br>
</li>
<li>
<strong>leadSNP</strong>: The designated proxy SNP per locus, which is the SNP with the smallest p-value by default.</li>
<li>
<strong>&lt;tool&gt;.CS</strong>: The 95% probability Credible Set (CS) to which a SNP belongs within a given fine-mapping tool’s results. If a SNP is not in any of the tool’s CS, it is assigned <code>NA</code> (or <code>0</code> for the purposes of plotting).<br>
</li>
<li>
<strong>&lt;tool&gt;.PP</strong>: The posterior probability that a SNP is causal for a given GWAS/QTL trait.<br>
</li>
<li>
<strong>Support</strong>: The total number of fine-mapping tools that include the SNP in its CS.</li>
<li>
<strong>Consensus_SNP</strong>: By default, defined as a SNP that is included in the CS of more than <code>N</code> fine-mapping tool(s), i.e. <code>Support&gt;1</code> (default: <code>N=1</code>).<br>
</li>
<li>
<strong>mean.PP</strong>: The mean SNP-wise PP across all fine-mapping tools used.</li>
<li>
<strong>mean.CS</strong>: If mean PP is greater than the 95% probability threshold (<code>mean.PP&gt;0.95</code>) then <code>mean.CS</code> is 1, else 0. This tends to be a very stringent threshold as it requires a high degree of agreement between fine-mapping tools.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="notes">Notes<a class="anchor" aria-label="anchor" href="#notes"></a>
</h3>
<ul>
<li>Separate multi-finemap files are generated for each LD reference panel used, which is included in the file name (e.g. <em>UKB_LD.Multi-finemap.tsv.gz</em>).</li>
<li>Each fine-mapping tool defines its CS and PP slightly differently, so please refer to the associated original publications for the exact details of how these are calculated (links provided above).</li>
</ul>
</div>
</details>
</div>
<div class="section level2">
<h2 id="fine-mapping-tools">Fine-mapping tools<a class="anchor" aria-label="anchor" href="#fine-mapping-tools"></a>
</h2>
<p>Fine-mapping functions are now implemented via <a href="https://github.com/RajLabMSSM/echofinemap" class="external-link"><code>echofinemap</code></a>:</p>
<details><ul>
<li>
<code>echolocatoR</code> will automatically check whether you have the necessary columns to run each tool you selected in <code>echolocatoR::finemap_loci(finemap_methods=...)</code>. It will remove any tools that for which there are missing necessary columns, and produces a message letting you know which columns are missing.</li>
<li>Note that some columns (e.g. <code>MAF</code>,<code>N</code>,<code>t-stat</code>) will be automatically inferred if missing.<br>
</li>
<li>For easy reference, we list the necessary columns here as well.<br>
See <code>?echodata::construct_colmap()</code> for descriptions of these columns.<br>
All methods require the columns: <code>SNP</code>,<code>CHR</code>,<code>POS</code>,<code>Effect</code>,<code>StdErr</code>
</li>
</ul></details><p><br></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fm_methods</span> <span class="op">&lt;-</span> <span class="fu">echofinemap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echofinemap/man/required_cols.html" class="external-link">required_cols</a></span><span class="op">(</span>add_versions <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                         embed_links <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                         verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Registered S3 method overwritten by 'GGally':</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="do">##   method from   </span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="do">##   +.gg   ggplot2</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">fm_methods</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="11%">
<col width="7%">
<col width="7%">
<col width="33%">
<col width="38%">
</colgroup>
<thead><tr class="header">
<th align="left">method</th>
<th align="left">required</th>
<th align="left">suggested</th>
<th align="left">source</th>
<th align="left">citation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ABF</td>
<td align="left">SNP, CHR….</td>
<td align="left"></td>
<td align="left"><a href="https://github.com/chr1swallace/coloc" class="external-link">source</a></td>
<td align="left"><a href="https://doi.org/10.1086%2F519024" class="external-link">cite</a></td>
</tr>
<tr class="even">
<td align="left">COJO_conditional</td>
<td align="left">SNP, CHR….</td>
<td align="left">Freq, P, N</td>
<td align="left"><a href="https://github.com/jianyangqt/gcta" class="external-link">source</a></td>
<td align="left"><a href="https://doi.org/10.1038/ng.2213" class="external-link">cite</a></td>
</tr>
<tr class="odd">
<td align="left">COJO_joint</td>
<td align="left">SNP, CHR….</td>
<td align="left">Freq, P, N</td>
<td align="left"><a href="https://github.com/jianyangqt/gcta" class="external-link">source</a></td>
<td align="left"><a href="https://doi.org/10.1038/ng.2213" class="external-link">cite</a></td>
</tr>
<tr class="even">
<td align="left">COJO_stepwise</td>
<td align="left">SNP, CHR….</td>
<td align="left">Freq, P, N</td>
<td align="left"><a href="https://github.com/jianyangqt/gcta" class="external-link">source</a></td>
<td align="left"><a href="https://doi.org/10.1038/ng.2213" class="external-link">cite</a></td>
</tr>
<tr class="odd">
<td align="left">FINEMAP</td>
<td align="left">SNP, CHR….</td>
<td align="left">A1, A2, ….</td>
<td align="left"><a href="http://www.christianbenner.com/" class="external-link">source</a></td>
<td align="left"><a href="https://doi.org/10.1093%2Fbioinformatics%2Fbtw018" class="external-link">cite</a></td>
</tr>
<tr class="even">
<td align="left">PAINTOR</td>
<td align="left">SNP, CHR….</td>
<td align="left">MAF</td>
<td align="left"><a href="https://github.com/gkichaev/PAINTOR_V3.0" class="external-link">source</a></td>
<td align="left"><a href="https://doi.org/10.1093/bioinformatics/btw615" class="external-link">cite</a></td>
</tr>
<tr class="odd">
<td align="left">POLYFUN_FINEMAP</td>
<td align="left">SNP, CHR….</td>
<td align="left">MAF, N</td>
<td align="left"><a href="https://github.com/omerwe/polyfun" class="external-link">source</a></td>
<td align="left"><a href="https://doi.org/10.1038/s41588-022-01036-9" class="external-link">cite</a></td>
</tr>
<tr class="even">
<td align="left">POLYFUN_SUSIE</td>
<td align="left">SNP, CHR….</td>
<td align="left">MAF, N</td>
<td align="left"><a href="https://github.com/omerwe/polyfun" class="external-link">source</a></td>
<td align="left"><a href="https://doi.org/10.1038/s41588-022-01036-9" class="external-link">cite</a></td>
</tr>
<tr class="odd">
<td align="left">SUSIE</td>
<td align="left">SNP, CHR….</td>
<td align="left">N</td>
<td align="left"><a href="https://github.com/stephenslab/susieR" class="external-link">source</a></td>
<td align="left"><a href="https://doi.org/10.1371/journal.pgen.1010299" class="external-link">cite</a></td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a>
</h2>
<p>Datasets are now stored/retrieved via the following <strong>echoverse</strong> subpackages:<br>
- <a href="https://github.com/RajLabMSSM/echodata" class="external-link"><code>echodata</code></a>: Pre-computed fine-mapping results. Also handles the semi-automated standardization of summary statistics.<br>
- <a href="https://github.com/RajLabMSSM/echoannot" class="external-link"><code>echoannot</code></a>: Annotates GWAS/QTL summary statistics using epigenomics, pre-compiled annotation matrices, and machine learning model predictions of variant-specific functional impacts.<br>
- <a href="https://github.com/RajLabMSSM/catalogueR" class="external-link"><code>catalogueR</code></a>: Large compendium of fully standardized e/s/t-QTL summary statistics.</p>
<p>For more detailed information about each dataset, use <code>?</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Examples ###</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RajLabMSSM/echoannot" class="external-link">echoannot</a></span><span class="op">)</span>   </span>
<span><span class="op">?</span><span class="va">NOTT_2019.interactome</span> <span class="co"># epigenomic annotations</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RajLabMSSM/echodata" class="external-link">echodata</a></span><span class="op">)</span> </span>
<span><span class="op">?</span><span class="va">BST1</span> <span class="co"># fine-mapping results </span></span></code></pre></div>
<details><div class="section level3">
<h3 id="mungesumstats">
<a href="https://github.com/neurogenomics/MungeSumstats" class="external-link"><strong><code>MungeSumstats</code></strong></a>:<a class="anchor" aria-label="anchor" href="#mungesumstats"></a>
</h3>
<ul>
<li>You can search, import, and standardize any GWAS in the <a href="https://gwas.mrcieu.ac.uk/" class="external-link"><em>Open GWAS</em></a> database via <a href="https://github.com/neurogenomics/MungeSumstats" class="external-link"><code>MungeSumstats</code></a>, specifically the functions <code>find_sumstats</code> and <code>import_sumstats</code>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="cataloguer-qtls">
<a href="https://github.com/RajLabMSSM/catalogueR" class="external-link"><code>catalogueR</code></a>: QTLs<a class="anchor" aria-label="anchor" href="#cataloguer-qtls"></a>
</h3>
<div class="section level4">
<h4 id="eqtl-catalogue-cataloguereqtl_cataloguequery">
<a href="https://www.ebi.ac.uk/eqtl/" class="external-link">eQTL Catalogue</a>: <code>catalogueR::eQTL_Catalogue.query()</code>
<a class="anchor" aria-label="anchor" href="#eqtl-catalogue-cataloguereqtl_cataloguequery"></a>
</h4>
<ul>
<li>API access to full summary statistics from many standardized e/s/t-QTL datasets.<br>
</li>
<li>Data access and colocalization tests facilitated through the <a href="https://github.com/RajLabMSSM/catalogueR" class="external-link"><code>catalogueR</code></a> R package.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="echodata-fine-mapping-results">
<a href="https://github.com/RajLabMSSM/catalogueR" class="external-link"><code>echodata</code></a>: fine-mapping results<a class="anchor" aria-label="anchor" href="#echodata-fine-mapping-results"></a>
</h3>
<div class="section level4">
<h4 id="echolocator-fine-mapping-portal-pre-computed-fine-mapping-results">
<a href="https://rajlab.shinyapps.io/Fine_Mapping_Shiny" class="external-link"><strong><em>echolocatoR Fine-mapping Portal</em></strong></a>: pre-computed fine-mapping results<a class="anchor" aria-label="anchor" href="#echolocator-fine-mapping-portal-pre-computed-fine-mapping-results"></a>
</h4>
<ul>
<li>You can visit the <em>echolocatoR Fine-mapping Portal</em> to interactively visualize and download pre-computed fine-mapping results across a variety of phenotypes.</li>
<li>This data can be searched and imported programmatically using <code><a href="https://rdrr.io/pkg/echodata/man/portal_query.html" class="external-link">echodata::portal_query()</a></code>.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="echoannot-epigenomic--genome-wide-annotations">
<a href="https://github.com/RajLabMSSM/echoannot" class="external-link"><code>echoannot</code></a>: Epigenomic &amp; genome-wide annotations<a class="anchor" aria-label="anchor" href="#echoannot-epigenomic--genome-wide-annotations"></a>
</h3>
<div class="section level4">
<h4 id="nott-et-al-2019-echoannotnott2019_">
<a href="https://science.sciencemag.org/content/366/6469/1134.abstract" class="external-link">Nott et al. (2019)</a>: <code>echoannot::NOTT2019_*()</code>
<a class="anchor" aria-label="anchor" href="#nott-et-al-2019-echoannotnott2019_"></a>
</h4>
<ul>
<li>Data from this publication contains results from cell type-specific (neurons, oligodendrocytes, astrocytes, microglia, &amp; peripheral myeloid cells) epigenomic assays (H3K27ac, ATAC, H3K4me3) from <em>ex vivo</em> pediatric human brain tissue.</li>
</ul>
</div>
<div class="section level4">
<h4 id="corces-et-al2020-echoannotcorces2020_">
<a href="https://doi.org/10.1038/s41588-020-00721-x" class="external-link">Corces et al.2020</a>: <code>echoannot::CORCES2020_*()</code>
<a class="anchor" aria-label="anchor" href="#corces-et-al2020-echoannotcorces2020_"></a>
</h4>
<ul>
<li>Data from this publication contains results from single-cell and bulk chromatin accessibility assays ([sc]ATAC-seq) and chromatin interactions ( <a href="https://ay-lab.github.io/FitHiChIP/" class="external-link"><code>FitHiChIP</code></a>) from <em>postmortem</em> adult human brain tissue.</li>
</ul>
</div>
<div class="section level4">
<h4 id="xgr-echoannotxgr_download_and_standardize">
<a href="http://xgr.r-forge.r-project.org" class="external-link">XGR</a>: <code>echoannot::XGR_download_and_standardize()</code>
<a class="anchor" aria-label="anchor" href="#xgr-echoannotxgr_download_and_standardize"></a>
</h4>
<ul>
<li>API access to a diverse library of cell type/line-specific epigenomic (e.g. <strong>ENCODE</strong>) and other genome-wide annotations.</li>
</ul>
</div>
<div class="section level4">
<h4 id="roadmap-echoannotroadmap_query">
<a href="http://www.roadmapepigenomics.org" class="external-link">Roadmap</a>: <code>echoannot::ROADMAP_query()</code>
<a class="anchor" aria-label="anchor" href="#roadmap-echoannotroadmap_query"></a>
</h4>
<ul>
<li>API access to cell type-specific epigenomic data.</li>
</ul>
</div>
<div class="section level4">
<h4 id="biomart-echoannotannotate_snps">
<a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html" class="external-link">biomaRt</a>: <code>echoannot::annotate_snps()</code>
<a class="anchor" aria-label="anchor" href="#biomart-echoannotannotate_snps"></a>
</h4>
<ul>
<li>API access to various genome-wide SNP annotations (e.g. missense, nonsynonmous, intronic, enhancer).</li>
</ul>
</div>
<div class="section level4">
<h4 id="haplor-echoannotannotate_snps">
<a href="https://cran.r-project.org/web/packages/haploR/vignettes/haplor-vignette.html" class="external-link">HaploR</a>: <code>echoannot::annotate_snps()</code>
<a class="anchor" aria-label="anchor" href="#haplor-echoannotannotate_snps"></a>
</h4>
<ul>
<li>API access to known per-SNP QTL and epigenomic data hits.</li>
</ul>
</div>
</div>
</details>
</div>
<div class="section level2">
<h2 id="enrichment-tools">Enrichment tools<a class="anchor" aria-label="anchor" href="#enrichment-tools"></a>
</h2>
<p>Annotation enrichment functions are now implemented via <a href="https://github.com/RajLabMSSM/echoannot" class="external-link"><code>echoannot</code></a>:</p>
<details><div class="section level3">
<h3 id="implemented">Implemented<a class="anchor" aria-label="anchor" href="#implemented"></a>
</h3>
<div class="section level4">
<h4 id="xgr-echoannotxgr_enrichment">
<a href="http://xgr.r-forge.r-project.org" class="external-link">XGR</a>: <code>echoannot::XGR_enrichment()</code>
<a class="anchor" aria-label="anchor" href="#xgr-echoannotxgr_enrichment"></a>
</h4>
<ul>
<li>Binomial enrichment tests between customisable foreground and background SNPs.</li>
</ul>
</div>
<div class="section level4">
<h4 id="motifbreakr-echoannotmotifbreakr">
<a href="https://github.com/Simon-Coetzee/motifBreakR" class="external-link">motifbreakR</a>: <code>echoannot::MOTIFBREAKR()</code>
<a class="anchor" aria-label="anchor" href="#motifbreakr-echoannotmotifbreakr"></a>
</h4>
<ul>
<li>Identification of transcript factor binding motifs (TFBM) and prediction of SNP disruption to said motifs.</li>
<li>Includes a comprehensive list of TFBM databases via <a href="https://bioconductor.org/packages/release/bioc/html/MotifDb.html" class="external-link">MotifDB</a> (9,900+ annotated position frequency matrices from 14 public sources, for multiple organisms).</li>
</ul>
</div>
<div class="section level4">
<h4 id="regioner-echoannottest_enrichment">
<a href="http://bioconductor.org/packages/release/bioc/html/regioneR.html" class="external-link">regioneR</a>: <code>echoannot::test_enrichment()</code>
<a class="anchor" aria-label="anchor" href="#regioner-echoannottest_enrichment"></a>
</h4>
<ul>
<li>Iterative pairwise permutation testing of overlap between all combinations of two <a href="https://biodatascience.github.io/compbio/bioc/GRL.html" class="external-link"><code>GRangesList</code></a> objects.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="under-construction">Under construction<a class="anchor" aria-label="anchor" href="#under-construction"></a>
</h3>
<div class="section level4">
<h4 id="garfield">
<a href="https://www.bioconductor.org/packages/release/bioc/html/garfield.html" class="external-link">GARFIELD</a><a class="anchor" aria-label="anchor" href="#garfield"></a>
</h4>
<ul>
<li>Genomic enrichment with LD-informed heuristics.</li>
</ul>
</div>
<div class="section level4">
<h4 id="goshifter">
<a href="https://github.com/immunogenomics/goshifter" class="external-link">GoShifter</a><a class="anchor" aria-label="anchor" href="#goshifter"></a>
</h4>
<ul>
<li>LD-informed iterative enrichment analysis.</li>
</ul>
</div>
<div class="section level4">
<h4 id="s-ldsc">
<a href="https://www.nature.com/articles/ng.3954" class="external-link">S-LDSC</a><a class="anchor" aria-label="anchor" href="#s-ldsc"></a>
</h4>
<ul>
<li>Genome-wide stratified LD score regression.</li>
<li>Inlccles 187-annotation baseline model from <a href="https://www.nature.com/articles/s41588-018-0231-8" class="external-link">Gazal et al. 2018</a>.<br>
</li>
<li>You can alternatively supply a custom annotations matrix.</li>
</ul>
</div>
</div>
</details>
</div>
<div class="section level2">
<h2 id="ld-reference-panels">LD reference panels<a class="anchor" aria-label="anchor" href="#ld-reference-panels"></a>
</h2>
<p>LD reference panels are now queried/processed by <a href="https://github.com/RajLabMSSM/echoLD" class="external-link"><code>echoLD</code></a>, specifically the function <code>get_LD()</code>:</p>
<details><div class="section level3">
<h3 id="uk-biobank">
<a href="https://www.ukbiobank.ac.uk" class="external-link">UK Biobank</a><a class="anchor" aria-label="anchor" href="#uk-biobank"></a>
</h3>
</div>
<div class="section level3">
<h3 id="id_1000-genomes-phase-1">
<a href="https://www.internationalgenome.org" class="external-link">1000 Genomes Phase 1</a><a class="anchor" aria-label="anchor" href="#id_1000-genomes-phase-1"></a>
</h3>
</div>
<div class="section level3">
<h3 id="id_1000-genomes-phase-3">
<a href="https://www.internationalgenome.org" class="external-link">1000 Genomes Phase 3</a><a class="anchor" aria-label="anchor" href="#id_1000-genomes-phase-3"></a>
</h3>
</div>
<div class="section level3">
<h3 id="custom-ld-panel">Custom LD panel:<a class="anchor" aria-label="anchor" href="#custom-ld-panel"></a>
</h3>
<ul>
<li>From user-supplied VCFs</li>
</ul>
</div>
<div class="section level3">
<h3 id="custom-ld-panel-1">Custom LD panel<a class="anchor" aria-label="anchor" href="#custom-ld-panel-1"></a>
</h3>
<ul>
<li>From user-supplied precomputed LD matrices</li>
</ul>
</div>
</details>
</div>
<div class="section level2">
<h2 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h2>
<p>Plotting functions are now implemented via:<br>
- <a href="https://github.com/RajLabMSSM/echoplot" class="external-link"><code>echoplot</code></a>: Multi-track locus plots with GWAS, fine-mapping results, and functional annotations (<code>plot_locus()</code>). Can also plot multi-GWAS/QTL and multi-ancestry results (<code>plot_locus_multi()</code>).<br>
- <a href="https://github.com/RajLabMSSM/echoannot" class="external-link"><code>echoannot</code></a>: Study-level summary plots showing aggregted info across many loci at once (<code><a href="https://rdrr.io/pkg/echoannot/man/super_summary_plot.html" class="external-link">super_summary_plot()</a></code>).<br>
- <a href="https://github.com/RajLabMSSM/echoLD" class="external-link"><code>echoLD</code></a>: Plot an LD matrix using one of several differnt plotting methods (<code>plot_LD()</code>).</p>
</div>
<div class="section level2">
<h2 id="tabix-queries">Tabix queries<a class="anchor" aria-label="anchor" href="#tabix-queries"></a>
</h2>
<p>All queries of <a href="http://www.htslib.org/doc/tabix.html" class="external-link"><code>tabix</code></a>-indexed files (for rapid data subset extraction) are implemented via <a href="https://github.com/RajLabMSSM/echotabix" class="external-link"><code>echotabix</code></a>.</p>
<details><ul>
<li>
<code><a href="https://rdrr.io/pkg/echotabix/man/convert_and_query.html" class="external-link">echotabix::convert_and_query()</a></code> detects whether the GWAS summary statistics file you provided is already <code>tabix</code>-indexed, and it not, automatically performs all steps necessary to convert it (sorting, <code>bgzip</code>-compression, indexing) across a wide variety of scenarios.<br>
</li>
<li>
<code><a href="https://rdrr.io/pkg/echotabix/man/query.html" class="external-link">echotabix::query()</a></code> contains many different methods for making tabix queries (e.g. <code>Rtracklayer</code>,<code>echoconda</code>,<code>VariantAnnotation</code>,<code>seqminer</code>), each of which fail in certain circumstances. To avoid this, <code>query()</code> automatically selects the method that will work for the particular file being queried and your machine’s particular versions of R/Bioconductor/OS, taking the guesswork and troubleshooting out of <code>tabix</code> queries.</li>
</ul></details>
</div>
<div class="section level2">
<h2 id="downloads">Downloads<a class="anchor" aria-label="anchor" href="#downloads"></a>
</h2>
<p>Single- and multi-threaded downloads are now implemented via <a href="https://github.com/RajLabMSSM/downloadR" class="external-link"><code>downloadR</code></a>.</p>
<details><ul>
<li>Multi-threaded downloading is performed using <a href="https://github.com/axel-download-accelerator/axel" class="external-link"><code>axel</code></a>, and is particularly useful for speeding up downloads of large files.</li>
<li>
<code>axel</code> is installed via the official <em>echoverse</em> <a href="https://docs.conda.io/en/latest/" class="external-link">conda</a> environment: “echoR_mini”. This environment is automatically created by the function <code><a href="https://rdrr.io/pkg/echoconda/man/yaml_to_env.html" class="external-link">echoconda::yaml_to_env()</a></code> when needed.</li>
</ul></details><hr>
</div>
<div class="section level1">
<div class="page-header"><h1 id="developer">Developer<a class="anchor" aria-label="anchor" href="#developer"></a>
</h1></div>
<p><a href="https://bschilder.github.io/BMSchilder/" target="_blank" class="external-link">Brian M. Schilder, Bioinformatician II</a><br><a href="https://rajlab.org" target="_blank" class="external-link">Raj Lab</a><br><a href="https://icahn.mssm.edu/about/departments/neuroscience" target="_blank" class="external-link">Department of Neuroscience, Icahn School of Medicine at Mount Sinai</a></p>
<hr>
</div>
<div class="section level1">
<h1 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h1>
<details><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.5 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] rmarkdown_2.17</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] utf8_1.2.2                  reticulate_1.26            </span></span>
<span><span class="co">##   [3] R.utils_2.12.1              tidyselect_1.2.0           </span></span>
<span><span class="co">##   [5] RSQLite_2.2.18              AnnotationDbi_1.60.0       </span></span>
<span><span class="co">##   [7] htmlwidgets_1.5.4           grid_4.2.1                 </span></span>
<span><span class="co">##   [9] BiocParallel_1.32.0         XGR_1.1.8                  </span></span>
<span><span class="co">##  [11] munsell_0.5.0               codetools_0.2-18           </span></span>
<span><span class="co">##  [13] interp_1.1-3                DT_0.26                    </span></span>
<span><span class="co">##  [15] colorspace_2.0-3            OrganismDbi_1.40.0         </span></span>
<span><span class="co">##  [17] Biobase_2.58.0              filelock_1.0.2             </span></span>
<span><span class="co">##  [19] highr_0.9                   knitr_1.40                 </span></span>
<span><span class="co">##  [21] supraHex_1.35.0             rstudioapi_0.14            </span></span>
<span><span class="co">##  [23] stats4_4.2.1                DescTools_0.99.47          </span></span>
<span><span class="co">##  [25] MatrixGenerics_1.10.0       GenomeInfoDbData_1.2.9     </span></span>
<span><span class="co">##  [27] mixsqp_0.3-43               bit64_4.0.5                </span></span>
<span><span class="co">##  [29] echoconda_0.99.8            rprojroot_2.0.3            </span></span>
<span><span class="co">##  [31] basilisk_1.9.12             vctrs_0.5.0                </span></span>
<span><span class="co">##  [33] generics_0.1.3              xfun_0.34                  </span></span>
<span><span class="co">##  [35] biovizBase_1.46.0           BiocFileCache_2.6.0        </span></span>
<span><span class="co">##  [37] R6_2.5.1                    GenomeInfoDb_1.34.1        </span></span>
<span><span class="co">##  [39] AnnotationFilter_1.22.0     bitops_1.0-7               </span></span>
<span><span class="co">##  [41] cachem_1.0.6                reshape_0.8.9              </span></span>
<span><span class="co">##  [43] DelayedArray_0.24.0         assertthat_0.2.1           </span></span>
<span><span class="co">##  [45] BiocIO_1.8.0                scales_1.2.1               </span></span>
<span><span class="co">##  [47] nnet_7.3-18                 rootSolve_1.8.2.3          </span></span>
<span><span class="co">##  [49] gtable_0.3.1                lmom_2.9                   </span></span>
<span><span class="co">##  [51] ggbio_1.46.0                ensembldb_2.22.0           </span></span>
<span><span class="co">##  [53] rlang_1.0.6                 echodata_0.99.15           </span></span>
<span><span class="co">##  [55] splines_4.2.1               lazyeval_0.2.2             </span></span>
<span><span class="co">##  [57] rtracklayer_1.58.0          dichromat_2.0-0.1          </span></span>
<span><span class="co">##  [59] hexbin_1.28.2               checkmate_2.1.0            </span></span>
<span><span class="co">##  [61] BiocManager_1.30.19         yaml_2.3.6                 </span></span>
<span><span class="co">##  [63] reshape2_1.4.4              backports_1.4.1            </span></span>
<span><span class="co">##  [65] snpStats_1.47.1             GenomicFeatures_1.50.2     </span></span>
<span><span class="co">##  [67] ggnetwork_0.5.10            Hmisc_4.7-1                </span></span>
<span><span class="co">##  [69] RBGL_1.74.0                 tools_4.2.1                </span></span>
<span><span class="co">##  [71] ggplot2_3.3.6               ellipsis_0.3.2             </span></span>
<span><span class="co">##  [73] RColorBrewer_1.1-3          proxy_0.4-27               </span></span>
<span><span class="co">##  [75] BiocGenerics_0.44.0         coloc_5.1.0.1              </span></span>
<span><span class="co">##  [77] Rcpp_1.0.9                  plyr_1.8.7                 </span></span>
<span><span class="co">##  [79] base64enc_0.1-3             progress_1.2.2             </span></span>
<span><span class="co">##  [81] zlibbioc_1.44.0             purrr_0.3.5                </span></span>
<span><span class="co">##  [83] RCurl_1.98-1.9              basilisk.utils_1.9.4       </span></span>
<span><span class="co">##  [85] prettyunits_1.1.1           rpart_4.1.19               </span></span>
<span><span class="co">##  [87] deldir_1.0-6                viridis_0.6.2              </span></span>
<span><span class="co">##  [89] S4Vectors_0.36.0            cluster_2.1.4              </span></span>
<span><span class="co">##  [91] SummarizedExperiment_1.28.0 ggrepel_0.9.1              </span></span>
<span><span class="co">##  [93] fs_1.5.2                    crul_1.3                   </span></span>
<span><span class="co">##  [95] magrittr_2.0.3              data.table_1.14.4          </span></span>
<span><span class="co">##  [97] echotabix_0.99.8            dnet_1.1.7                 </span></span>
<span><span class="co">##  [99] openxlsx_4.2.5.1            mvtnorm_1.1-3              </span></span>
<span><span class="co">## [101] ProtGenerics_1.30.0         matrixStats_0.62.0         </span></span>
<span><span class="co">## [103] patchwork_1.1.2             hms_1.1.2                  </span></span>
<span><span class="co">## [105] evaluate_0.17               XML_3.99-0.12              </span></span>
<span><span class="co">## [107] jpeg_0.1-9                  readxl_1.4.1               </span></span>
<span><span class="co">## [109] IRanges_2.32.0              gridExtra_2.3              </span></span>
<span><span class="co">## [111] compiler_4.2.1              biomaRt_2.54.0             </span></span>
<span><span class="co">## [113] tibble_3.1.8                crayon_1.5.2               </span></span>
<span><span class="co">## [115] R.oo_1.25.0                 htmltools_0.5.3            </span></span>
<span><span class="co">## [117] echoannot_0.99.10           tzdb_0.3.0                 </span></span>
<span><span class="co">## [119] Formula_1.2-4               tidyr_1.2.1                </span></span>
<span><span class="co">## [121] expm_0.999-6                Exact_3.2                  </span></span>
<span><span class="co">## [123] DBI_1.1.3                   dbplyr_2.2.1               </span></span>
<span><span class="co">## [125] MASS_7.3-58.1               rappdirs_0.3.3             </span></span>
<span><span class="co">## [127] boot_1.3-28                 dlstats_0.1.5              </span></span>
<span><span class="co">## [129] Matrix_1.5-1                badger_0.2.1               </span></span>
<span><span class="co">## [131] readr_2.1.3                 piggyback_0.1.4            </span></span>
<span><span class="co">## [133] cli_3.4.1                   R.methodsS3_1.8.2          </span></span>
<span><span class="co">## [135] parallel_4.2.1              echofinemap_0.99.4         </span></span>
<span><span class="co">## [137] igraph_1.3.5                GenomicRanges_1.50.0       </span></span>
<span><span class="co">## [139] pkgconfig_2.0.3             rvcheck_0.2.1              </span></span>
<span><span class="co">## [141] GenomicAlignments_1.34.0    dir.expiry_1.5.1           </span></span>
<span><span class="co">## [143] RCircos_1.2.2               foreign_0.8-83             </span></span>
<span><span class="co">## [145] osfr_0.2.9                  xml2_1.3.3                 </span></span>
<span><span class="co">## [147] XVector_0.38.0              yulab.utils_0.0.5          </span></span>
<span><span class="co">## [149] echoLD_0.99.8               stringr_1.4.1              </span></span>
<span><span class="co">## [151] VariantAnnotation_1.44.0    digest_0.6.30              </span></span>
<span><span class="co">## [153] graph_1.76.0                httpcode_0.3.0             </span></span>
<span><span class="co">## [155] Biostrings_2.66.0           cellranger_1.1.0           </span></span>
<span><span class="co">## [157] htmlTable_2.4.1             gld_2.6.6                  </span></span>
<span><span class="co">## [159] restfulr_0.0.15             curl_4.3.3                 </span></span>
<span><span class="co">## [161] Rsamtools_2.14.0            rjson_0.2.21               </span></span>
<span><span class="co">## [163] lifecycle_1.0.3             nlme_3.1-160               </span></span>
<span><span class="co">## [165] jsonlite_1.8.3              desc_1.4.2                 </span></span>
<span><span class="co">## [167] viridisLite_0.4.1           BSgenome_1.66.1            </span></span>
<span><span class="co">## [169] fansi_1.0.3                 downloadR_0.99.5           </span></span>
<span><span class="co">## [171] pillar_1.8.1                susieR_0.12.27             </span></span>
<span><span class="co">## [173] lattice_0.20-45             GGally_2.1.2               </span></span>
<span><span class="co">## [175] KEGGREST_1.38.0             fastmap_1.1.0              </span></span>
<span><span class="co">## [177] httr_1.4.4                  survival_3.4-0             </span></span>
<span><span class="co">## [179] glue_1.6.2                  zip_2.2.2                  </span></span>
<span><span class="co">## [181] png_0.1-7                   bit_4.0.4                  </span></span>
<span><span class="co">## [183] Rgraphviz_2.41.2            class_7.3-20               </span></span>
<span><span class="co">## [185] stringi_1.7.8               blob_1.2.3                 </span></span>
<span><span class="co">## [187] latticeExtra_0.6-30         memoise_2.0.1              </span></span>
<span><span class="co">## [189] dplyr_1.0.10                irlba_2.3.5.1              </span></span>
<span><span class="co">## [191] e1071_1.7-12                ape_5.6-2</span></span></code></pre>
</details><p><br></p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/RajLabMSSM/echolocatoR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/RajLabMSSM/echolocatoR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing echolocatoR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Brian Schilder <br><small class="roles"> Maintainer </small> <a href="https://orcid.org/0000-0001-5949-2191" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jack Humphrey <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6274-6620" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Towfique Raj <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-9355-5704" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
