---
title: "Getting Started" 
author: "<h4>Author: <i>Brian M. Schilder</i></h4>" 
date: "<h4>Updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output:
  BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{echolocatoR} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---


```{r setup}
library(echolocatoR) 
```

# Full pipeline  

All examples below use data from the Parkinson's disease GWAS by Nalls et al. (2019).  

## Prepare `top_SNPs` data.frame   

* To enable rapid fine-mapping of many loci, you can create a `top_SNPs` data.frame  
which contains the position of the lead/index SNP within each locus.
* `finemap_loci()` (see next step) will then use this info to extract subsets of the   
full GWAS/QTL summary statistics using windows centered on each lead/index SNP.
* The `topSS` argument can either be a data.frame, or a path to a topSS file saved somehwere. Most common tabular data formats (e.g. .tsv, .csv, .xlsx) are accepted.  



```{r  Prepare `top_SNPs` data.frame} 
topSS <- echodata::topSNPs_Nalls2019_raw
top_SNPs <- echodata::import_topSNPs( 
  topSS = topSS,
  munge = TRUE,
  pval_col="P, all studies", 
  effect_col="Beta, all studies", 
  gene_col="Nearest Gene", 
  locus_col = "Nearest Gene",
  grouping_vars = c("Locus"),
  remove_variants = "rs34637584") 
head(top_SNPs)
```



## Path to full summary stats file  

* Since a full GWAS summary stats file would be too large to include within *echolocatoR*,  
we instead provide an example subset of the full summary stats.  

* To simulate how you'd actually use your own full summary stats file, we will save our example dataset to your computer (you can change the path to wherever you like). 

* We highly recommend munging your full summary stats using the Bioconductor package [`MungeSumstats`](https://github.com/neurogenomics/MungeSumstats) first. It's easy to use and very robust. It also means you don't have to provide most column mapping arguments in `finemap_loci` when `munged=TRUE`. 

Here's an example of how to munge your full summary stats file: 

```
fullSS_path <- echodata::example_fullSS(munged = FALSE)
fullSS_path <- MungeSumstats::format_sumstats(path = fullSS_path, ref_genome = "GRCH37")
```

We have already munged the following example summary stats for you.

```{r fullSS} 
fullSS_path <- echodata::example_fullSS(dataset = "Nalls2019")
```

## Run fine-mapping pipeline  

For a full description of all arguments, see `?finemap_loci`.  

Here are some key arguments:  

* *results_dir*: Where you want to store all of your results.  
* *finemap_methods*: Which fine-mapping methods you want to run (currently includes ABF, FINEMAP, SUSIE, POLYFUN_SUSIE, and COJO).  
* *bp_distance*: Controls window size. Specifically, `bp_distance` is the number of basepairs upstream/downstream you want to extract for each locus. For example, if you want a 2Mb window (+/- 1Mb from the lead/index SNP in `top_SNPs`), set `bp_distance=1e+06`.  
* *plot.zoom*: Zoom in/out from the center of each locus when producing the multiview plot.  
You can adjust this separately from `bp_distance` so that you don't have rerun the whole pipeline each time (locus subsets, LD matrices, and fine-mapping results are all automatically saved in locus-specific folders).  

**NOTE**: This example assumes you have already installed tabix (via htslib) and bcftoools.

**WARNING**: Please use the full absolute paths (instead of relative paths) wherever possible (e.g. `results_dir`). This is especially important for the tool *FINEMAP*.

```{r Run fine-mapping pipeline}
results <- echolocatoR::finemap_loci(
 fullSS_path = fullSS_path,
 top_SNPs = top_SNPs,
 loci = c("BST1","MEX3C"),
 dataset_name = "Nalls23andMe_2019",
 fullSS_genome_build = "hg19",
 bp_distance = 250000,
 munged = TRUE)
```

<hr>

# Session info  

<details>

```{r Session Info, attr.output='style="max-height: 200px;"'}
utils::sessionInfo()
```

</details>

<br>

