---
title: "echolocatoR Fine-mapping Portal"
author: "<h4>Author: <i>Brian M. Schilder</i></h4>" 
date: "<h4>Updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output:
  BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{echolocatoR Fine-mapping Portal} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---
 
```{r setup, include=TRUE}
library(echolocatoR)   
```

Here we take advantage of the fine-mapping results files already available on the [echolocatoR Fine-mapping Portal](https://rajlab.shinyapps.io/Fine_Mapping_Shiny).

# Download data  

## Select loci 

```{r loci}
## Limit the number of loci for demo purposes
loci <- echodata::topSNPs_Nalls2019$Locus[1:3] 
```

## Portal query 

We can query multiple studies, loci, and even data types at once. 
When `as_datatable=TRUE`, all file paths and metadata are conveniently 
organized into a `data.table.`

```{r Download data}  
results_dir <- tempdir()
local_files <- echodata::portal_query(dataset_types="GWAS",
                                      phenotypes = c("parkinson"),
                                      file_types = c("multi_finemap","LD"), 
                                      loci = loci,
                                      LD_panels=c("UKB"),
                                      results_dir = results_dir, 
                                      as_datatable = TRUE) 
```

### Merge fine-mapping results  

Next, we can gather all of the fine-mapping results generated by
`finemap_loci()` previously.  
`merge_finemapping_results` recursively searches for the correct files
within a hierarchical folder structure and imports only the multi-finemap files. 

```{r merge_finemapping_results()} 
merged_DT <- echodata::merge_finemapping_results(dataset = results_dir,   
                                                 minimum_support = 0,
                                                 include_leadSNPs = TRUE,
                                                 consensus_thresh = 2)
echodata::results_report(merged_DT)
```

## Import LD 

Next, we import the a subset of the LD matrices for only the lead SNP.

```{r}
ld_files <- local_files[file_type=="LD",]
ld_matrices <- lapply(stats::setNames(ld_files$local_file, 
                                      ld_files$locus),
                      function(x){
  data.table::fread(x)
}) 

knitr::kable(head(ld_matrices$ASXL3))
```



# Plot locus

Now let's plot one locus as an example. 

```{r LRRK2, eval=FALSE} 
locus <- unique(merged_DT$Locus)[1] # Pick the first locus
dat <- merged_DT[Locus==locus,] 
LD_matrix = ld_matrices[[locus]]
locus_dir <- file.path(tempdir(),locus)

plt <- echoplot::plot_locus(dat = dat,   
                    LD_matrix = LD_matrix,
                    LD_reference = "UKB",
                    locus_dir = locus_dir, 
                      
                    nott_epigenome = TRUE,   
                    nott_regulatory_rects = TRUE, 
                    nott_show_placseq = TRUE,
                    zoom = c("20x")) 
```

# Session info

<details> 

```{r Session Info, attr.output='style="max-height: 400px;"'}
utils::sessionInfo()
```

</details>

<br>

