---
title: "PD loci" 
author: "<h4>Author: <i>Brian M. Schilder</i></h4>" 
date: "<h4>Updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output:
  BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{PD_loci} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = T, error = F, dpi=300}
root.dir <- tempdir()
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  root.dir = root.dir, 
  fig.height = 12,
  fig.width = 10
)  
knitr::opts_knit$set(
  root.dir = root.dir
  )
# knitr::opts_chunk$get("root.dir")
# devtools::build_vignettes(quiet = F)
```

```{r setup, eval=F}
library(echolocatoR) 
library(data.table)

results_dir <- file.path(root.dir,"results")
```


Here we take advantage of the fine-mapping results files already available on the [echolocatoR Fine-mapping Portal](https://rajlab.shinyapps.io/Fine_Mapping_Shiny).

# Download data  


*echolocatoR* includes some functions to search for and extract files from GitHub repos. In this case we'll get files from the [echolocatoR Fine-mapping Portal](https://rajlab.shinyapps.io/Fine_Mapping_Shiny), but these functions can be used on any public GitHub repo.  

* `GITHUB.list_files()`: Searches the repo for files matching your regex query.
* `GITHUB.download_files()`: Not only does this download your files, but it reconstructs the original folder structure they were found in. This is useful for *echolocatoR* output files, which are automatically organized in a hierarchical folder structure (e.g. *results/GWAS/Nalls23andMe_2019*).


```{r locus_list}
locus_list <- c("LRRK2","MBNL2","DYRK1A","FCGR2A","MED12L")  
```

## Multi-finemap data  

First, we'll download the multi-finemap results files.  

```{r multi_finemap, eval=F}
local_finemap <- GITHUB.portal_query(results_dir = results_dir, 
                                     dataset_types="GWAS",
                                     phenotypes = c("parkinson"),
                                     file_types = "multi_finemap",
                                     loci = locus_list,
                                     LD_panels=c("UKB"))  
```

## LD data

Next, we'll download the files containing LD with the lead GWAS SNP. 

```{r LD, eval=F}
local_ld <- GITHUB.portal_query(results_dir = results_dir, 
                                 dataset_types="GWAS",
                                 phenotypes = c("parkinson"),
                                 file_types = "LD",
                                 loci = locus_list,
                                 LD_panels=c("UKB"))  
```

## Data dictionary

Now let's put them in a data dictionary for easy retrieval later.
This way, you can simply find the right file path my locus name.

This function also infers a number of other useful variables from this data. 

```{r Data dictionary, eval=F}
named_list <- list("finemap"=local_finemap,
                   "LD"=local_ld)
data_dict <- GITHUB.make_data_dict(named_list)
```


```{r LRRK2, attr.output='style="max-height: 200px;"', eval=F} 
locus <- "LRRK2"

track_order <- names(PLOT.heights_dict())
track_order <- R.utils::insert(track_order[track_order!="Genes"], 1, "Genes")

LRRK2 <- PLOT.locus(finemap_dat=fread(data_dict$finemap[[locus]]),   
                    LD_matrix=fread(data_dict$LD[[locus]]),
                    LD_reference="UKB",
                    locus_dir=data_dict$locus_dir[[locus]], 
                    
                    max_transcripts = 3,
                    Nott_binwidth = 200,
                    Nott_epigenome=T,   
                    Nott_regulatory_rects = T, 
                    Nott_show_placseq = T,
                    
                    save_plot=F, 
                    return_list=T, 
                    zoom_exceptions_str = "*full window$|zoom_polygon|Genes",
                    track_order = track_order,
                    plot.zoom=c("1x","20x","25x")) 
```

```{r MBNL2, attr.output='style="max-height: 200px;"', eval=F} 
locus <- "MBNL2"

track_order <- names(PLOT.heights_dict())
track_order <- R.utils::insert(track_order[track_order!="Genes"], 1, "Genes")


MBNL2 <- PLOT.locus(finemap_dat=fread(data_dict$finemap[[locus]]),   
                    LD_matrix=fread(data_dict$LD[[locus]]),
                    LD_reference="UKB",
                    locus_dir=data_dict$locus_dir[[locus]], 
                    
                    max_transcripts = 1,
                    Nott_epigenome=T,  
                    Nott_binwidth = 200,
                    Nott_regulatory_rects = T, 
                    Nott_show_placseq = T,
                    
                    save_plot=F, 
                    return_list=T,
                    zoom_exceptions_str = "*full window$|zoom_polygon|Genes",
                    track_order = track_order,
                    plot.zoom=c("1x","10x","20x","25x"))
```
 
 
```{r DYRK1A, attr.output='style="max-height: 200px;"', eval=F} 
locus <- "DYRK1A"

DYRK1A <- PLOT.locus(finemap_dat=fread(data_dict$finemap[[locus]]),   
                    LD_matrix=fread(data_dict$LD[[locus]]),
                    LD_reference="UKB",
                    locus_dir=data_dict$locus_dir[[locus]], 
                    
                    max_transcripts = 3,
                    Nott_epigenome=T,  
                    Nott_binwidth = 500,
                    Nott_regulatory_rects = T, 
                    Nott_show_placseq = T,
                    
                    save_plot=F, 
                    return_list=T,
                    plot.zoom=c("25x"))
```
 
This region has a lot of genes so we're going to subset the main view a bit. 
 
```{r FCGR2A, attr.output='style="max-height: 200px;"', eval=F} 
locus <- "FCGR2A"
# 
# finemap_dat <- fread(data_dict$finemap[[locus]])
# xlims <- PLOT.get_window_limits(finemap_dat = finemap_dat,  
#                                 plot.zoom = "3x")
# finemap_dat <- subset(finemap_dat, Mb>xlims[1] & Mb<xlims[2] )

FCGR2A <- PLOT.locus(finemap_dat=fread(data_dict$finemap[[locus]]),   
                    LD_matrix=fread(data_dict$LD[[locus]]),
                    LD_reference="UKB",
                    locus_dir=data_dict$locus_dir[[locus]], 
                    
                    max_transcripts = 3,
                    Nott_epigenome=T,  
                    # Nott_binwidth = 500,
                    Nott_regulatory_rects = T, 
                    Nott_show_placseq = T,
                    
                    save_plot=F, 
                    return_list=T,
                    plot.zoom=c("25x")) 
``` 

# Modify plots

```{r, eval=F}
zoom <- "25x"
locus <- "FCGR2A"#"FCGR2A"
plot_zoom_list <- FCGR2A

mod_plot <- function(plot_zoom_list, locus, zoom){
  # deparse(substitute(FCGR2A))
  TRKS <- plot_zoom_list[[zoom]]
  finemap_dat <- fread(data_dict$finemap[[locus]])
  
  n_snps <- if(data_dict$dataset_type %in% names(TRKS)){
          paste0("n SNPs: ",
                 nrow(ggplot2::ggplot_build(TRKS[[data_dict$dataset_type]])$data[[2]]),", ")
        } else {NULL}
  title_text <- paste0(basename(data_dict$locus_dir[[locus]]),"   (",n_snps,"zoom: ",zoom,")")
     
  # Alter track
  TRKS[["Genes"]] <- TRKS[["Genes"]] +  
    scale_y_discrete(expand = expansion(add  = c(.0,-1)))
  
  
  # Fuse tracks
  LOCUS_zoom <- patchwork::wrap_plots(TRKS, 
                                       ncol = 1, 
                                       heights = heights) +
      patchwork::plot_annotation(title = title_text)
   
  # Save plot
  ggsave(filename = file.path(results_dir,"GWAS/Nalls23andMe_2019",locus,
                              paste0("multiview.",locus,
                                     ".UKB.",zoom,"_mod.jpg")), 
         plot = LOCUS_zoom, 
         dpi = 300, 
         height=12,
         width=10)
  return(LOCUS_zoom)
}


```

# Session info

<details> 

```{r Session Info, attr.output='style="max-height: 200px;"'}
utils::sessionInfo()
```

</details>

