---
title: "QTL pipeline vignette" 
author: "Brian M. Schilder"
date: "Most recent update:<br> `r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{QTL_pipeline_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = T, error = T, message=F} 
root.dir <- tempdir()
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  root.dir = root.dir,
  fig.height = 12,
  fig.width = 10
)  
knitr::opts_knit$set(root.dir = root.dir)

library(echolocatoR) 
library(dplyr)
results_dir <- file.path(root.dir,"results") 
dir.create(results_dir, showWarnings = F, recursive = F)
```

 
# QTL pipeline

- Here, we will use GWAS-eQTL colocalization results provided via the [*echolocatoR Fine-mapping Portal*](https://github.com/RajLabMSSM/Fine_Mapping_Shiny), from the preprint:
> K de Paiva Lopes, GJL Snijders, J Humphrey, A Allan, M Sneeboer, E Navarro, BM Schilderâ€¦T Raj (2020) Atlas of Genetic Effects in Human Microglia Transcriptome across Brain Regions, Aging and Disease Pathologies. bioRxiv; [https://doi.org/10.1101/2020.10.27.356113](https://doi.org/10.1101/2020.10.27.356113).


## Prepare `top_SNPs` data.frame   

- In this case, we don't have a top SNPs file ready.
So we're just going to make one directly from the full summary stats file itself (*NOTE*: You can only use this approach if you can fit the entire file in memory).  
- In this case, you'll want to make sure to set `grouping_vars=c("Locus","Gene")` so that you get top SNPs for each eGene-locus pair (not just one SNP per locus).  

```{r  Prepare `top_SNPs` data.frame}
fullSS_path <- file.path(root.dir,"Microglia_all_regions_Kunkle_2019_COLOC.snp-level_select.tsv.gz")
## Download example data 
if(!file.exists(fullSS_path)){
  download.file("https://github.com/RajLabMSSM/Fine_Mapping_Shiny/raw/master/www/Microglia_all_regions_Kunkle_2019_COLOC.snp-level_select.tsv.gz", fullSS_path)
}


top_SNPs <- import_topSNPs(  
  topSS = fullSS_path,
  chrom_col = "chr", 
  position_col = "pos",
  snp_col="snp",
  pval_col="gwas.pvalues", 
  effect_col="gwas.beta", 
  gene_col="gene",
  locus_col = "Locus",
  grouping_vars = c("Locus","Gene")) 
head(top_SNPs)
```

## eGene-locus list  

- In `finemap_loci()` we usually supply `loci=` with a list of locus names. 
However,when fine-mapping GWAS-QTL jointly, you'll need to specify which
QTL eGene-GWAS locus pairs you want to run.
- You can easy get this named list of all eGene-locus pairs with `gene_locus_list()`. The returned list's values are the GWAS loci, while the names are the QTL eGenes.  
- If you don't want to test all pairs, you can filter the `top_SNPs` object first. In this example, I'm going to remove any pairs that don't have matching eGene-locus names.  
We'll also limit the anaylsis to just 3 loci.  

```{r} 
loci <- gene_locus_list(top_SNPs)
# We'll just fine-map 2 loci to start
loci <- loci[1:3] 
loci
```


## Run fine-mapping pipeline  

**NOTE**: Currently this `finemap_loci()` will only fine-map the GWAS results. However you can still plot the fine-mapped GWAS results using the `QTL_prefixes` argument.  

```{r}
Kunkle_2019.microgliaQTL <- finemap_loci(# GENERAL ARGUMENTS 
                                        top_SNPs = top_SNPs,  
                                        results_dir = results_dir,
                                        
                                        loci = loci,
                                        dataset_name = "Kunkle_2019.microgliaQTL",
                                        dataset_type = "GWAS",  
                                        sample_size = 21982 + 41944,
                                        proportion_cases = 21982 / 41944,
                                        force_new_subset = F,
                                        force_new_LD = F,
                                        force_new_finemap = F,
                                        remove_tmps = T,
                                          
                 # SUMMARY STATS ARGUMENTS
                 fullSS_path = fullSS_path,
                 query_by ="fullSS",
                 chrom_col = "chr", 
                 position_col = "pos", 
                 snp_col = "snp",
                 pval_col = "gwas.pvalues", 
                 effect_col = "gwas.beta", 
                 stderr_col = "gwas.varbeta", 
                 MAF_col = "gwas.MAF", 
                 gene_col = "gene",
                 
                 # QTL prefixes
                 QTL_prefixes = c("qtl."),
                 
                 # FILTERING ARGUMENTS
                 bp_distance = 500000, #100000,
                 min_MAF = 0.001, 
                 
                 # FINE-MAPPING ARGUMENTS
                 finemap_methods = c("ABF","FINEMAP","SUSIE","POLYFUN_SUSIE"),
                 n_causal = 5,
                 credset_thresh = .95,
                 
                 # LD ARGUMENTS 
                 LD_reference = "1KGphase1",
                 superpopulation = "EUR",
                 download_method = "axel",
                 
                 # PLOT ARGUMENTS 
                 plot.types=c("simple"),
                 plot.zoom = "1x" 
                 )
```


# Session info

<details> 

```{r Session Info, attr.output='style="max-height: 200px;"'}
utils::sessionInfo()
```

</details>
