---
title: "Summarise"
author: "<h4>Author: <i>Brian M. Schilder</i></h4>" 
date: "<h4>Updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output:
  BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{BD_finemapping} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = T, message = F, error = T}
root.dir <- tempdir()
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  root.dir = root.dir, 
  fig.height = 12,
  fig.width = 10
)  
knitr::opts_knit$set(
  root.dir = root.dir
  )
library(echolocatoR) 
results_dir <- file.path(root.dir,"results")
```


# Download data  

* Many fine-mapping results can be found on the [echolocatoR Fine-mapping Portal](https://rajlab.shinyapps.io/Fine_Mapping_Shiny).  
* We will download some of those files here using some built-in APIs.  

```{r Download data}  
local_files <- echodata::portal_query(results_dir = results_dir, 
                                      dataset_types="GWAS",
                                      phenotypes = c("parkinson"),
                                      file_types = "multi_finemap", 
                                      LD_panels=c("UKB")) 
```

# Merge `echolocatoR` results  

Gather all of the fine-mapping results generated by `finemap_loci()` previously.  

```{r merge_finemapping_results()} 
merged_DT <- echoannot::merge_finemapping_results(minimum_support=0,
                                       include_leadSNPs=TRUE,
                                       dataset = file.path(results_dir,
                                                           "GWAS/Nalls23andMe_2019"),   
                                       consensus_thresh = 2)    

echoannot:::results_report(merged_DT)
```

# Summarise

## `get_SNPgroup_counts()`

Get the number of SNPs for each SNP group per locus.  
It also prints the mean number of SNPs for each SNP group across all loci.  
**NOTE**: You will need to make sure to set `merge_finemapping_results(minimum_support=1)`
in the above step to get accurate counts for all SNP groups.  

```{r get_SNPgroup_counts()}
snp_groups <- echoannot::get_SNPgroup_counts(merged_DT=merged_DT)   
createDT(snp_groups)
```

## `get_CS_counts()` 

County the number of tool-specific and UCS Credible Set SNPs per locus.  

```{r get_CS_counts()}
UCS_counts <- echodata::get_CS_counts(merged_DT = merged_DT)
print(UCS_counts)
```


# Plot  

- The following functions each return a list containing both the `...$plot` and the `...$data`
used to make the plot.  
- Where available, `snp_filter` allows user to use any filtering argument (supplied as a string)
to subset the data they want to use in the plot/data.  


## `SUMMARISE.coloc_nominated_eGenes()`  

If you ran colocalization tests with `echolocatoR` (via `catalogueR`) you can use those results to 
come up with a top QTL nominated gene for each locus (potentially implicating that gene in your phenotype).  

### Download COLOC results from [echolocatoR Fine-mapping Portal](https://rajlab.shinyapps.io/Fine_Mapping_Shiny).    

```{r}
coloc_res <- echodata::get_Nalls2019_coloc() 
```


## Multi-plot  
 

```{r super_summary_plot()}
super_plot <- echoannot::super_summary_plot(
  merged_DT = merged_DT,
  snp_filter="Consensus_SNP==T",
  coloc_results=coloc_res,
  plot_missense=FALSE)
```


# Session info 

<details>

```{r Session Info, attr.output='style="max-height: 200px;"'}
utils::sessionInfo()
```

</details>

