<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Check genome build — check_genome • echolocatoR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Check genome build — check_genome"><meta property="og:description" content="If fullSS_genome_build==NULL and munged=TRUE,
infers genome build (hg19 vs. hg38)
from summary statistics using get_genome_builds.
This can only be done with summary statistics that have already been
munged by format_sumstats.
When fullSS_genome_build is a synonym of hg19 or hg38, this function
simply returns a standardized version of the user-provided
genome build."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">echolocatoR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/echolocatoR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/QTLs.html">QTLs</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="../articles/summarise.html">Summarise</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/RajLabMSSM/echolocatoR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Check genome build</h1>
    <small class="dont-index">Source: <a href="https://github.com/RajLabMSSM/echolocatoR/blob/HEAD/R/check_genome.R" class="external-link"><code>R/check_genome.R</code></a></small>
    <div class="hidden name"><code>check_genome.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>If <code>fullSS_genome_build==NULL</code> and <code>munged=TRUE</code>,
infers genome build (hg19 vs. hg38)
from summary statistics using <a href="https://rdrr.io/pkg/MungeSumstats/man/get_genome_builds.html" class="external-link">get_genome_builds</a>.
This can only be done with summary statistics that have already been
munged by <a href="https://rdrr.io/pkg/MungeSumstats/man/format_sumstats.html" class="external-link">format_sumstats</a>.
When <code>fullSS_genome_build</code> is a synonym of hg19 or hg38, this function
simply returns a standardized version of the user-provided
genome build.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">check_genome</span><span class="op">(</span></span>
<span>  fullSS_genome_build <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  munged <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fullSS_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampled_snps <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  names_from_paths <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dbSNP <span class="op">=</span> <span class="fl">155</span>,</span>
<span>  nThread <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>fullSS_genome_build</dt>
<dd><p>Genome build of the full summary statistics
 (<code>fullSS_path</code>). Can be "GRCH37" or "GRCH38" or one of their synonyms..
If <code>fullSS_genome_build==NULL</code> and <code>munged=TRUE</code>,
infers genome build (hg19 vs. hg38)
from summary statistics using <a href="https://rdrr.io/pkg/MungeSumstats/man/get_genome_builds.html" class="external-link">get_genome_builds</a>.</p></dd>


<dt>munged</dt>
<dd><p>Whether <code>fullSS_path</code> have already been
standardised/filtered full summary stats
with <a href="https://rdrr.io/pkg/MungeSumstats/man/format_sumstats.html" class="external-link">format_sumstats</a>.
If <code>munged=FALSE</code> you'll need to provide the necessary
 column names to the <code>colmap</code> argument.</p></dd>


<dt>fullSS_path</dt>
<dd><p>Path to the full summary statistics file (GWAS or QTL)
that you want to fine-map.
It is usually best to provide the absolute path rather
than the relative path.</p></dd>


<dt>sampled_snps</dt>
<dd><p>Downsample the number of SNPs used when inferring genome
build to save time.</p></dd>


<dt>names_from_paths</dt>
<dd><p>Infer the name of each item in <code>sumstats_list</code>
from its respective file path.
Only works if <code>sumstats_list</code> is a list of paths.</p></dd>


<dt>dbSNP</dt>
<dd><p>version of dbSNP to be used (144 or 155). Default is 155.</p></dd>


<dt>nThread</dt>
<dd><p>Number of threads to parallelise saving across.</p></dd>


<dt>verbose</dt>
<dd><p>Print messages.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Character string indicating genome build.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">fullSS_path</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/example_fullSS.html" class="external-link">example_fullSS</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Writing file to ==&gt; /tmp/Rtmp7rpVnL/nalls2019.fullSS_subset.tsv</span>
<span class="r-in"><span><span class="va">build</span> <span class="op">&lt;-</span> <span class="fu">check_genome</span><span class="op">(</span>fullSS_genome_build<span class="op">=</span><span class="st">"hg19"</span>,</span></span>
<span class="r-in"><span>                      fullSS_path<span class="op">=</span><span class="va">fullSS_path</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

