<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Extract a subset of the summary stats — extract_snp_subset • echolocatoR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Extract a subset of the summary stats — extract_snp_subset"><meta property="og:description" content="Use tabix to extract a locus subset
 from the full summary statistics file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">echolocatoR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/echolocatoR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/QTLs.html">QTLs</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="../articles/summarise.html">Summarise</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/RajLabMSSM/echolocatoR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract a subset of the summary stats</h1>
    <small class="dont-index">Source: <a href="https://github.com/RajLabMSSM/echolocatoR/blob/HEAD/R/extract_snp_subset.R" class="external-link"><code>R/extract_snp_subset.R</code></a></small>
    <div class="hidden name"><code>extract_SNP_subset.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Use <em>tabix</em> to extract a locus subset
 from the full summary statistics file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">extract_snp_subset</span><span class="op">(</span></span>
<span>  <span class="va">subset_path</span>,</span>
<span>  locus <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colmap <span class="op">=</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/construct_colmap.html" class="external-link">construct_colmap</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">fullSS_path</span>,</span>
<span>  <span class="va">topSNPs</span>,</span>
<span>  <span class="va">LD_reference</span>,</span>
<span>  force_new_subset <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  force_new_maf <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bp_distance <span class="op">=</span> <span class="fl">5e+05</span>,</span>
<span>  superpopulation <span class="op">=</span> <span class="st">"EUR"</span>,</span>
<span>  compute_n <span class="op">=</span> <span class="st">"ldsc"</span>,</span>
<span>  query_by <span class="op">=</span> <span class="st">"tabix"</span>,</span>
<span>  download_method <span class="op">=</span> <span class="st">"axel"</span>,</span>
<span>  nThread <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  conda_env <span class="op">=</span> <span class="st">"echoR_mini"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>subset_path</dt>
<dd><p>Path where the <code>query</code> should be 
saved after standardization.</p></dd>


<dt>locus</dt>
<dd><p>Locus name to fine-map (e.g. <code>"BIN1"</code>).
Can be named to indicate a specific gene within a QTL locus
(e.g. <code>c(ENSG00000136731="BIN1")</code>).</p></dd>


<dt>colmap</dt>
<dd><p>Column name mappings in in <code>fullSS_path</code>. Must be a named
list. Can use <a href="https://rdrr.io/pkg/echodata/man/construct_colmap.html" class="external-link">construct_colmap</a> to assist with this. This
function can be used in two different ways:</p><ul><li><p><code>munged=FALSE</code> : When <code>munged=FALSE</code>,
 you will need to provide the necessary column names to the
 <code>colmap</code> argument (<em>default</em>).</p></li>
<li><p><code>munged=TRUE</code> :  Alternatively, instead of filling out
 each argument in
<a href="https://rdrr.io/pkg/echodata/man/construct_colmap.html" class="external-link">construct_colmap</a>, you can simply set <code>munged=TRUE</code>
 if  <code>fullSS_path</code> has already been munged with
 <a href="https://rdrr.io/pkg/MungeSumstats/man/format_sumstats.html" class="external-link">format_sumstats</a>.</p></li>
</ul></dd>


<dt>fullSS_path</dt>
<dd><p>Path to the full summary statistics file (GWAS or QTL)
that you want to fine-map.
It is usually best to provide the absolute path rather
than the relative path.</p></dd>


<dt>topSNPs</dt>
<dd><p>A data.frame with the genomic coordinates of the lead SNP
for each locus.
The lead SNP will be used as the center of the window when extracting
subset from the full GWAS/QTL summary statistics file.
Only one SNP per <strong>Locus</strong> should be included.
At minimum, <code>topSNPs</code> should include the following columns:</p><dl><dt><em>Locus</em></dt>
<dd><p>A unique name for each locus. Often,
 loci are named after a relevant gene (e.g. LRRK2) or based on
  the name/coordinates of the lead SNP (e.g. locus_chr12_40734202)</p></dd>

<dt><em>CHR</em></dt>
<dd><p>The chromosome that the SNP is on.
 Can be "chr12" or "12" format.</p></dd>

<dt><em>POS</em></dt>
<dd><p>The genomic position of the SNP (in basepairs)</p></dd>


</dl></dd>


<dt>LD_reference</dt>
<dd><p>LD reference to use:</p><ul><li><p>"1KGphase1" : 1000 Genomes Project Phase 1 (genome build: hg19).</p></li>
<li><p>"1KGphase3" : 1000 Genomes Project Phase 3 (genome build: hg19).</p></li>
<li><p>"UKB" : Pre-computed LD from a British
European-decent subset of UK Biobank.
<em>Genome build</em> : hg19</p></li>
<li><p>"&lt;vcf_path&gt;" : User-supplied path to a custom VCF file 
to compute LD matrix from.<br><em>Accepted formats</em>: <em>.vcf</em> / <em>.vcf.gz</em> / <em>.vcf.bgz</em><br><em>Genome build</em> : defined by user with <code>target_genome</code>.</p></li>
<li><p>"&lt;matrix_path&gt;" : User-supplied path to a pre-computed LD matrix   
<em>Accepted formats</em>:  <em>.rds</em> / <em>.rda</em> / <em>.csv</em> / 
<em>.tsv</em> / <em>.txt</em><br><em>Genome build</em> : defined by user with <code>target_genome</code>.</p></li>
</ul></dd>


<dt>force_new_subset</dt>
<dd><p>By default, if a subset of the full
 summary stats file for a given locus is already present,
then <span class="pkg">echolocatoR</span> will just use the pre-existing file.
Set <code>force_new_subset=T</code> to override this and extract a new subset.
Subsets are saved in the following path structure:
<em>Data/\&lt;dataset_type\&gt;/\&lt;dataset_name\&gt;/\&lt;locus\&gt;/Multi-finemap/
\&lt;locus\&gt;_\&lt;dataset_name\&gt;_Multi-finemap.tsv.gz</em></p></dd>


<dt>force_new_maf</dt>
<dd><p>Download UKB_MAF file again.</p></dd>


<dt>bp_distance</dt>
<dd><p>Distance around the lead SNP to include.</p></dd>


<dt>superpopulation</dt>
<dd><p>Superpopulation to subset LD panel by
(used only if <code>LD_reference</code> is "1KGphase1" or "1KGphase3").
See <a href="https://rdrr.io/pkg/echoLD/man/popDat_1KGphase1.html" class="external-link">popDat_1KGphase1</a> and <a href="https://rdrr.io/pkg/echoLD/man/popDat_1KGphase3.html" class="external-link">popDat_1KGphase3</a>
for full tables of their respective samples.</p></dd>


<dt>compute_n</dt>
<dd><p>How to compute per-SNP sample size (new column "N").<br>
If the column "N" is already present in <code>dat</code>, this column 
will be used to extract per-SNP sample sizes
and the argument <code>compute_n</code> will be ignored.<br>
If the column "N" is <em>not</em> present in <code>dat</code>, one of the following 
options can be supplied to <code>compute_n</code>:</p><ul><li><p><code>0</code>: N will not be computed.</p></li>
<li><p><code>&gt;0</code>: If any number &gt;0 is provided,
that value will be set as N for every row.
**Note**: Computing N this way is incorrect and should be avoided
if at all possible.</p></li>
<li><p><code>"sum"</code>: N will be computed as:
cases (N_CAS) + controls (N_CON), so long as both columns are present.</p></li>
<li><p><code>"ldsc"</code>: N will be computed as effective sample size:
Neff =(N_CAS+N_CON)*(N_CAS/(N_CAS+N_CON)) / mean((N_CAS/(N_CAS+N_CON))(N_CAS+N_CON)==max(N_CAS+N_CON)).</p></li>
<li><p><code>"giant"</code>: N will be computed as effective sample size:
Neff = 2 / (1/N_CAS + 1/N_CON).</p></li>
<li><p><code>"metal"</code>: N will be computed as effective sample size:
Neff = 4 / (1/N_CAS + 1/N_CON).</p></li>
</ul></dd>


<dt>query_by</dt>
<dd><p>Choose which method you want to use to extract
 locus subsets from the full summary stats file.
Methods include:</p><dl><dt>"tabix"</dt>
<dd><p>Convert the full summary stats file in an indexed tabix file.
 Makes querying lightning fast after the initial conversion is done.
  (<em>default</em>)</p></dd>

<dt>"coordinates"</dt>
<dd><p>Extract locus subsets using min/max genomic
coordinates with <em>awk</em>.</p></dd>


</dl></dd>


<dt>download_method</dt>
<dd><ul><li><p><code>"axel"</code> : Multi-threaded</p></li>
<li><p><code>"wget"</code> : Single-threaded</p></li>
<li><p><code>"download.file"</code> : Single-threaded</p></li>
<li><p><code>"internal"</code> : Single-threaded 
(passed to <a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a>)</p></li>
<li><p><code>"wininet"</code> : Single-threaded 
(passed to <a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a>)</p></li>
<li><p><code>"libcurl"</code> : Single-threaded 
(passed to <a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a>)</p></li>
<li><p><code>"curl"</code> : Single-threaded 
(passed to <a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a>)</p></li>
</ul></dd>


<dt>nThread</dt>
<dd><p>Number of threads to parallelise saving across.</p></dd>


<dt>conda_env</dt>
<dd><p>Conda environment to use.</p></dd>


<dt>verbose</dt>
<dd><p>Print messages.</p></dd>

</dl></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Other query functions: 
<code><a href="query_handler.html">query_handler</a>()</code></p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

