<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Fine-map multiple loci — finemap_loci • echolocatoR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Fine-map multiple loci — finemap_loci"><meta property="og:description" content="echolocatoR will automatically fine-map each locus.
Uses the topSNPs data.frame to define locus coordinates."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">echolocatoR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/echolocatoR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/QTLs.html">QTLs</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="../articles/finemapping_portal.html">echolocatoR Fine-mapping Portal</a>
    </li>
    <li>
      <a href="../articles/plot_locus.html">Plot locus</a>
    </li>
    <li>
      <a href="../articles/summarise.html">Summarise</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/RajLabMSSM/echolocatoR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fine-map multiple loci</h1>
    <small class="dont-index">Source: <a href="https://github.com/RajLabMSSM/echolocatoR/blob/HEAD/R/finemap_loci.R" class="external-link"><code>R/finemap_loci.R</code></a></small>
    <div class="hidden name"><code>finemap_loci.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><span class="pkg">echolocatoR</span> will automatically fine-map each locus.
Uses the <code>topSNPs</code> data.frame to define locus coordinates.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">finemap_loci</span><span class="op">(</span></span>
<span>  loci <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">fullSS_path</span>,</span>
<span>  fullSS_genome_build <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  results_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"results"</span><span class="op">)</span>,</span>
<span>  dataset_name <span class="op">=</span> <span class="st">"dataset_name"</span>,</span>
<span>  dataset_type <span class="op">=</span> <span class="st">"GWAS"</span>,</span>
<span>  topSNPs <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  force_new_subset <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  force_new_LD <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  force_new_finemap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  finemap_methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ABF"</span>, <span class="st">"FINEMAP"</span>, <span class="st">"SUSIE"</span><span class="op">)</span>,</span>
<span>  finemap_args <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_causal <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  credset_thresh <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  consensus_thresh <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  fillNA <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  conditioned_snps <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  priors_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  munged <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  colmap <span class="op">=</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/construct_colmap.html" class="external-link">construct_colmap</a></span><span class="op">(</span>munged <span class="op">=</span> <span class="va">munged</span><span class="op">)</span>,</span>
<span>  compute_n <span class="op">=</span> <span class="st">"ldsc"</span>,</span>
<span>  LD_reference <span class="op">=</span> <span class="st">"1KGphase3"</span>,</span>
<span>  LD_genome_build <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  leadSNP_LD_block <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  superpopulation <span class="op">=</span> <span class="st">"EUR"</span>,</span>
<span>  download_method <span class="op">=</span> <span class="st">"axel"</span>,</span>
<span>  bp_distance <span class="op">=</span> <span class="fl">5e+05</span>,</span>
<span>  min_POS <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  max_POS <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  min_MAF <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  trim_gene_limits <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  max_snps <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_r2 <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  remove_variants <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  remove_correlates <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  query_by <span class="op">=</span> <span class="st">"tabix"</span>,</span>
<span>  case_control <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  qtl_suffixes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"simple"</span><span class="op">)</span>,</span>
<span>  show_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  zoom <span class="op">=</span> <span class="st">"1x"</span>,</span>
<span>  tx_biotypes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nott_epigenome <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nott_show_placseq <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nott_binwidth <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  nott_bigwig_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  xgr_libnames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  roadmap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  roadmap_query <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  remove_tmps <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  conda_env <span class="op">=</span> <span class="st">"echoR_mini"</span>,</span>
<span>  return_all <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use_tryCatch <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">2022</span>,</span>
<span>  nThread <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  top_SNPs <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  PP_threshold <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  consensus_threshold <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot.Nott_epigenome <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot.Nott_show_placseq <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot.Nott_binwidth <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot.Nott_bigwig_dir <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot.Roadmap <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot.Roadmap_query <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot.XGR_libnames <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  server <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot.types <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot.zoom <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  QTL_prefixes <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vcf_folder <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  probe_path <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  file_sep <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  chrom_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  chrom_type <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  position_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  snp_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  pval_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  effect_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  stderr_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  tstat_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  locus_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  freq_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  MAF_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  A1_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  A2_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  N_cases_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  N_controls_col <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  N_cases <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  N_controls <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  proportion_cases <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  sample_size <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  PAINTOR_QTL_datasets <span class="op">=</span> <span class="fu"><a href="deprecated.html">deprecated</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>loci</dt>
<dd><p>Character list of loci in <strong>Locus</strong> col of <code>topSNPs</code>.</p></dd>


<dt>fullSS_path</dt>
<dd><p>Path to the full summary statistics file (GWAS or QTL)
that you want to fine-map.
It is usually best to provide the absolute path rather
than the relative path.</p></dd>


<dt>fullSS_genome_build</dt>
<dd><p>Genome build of the full summary statistics
 (<code>fullSS_path</code>). Can be "GRCH37" or "GRCH38" or one of their synonyms..
If <code>fullSS_genome_build==NULL</code> and <code>munged=TRUE</code>,
infers genome build (hg19 vs. hg38)
from summary statistics using <a href="https://rdrr.io/pkg/MungeSumstats/man/get_genome_builds.html" class="external-link">get_genome_builds</a>.</p></dd>


<dt>results_dir</dt>
<dd><p>Where to store all results.
<strong>IMPORTANT!:</strong> It is usually best to provide the absolute path
rather than the relative path.
This is especially important for <em>FINEMAP</em>.</p></dd>


<dt>dataset_name</dt>
<dd><p>The name you want to assign to the dataset
being fine-mapped,
This will be used to name the subdirectory where your
results will be stored
(e.g. <em>Data/GWAS/&lt;dataset_name&gt;</em>).
Don't use special characters (e.g.".", "/").</p></dd>


<dt>dataset_type</dt>
<dd><p>The kind dataset you're fine-mapping
(e.g. GWAS, eQTL, tQTL).
This will also be used when creating the subdirectory where your results
will be stored
(e.g. <em>Data/&lt;dataset_type&gt;/Kunkle_2019</em>).</p></dd>


<dt>topSNPs</dt>
<dd><p>A data.frame with the genomic coordinates of the lead SNP
for each locus.
The lead SNP will be used as the center of the window when extracting
subset from the full GWAS/QTL summary statistics file.
Only one SNP per <strong>Locus</strong> should be included.
At minimum, <code>topSNPs</code> should include the following columns:</p><dl><dt><em>Locus</em></dt>
<dd><p>A unique name for each locus. Often,
 loci are named after a relevant gene (e.g. LRRK2) or based on
  the name/coordinates of the lead SNP (e.g. locus_chr12_40734202)</p></dd>

<dt><em>CHR</em></dt>
<dd><p>The chromosome that the SNP is on.
 Can be "chr12" or "12" format.</p></dd>

<dt><em>POS</em></dt>
<dd><p>The genomic position of the SNP (in basepairs)</p></dd>


</dl></dd>


<dt>force_new_subset</dt>
<dd><p>By default, if a subset of the full
 summary stats file for a given locus is already present,
then <span class="pkg">echolocatoR</span> will just use the pre-existing file.
Set <code>force_new_subset=T</code> to override this and extract a new subset.
Subsets are saved in the following path structure:
<em>Data/\&lt;dataset_type\&gt;/\&lt;dataset_name\&gt;/\&lt;locus\&gt;/Multi-finemap/
\&lt;locus\&gt;_\&lt;dataset_name\&gt;_Multi-finemap.tsv.gz</em></p></dd>


<dt>force_new_LD</dt>
<dd><p>Force new LD subset.</p></dd>


<dt>force_new_finemap</dt>
<dd><p>By default, if an fine-mapping results file for
 a given locus is already present,
then <span class="pkg">echolocatoR</span> will just use the preexisting file.
Set <code>force_new_finemap=T</code> to override this and re-run fine-mapping.</p></dd>


<dt>finemap_methods</dt>
<dd><p>Which fine-mapping methods you want to use.</p></dd>


<dt>finemap_args</dt>
<dd><p>A named nested list containing additional arguments 
for each fine-mapping method. e.g.
<code>finemap_args = list(FINEMAP=list(), PAINTOR=list(method=""))</code></p></dd>


<dt>n_causal</dt>
<dd><p>The maximum number of potential causal SNPs per locus.
This parameter is used somewhat differently by different fine-mapping tools.
See tool-specific functions for details.</p></dd>


<dt>credset_thresh</dt>
<dd><p>The minimum fine-mapped posterior probability
 for a SNP to be considered part of a Credible Set.
For example, <code>credset_thresh=.95</code> means that all Credible Set SNPs
will be 95% Credible Set SNPs.</p></dd>


<dt>consensus_thresh</dt>
<dd><p>The minimum number of fine-mapping tools 
in which a SNP is in the Credible Set in order to be included in the 
"Consensus_SNP" column.</p></dd>


<dt>fillNA</dt>
<dd><p>Value to fill LD matrix NAs with.</p></dd>


<dt>conditioned_snps</dt>
<dd><p>Which SNPs to conditions on when fine-mapping
with (e.g. <em>COJO</em>).</p></dd>


<dt>priors_col</dt>
<dd><p>[Optional] Name of the a column in 
<code>dat</code> to extract SNP-wise prior probabilities from.</p></dd>


<dt>munged</dt>
<dd><p>Whether <code>fullSS_path</code> have already been
standardised/filtered full summary stats
with <a href="https://rdrr.io/pkg/MungeSumstats/man/format_sumstats.html" class="external-link">format_sumstats</a>.
If <code>munged=FALSE</code> you'll need to provide the necessary
 column names to the <code>colmap</code> argument.</p></dd>


<dt>colmap</dt>
<dd><p>Column name mappings in in <code>fullSS_path</code>. Must be a named
list. Can use <a href="https://rdrr.io/pkg/echodata/man/construct_colmap.html" class="external-link">construct_colmap</a> to assist with this. This
function can be used in two different ways:</p><ul><li><p><code>munged=FALSE</code> : When <code>munged=FALSE</code>,
 you will need to provide the necessary column names to the
 <code>colmap</code> argument (<em>default</em>).</p></li>
<li><p><code>munged=TRUE</code> :  Alternatively, instead of filling out
 each argument in
<a href="https://rdrr.io/pkg/echodata/man/construct_colmap.html" class="external-link">construct_colmap</a>, you can simply set <code>munged=TRUE</code>
 if  <code>fullSS_path</code> has already been munged with
 <a href="https://rdrr.io/pkg/MungeSumstats/man/format_sumstats.html" class="external-link">format_sumstats</a>.</p></li>
</ul></dd>


<dt>compute_n</dt>
<dd><p>How to compute per-SNP sample size (new column "N").<br>
If the column "N" is already present in <code>dat</code>, this column 
will be used to extract per-SNP sample sizes
and the argument <code>compute_n</code> will be ignored.<br>
If the column "N" is <em>not</em> present in <code>dat</code>, one of the following 
options can be supplied to <code>compute_n</code>:</p><ul><li><p><code>0</code>: N will not be computed.</p></li>
<li><p><code>&gt;0</code>: If any number &gt;0 is provided,
that value will be set as N for every row.
**Note**: Computing N this way is incorrect and should be avoided
if at all possible.</p></li>
<li><p><code>"sum"</code>: N will be computed as:
cases (N_CAS) + controls (N_CON), so long as both columns are present.</p></li>
<li><p><code>"ldsc"</code>: N will be computed as effective sample size:
Neff =(N_CAS+N_CON)*(N_CAS/(N_CAS+N_CON)) / mean((N_CAS/(N_CAS+N_CON))(N_CAS+N_CON)==max(N_CAS+N_CON)).</p></li>
<li><p><code>"giant"</code>: N will be computed as effective sample size:
Neff = 2 / (1/N_CAS + 1/N_CON).</p></li>
<li><p><code>"metal"</code>: N will be computed as effective sample size:
Neff = 4 / (1/N_CAS + 1/N_CON).</p></li>
</ul></dd>


<dt>LD_reference</dt>
<dd><p>LD reference to use:</p><ul><li><p>"1KGphase1" : 1000 Genomes Project Phase 1 (genome build: hg19).</p></li>
<li><p>"1KGphase3" : 1000 Genomes Project Phase 3 (genome build: hg19).</p></li>
<li><p>"UKB" : Pre-computed LD from a British
European-decent subset of UK Biobank.
<em>Genome build</em> : hg19</p></li>
<li><p>"&lt;vcf_path&gt;" : User-supplied path to a custom VCF file 
to compute LD matrix from.<br><em>Accepted formats</em>: <em>.vcf</em> / <em>.vcf.gz</em> / <em>.vcf.bgz</em><br><em>Genome build</em> : defined by user with <code>target_genome</code>.</p></li>
<li><p>"&lt;matrix_path&gt;" : User-supplied path to a pre-computed LD matrix   
<em>Accepted formats</em>:  <em>.rds</em> / <em>.rda</em> / <em>.csv</em> / 
<em>.tsv</em> / <em>.txt</em><br><em>Genome build</em> : defined by user with <code>target_genome</code>.</p></li>
</ul></dd>


<dt>LD_genome_build</dt>
<dd><p>Genome build of the LD panel.
This is automatically assigned to the correct genome build for each
LD panel except when the user supplies custom vcf/LD files.</p></dd>


<dt>leadSNP_LD_block</dt>
<dd><p>Only return SNPs within the same LD block
as the lead SNP (the SNP with the smallest p-value).</p></dd>


<dt>superpopulation</dt>
<dd><p>Superpopulation to subset LD panel by
(used only if <code>LD_reference</code> is "1KGphase1" or "1KGphase3").
See <a href="https://rdrr.io/pkg/echoLD/man/popDat_1KGphase1.html" class="external-link">popDat_1KGphase1</a> and <a href="https://rdrr.io/pkg/echoLD/man/popDat_1KGphase3.html" class="external-link">popDat_1KGphase3</a>
for full tables of their respective samples.</p></dd>


<dt>download_method</dt>
<dd><ul><li><p><code>"axel"</code> : Multi-threaded</p></li>
<li><p><code>"wget"</code> : Single-threaded</p></li>
<li><p><code>"download.file"</code> : Single-threaded</p></li>
<li><p><code>"internal"</code> : Single-threaded 
(passed to <a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a>)</p></li>
<li><p><code>"wininet"</code> : Single-threaded 
(passed to <a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a>)</p></li>
<li><p><code>"libcurl"</code> : Single-threaded 
(passed to <a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a>)</p></li>
<li><p><code>"curl"</code> : Single-threaded 
(passed to <a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a>)</p></li>
</ul></dd>


<dt>bp_distance</dt>
<dd><p>Distance around the lead SNP to include.</p></dd>


<dt>min_POS</dt>
<dd><p>Minimum genomic position to include.</p></dd>


<dt>max_POS</dt>
<dd><p>Maximum genomic position to include.</p></dd>


<dt>min_MAF</dt>
<dd><p>Minimum Minor Allele Frequency (MAF) of SNPs to include.</p></dd>


<dt>trim_gene_limits</dt>
<dd><p>If a gene name is supplied to this argument 
(e.g. <code>trim_gene_limits="BST"</code>), only SNPs within the gene body
 will be included.</p></dd>


<dt>max_snps</dt>
<dd><p>Maximum number of SNPs to include.</p></dd>


<dt>min_r2</dt>
<dd><p>Correlation threshold for <code>remove_correlates</code>.</p></dd>


<dt>remove_variants</dt>
<dd><p>A list of SNP RSIDs to remove.</p></dd>


<dt>remove_correlates</dt>
<dd><p>A list of SNPs.
If provided, all SNPs that correlates with these SNPs (at r2&gt;=<code>min_r2</code>)
will be removed from both <code>dat</code> and <code>LD</code> list items..</p></dd>


<dt>query_by</dt>
<dd><p>Choose which method you want to use to extract
 locus subsets from the full summary stats file.
Methods include:</p><dl><dt>"tabix"</dt>
<dd><p>Convert the full summary stats file in an indexed tabix file.
 Makes querying lightning fast after the initial conversion is done.
  (<em>default</em>)</p></dd>

<dt>"coordinates"</dt>
<dd><p>Extract locus subsets using min/max genomic
coordinates with <em>awk</em>.</p></dd>


</dl></dd>


<dt>case_control</dt>
<dd><p>Whether the summary statistics come from a case-control
study (e.g. a GWAS of having Alzheimer's Disease or not) (<code>TRUE</code>)
or a quantitative study (e.g. a GWAS of height, or an eQTL) (<code>FALSE</code>).</p></dd>


<dt>qtl_suffixes</dt>
<dd><p>If columns with QTL data is included in <code>dat</code>,
you can indicate which columns those are with one or more string suffixes
(e.g. <code>qtl_suffixes=c(".eQTL1",".eQTL2")</code> to use the columns 
"P.QTL1", "Effect.QTL1", "P.QTL2", "Effect.QTL2").</p></dd>


<dt>plot_types</dt>
<dd><p>Which kinds of plots to include.
Options:</p><ul><li><p>"simple"Just plot the following tracks: GWAS,
fine-mapping, gene models</p></li>
<li><p>"fancy"Additionally plot XGR annotation tracks
(XGR, Roadmap, Nott2019).
'</p></li>
<li><p>"LD"LD heatmap showing the 10 SNPs surrounding the lead SNP.</p></li>
</ul></dd>


<dt>show_plot</dt>
<dd><p>Print plot to screen.</p></dd>


<dt>zoom</dt>
<dd><p>Zoom into the center of the locus when plotting
(without editing the fine-mapping results file).
You can provide either:</p><ul><li><p>The size of your plot window in terms of basepairs
(e.g. <code>zoom=50000</code> for a 50kb window).</p></li>
<li><p>How much you want to zoom in (e.g. <code>zoom="1x"</code>
for the full locus, <code>zoom="2x"</code>
for 2x zoom into the center of the locus, etc.).</p></li>
</ul><p>You can pass a list of window sizes (e.g. <code>c(50000,100000,500000)</code>)
to automatically generate
multiple views of each locus.
This can even be a mix of different style inputs: e.g.
 <code>c("1x","4.5x",25000)</code>.</p></dd>


<dt>tx_biotypes</dt>
<dd><p>Transcript biotypes to include in the gene model track. 
By default (<code>NULL</code>), all transcript biotypes will be included.
See <a href="https://rdrr.io/pkg/echoplot/man/get_tx_biotypes.html" class="external-link">get_tx_biotypes</a> for a full list of 
all available biotypes</p></dd>


<dt>nott_epigenome</dt>
<dd><p>Include tracks showing brain cell-type-specific 
epigenomic data from 
<a href="https://doi.org/10.1126/science.aay0793" class="external-link">Nott et al. (2019)</a>.</p></dd>


<dt>nott_show_placseq</dt>
<dd><p>Include track generated by
<a href="https://rdrr.io/pkg/echoannot/man/NOTT2019_plac_seq_plot.html" class="external-link">NOTT2019_plac_seq_plot</a>.</p></dd>


<dt>nott_binwidth</dt>
<dd><p>When including Nott et al. (2019)
epigenomic data in the track plots,
adjust the bin width of the histograms.</p></dd>


<dt>nott_bigwig_dir</dt>
<dd><p>Instead of pulling Nott et al. (2019)
 epigenomic data
from the <em>UCSC Genome Browser</em>, use a set of local bigwig files.</p></dd>


<dt>xgr_libnames</dt>
<dd><p>Passed to <a href="https://rdrr.io/pkg/echoplot/man/XGR_plot.html" class="external-link">XGR_plot</a>.
Which XGR annotations to check overlap with.
For full list of libraries see
<a href="http://XGR_r-forge.r-project.org/#annotations-at-the-genomic-region-level" class="external-link">
 here</a>.
 Passed to the <code>RData.customised</code> argument in <a href="https://rdrr.io/pkg/XGR/man/xRDataLoader.html" class="external-link">xRDataLoader</a>.</p></dd>


<dt>roadmap</dt>
<dd><p>Find and plot annotations from Roadmap.</p></dd>


<dt>roadmap_query</dt>
<dd><p>Only plot annotations from Roadmap whose
metadata contains a string or any items from  a list of strings
(e.g. <code>"brain"</code> or <code>c("brain","liver","monocytes")</code>).</p></dd>


<dt>remove_tmps</dt>
<dd><p>Whether to remove any temporary files
(e.g. FINEMAP output files) after the pipeline is done running.</p></dd>


<dt>conda_env</dt>
<dd><p>Conda environment to use.</p></dd>


<dt>return_all</dt>
<dd><p>Return a nested list of various the pipeline's outputs
including plots, tables, and file paths (default: <code>TRUE</code>).
If <code>FALSE</code>, instead only returns a single merged
 <a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a> containing the results from all loci.</p></dd>


<dt>use_tryCatch</dt>
<dd><p>If an error is encountered in one locus,
the pipeline will continue to try running the rest of the loci
 (default: <code>use_tryCatch=TRUE</code>). This avoid stopping all analyses due
 to errors that only affect some loci,
 but currently prevents debugging via traceback.</p></dd>


<dt>seed</dt>
<dd><p>Set the seed for all functions where this is possible.</p></dd>


<dt>nThread</dt>
<dd><p>Number of threads to parallelise saving across.</p></dd>


<dt>verbose</dt>
<dd><p>Print messages.</p></dd>


<dt>top_SNPs</dt>
<dd><p>[deprecated]</p></dd>


<dt>PP_threshold</dt>
<dd><p>[deprecated]</p></dd>


<dt>consensus_threshold</dt>
<dd><p>[deprecated]</p></dd>


<dt>plot.Nott_epigenome</dt>
<dd><p>[deprecated]</p></dd>


<dt>plot.Nott_show_placseq</dt>
<dd><p>[deprecated]</p></dd>


<dt>plot.Nott_binwidth</dt>
<dd><p>[deprecated]</p></dd>


<dt>plot.Nott_bigwig_dir</dt>
<dd><p>[deprecated]</p></dd>


<dt>plot.Roadmap</dt>
<dd><p>[deprecated]</p></dd>


<dt>plot.Roadmap_query</dt>
<dd><p>[deprecated]</p></dd>


<dt>plot.XGR_libnames</dt>
<dd><p>[deprecated]</p></dd>


<dt>server</dt>
<dd><p>[deprecated]</p></dd>


<dt>plot.types</dt>
<dd><p>[deprecated]</p></dd>


<dt>plot.zoom</dt>
<dd><p>[deprecated]</p></dd>


<dt>QTL_prefixes</dt>
<dd><p>[deprecated]</p></dd>


<dt>vcf_folder</dt>
<dd><p>[deprecated]</p></dd>


<dt>probe_path</dt>
<dd><p>[deprecated]</p></dd>


<dt>file_sep</dt>
<dd><p>[deprecated]</p></dd>


<dt>chrom_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>chrom_type</dt>
<dd><p>[deprecated]</p></dd>


<dt>position_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>snp_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>pval_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>effect_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>stderr_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>tstat_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>locus_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>freq_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>MAF_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>A1_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>A2_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>gene_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>N_cases_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>N_controls_col</dt>
<dd><p>[deprecated]</p></dd>


<dt>N_cases</dt>
<dd><p>[deprecated]</p></dd>


<dt>N_controls</dt>
<dd><p>[deprecated]</p></dd>


<dt>proportion_cases</dt>
<dd><p>[deprecated]</p></dd>


<dt>sample_size</dt>
<dd><p>[deprecated]</p></dd>


<dt>PAINTOR_QTL_datasets</dt>
<dd><p>[deprecated]</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>By default, returns a nested list containing grouped by locus names
(e.g. <code>BST1</code>, <code>MEX3C</code>). The results of each locus contain
 the following elements:</p><ul><li><p><code>finemap_dat</code> :  Fine-mapping results from all selected methods
 merged with the original summary statistics
 (i.e. <strong>Multi-finemap results</strong>).</p></li>
<li><p><code>locus_plot</code> :  A nested list containing one or more
zoomed views of locus plots.</p></li>
<li><p><code>LD_matrix</code> :  The post-processed LD matrix used
for fine-mapping.</p></li>
<li><p><code>LD_plot</code> :  An LD plot (if used).</p></li>
<li><p><code>locus_dir</code> :  Locus directory results are saved in.</p></li>
<li><p><code>arguments</code> : A record of the arguments supplied to
finemap_loci.</p></li>
</ul><p>In addition, the following object summarizes the results
from all the locus-specific results:</p><ul><li><p><code>merged_dat</code> : A merged <a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a>
 with all fine-mapping results from all loci.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Other MAIN: 
<code><a href="finemap_locus.html">finemap_locus</a>()</code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">topSNPs</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/echodata/man/topSNPs_Nalls2019.html" class="external-link">topSNPs_Nalls2019</a></span></span></span>
<span class="r-in"><span><span class="va">fullSS_path</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/example_fullSS.html" class="external-link">example_fullSS</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"Nalls2019"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Writing file to ==&gt; /tmp/RtmpZmsFJq/nalls2019.fullSS_subset.tsv</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">echolocatoR</span><span class="fu">::</span><span class="fu">finemap_loci</span><span class="op">(</span></span></span>
<span class="r-in"><span>  fullSS_path <span class="op">=</span> <span class="va">fullSS_path</span>,</span></span>
<span class="r-in"><span>  topSNPs <span class="op">=</span> <span class="va">topSNPs</span>,</span></span>
<span class="r-in"><span>  loci <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BST1"</span>,<span class="st">"MEX3C"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  finemap_methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ABF"</span>,<span class="st">"FINEMAP"</span>,<span class="st">"SUSIE"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  dataset_name <span class="op">=</span> <span class="st">"Nalls23andMe_2019"</span>,</span></span>
<span class="r-in"><span>  fullSS_genome_build <span class="op">=</span> <span class="st">"hg19"</span>,</span></span>
<span class="r-in"><span>  bp_distance <span class="op">=</span> <span class="fl">1000</span>,</span></span>
<span class="r-in"><span>  munged <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "+ Assigning Gene and Locus independently."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Standardising column headers.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> First line of summary statistics file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Locus	Gene	CHR	POS	SNP	P	Effect	</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Returning unmapped column names without making them uppercase.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Mapping colnames from MungeSumstats ==&gt; echolocatoR</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ┌────────────────────────────────────────┐</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> │                                        │</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> │   <span style="color: #FFFFFF; font-weight: 100;">)))&gt;</span><span style="color: #55FFFF;"> 🦇 BST1 [locus 1 / 2] 🦇 </span><span style="color: #FFFFFF; font-weight: 100;">&lt;(((</span>   │</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> │                                        │</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> └────────────────────────────────────────┘</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 1 ▶▶▶ Query 🔎</span> <span style="color: #00BBBB;">─────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Query Method: tabix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing GRanges query using min/max ranges within a single chromosome.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> query_dat is already a GRanges object. Returning directly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========= echotabix::convert =========</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting full summary stats file to tabix format for fast querying.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferred format: 'table'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Explicit format: 'table'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferring comment_char from tabular header: 'SNP'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Determining chrom type from file header.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assuming fullSS_path summary stats have already been processed with MungeSumstats.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Chromosome format: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detecting column delimiter.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified column separator: \t</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sorting rows by coordinates via bash.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Searching for header row with grep.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ( grep ^'SNP' .../nalls2019.fullSS_subset.tsv; grep</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     -v ^'SNP' .../nalls2019.fullSS_subset.tsv | sort</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     -k2,2n</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     -k3,3n ) &gt; .../file1eda3c341338_sorted.tsv</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing outputs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> bgzipping file with Rsamtools.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading bgzipped file using: data.table</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Header preview:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>           SNP CHR       BP A1 A2   FREQ    BETA     SE      P N_CAS   N_CON</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1:   rs741214   4 13737637  C  A 0.2341  0.0091 0.0119 0.4460 49053 1411006</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2: rs79385021   4 13737659  T  C 0.9109  0.0009 0.0198 0.9642 30435  974587</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3:   rs741215   4 13737767  C  T 0.0733  0.0190 0.0194 0.3284 49053 1411006</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 4:   rs759261   4 13737915  G  A 0.4511  0.0075 0.0100 0.4543 49053 1411006</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 5: rs35006360   4 13738014  G  A 0.1148 -0.0195 0.0160 0.2208 49053 1411006</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Tabix-indexing file using: Rsamtools</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Data successfully converted to bgzip-compressed, tabix-indexed format.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========= echotabix::query =========</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> query_dat is already a GRanges object. Returning directly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferred format: 'table'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying tabular tabix file using: Rsamtools.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking query chromosome style is correct.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Chromosome format: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieving data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting query results to data.table.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing query: 4:15736348-15738348</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding 'query' column to results.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved data with 17 rows</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving query ==&gt; /tmp/RtmpZmsFJq/results/GWAS/Nalls23andMe_2019/BST1/BST1_Nalls23andMe_2019_subset.tsv.gz</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Query: 17 SNPs x 11 columns.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Standardizing summary statistics subset.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Standardizing main column names.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Preparing A1,A1 cols</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Preparing MAF,Freq cols.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Could not infer MAF from Freq.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Preparing N_cases,N_controls cols.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Preparing proportion_cases col.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Calculating proportion_cases from N_cases and N_controls.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing sample size column (N).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing effective sample size using the LDSC method:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Neff = (N_CAS+N_CON) * (N_CAS/(N_CAS+N_CON)) / mean((N_CAS/(N_CAS+N_CON))[(N_CAS+N_CON)==max(N_CAS+N_CON)]))</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Mapping colnames from MungeSumstats ==&gt; echolocatoR</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Imputing t-statistic from Effect and StdErr.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + leadSNP missing. Assigning new one by min p-value.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Ensuring Effect,StdErr,P are numeric.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Ensuring 1 SNP per row and per genomic coordinate.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Removing extra whitespace</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Standardized query: 17 SNPs x 15 columns.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Saving standardized query ==&gt; /tmp/RtmpZmsFJq/results/GWAS/Nalls23andMe_2019/BST1/BST1_Nalls23andMe_2019_subset.tsv.gz</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 2 ▶▶▶ Extract Linkage Disequilibrium 🔗</span> <span style="color: #00BBBB;">────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> LD_reference identified as: 1kg.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 1000Genomes as LD reference panel.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing GRanges query using min/max ranges across one or more chromosomes.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + as_blocks=TRUE: Will query a single range per chromosome that covers all regions requested (plus anything in between).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> LD Reference Panel = 1KGphase3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying 1KG remote server.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selecting 503 EUR individuals from 1kgphase3.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========= echotabix::query =========</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> query_dat is already a GRanges object. Returning directly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Explicit format: 'vcf'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying VCF tabix file.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying VCF file using: VariantAnnotation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking query chromosome style is correct.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Chromosome format: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering query to 503 samples and returning ScanVcfParam object.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieving data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 1.7 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing 17 / 66 non-overlapping SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving VCF subset ==&gt; /tmp/RtmpZmsFJq/VCF/RtmpZmsFJq.chr4-15736493-15738253.ALL.chr4.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.bgz</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>converting NULL pointer to R NULL</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 0.3 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved data with 17 rows across 503 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> echoLD::snpStats:: Filling `MAF` column with MAF from LD panel.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> echoLD:snpStats:: Computing pairwise LD between 17 SNPs across 503 individuals (stats = R).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 17 x 17 LD_matrix (sparse)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting obj to sparseMatrix.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving sparse LD matrix ==&gt; /tmp/RtmpZmsFJq/results/GWAS/Nalls23andMe_2019/BST1/LD/BST1.1KGphase3_LD.RDS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing 1 temp files.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + FILTER:: Filtering by LD features.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 3 ▶▶▶ Filter SNPs 🚰</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> FILTER:: Filtering by SNP features.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + FILTER:: Post-filtered data: 17 x 16</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Subsetting LD matrix and dat to common SNPs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing unnamed rows/cols</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Replacing NAs with 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + LD_matrix = 17 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + dat = 17 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 17 SNPs in common.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting obj to sparseMatrix.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 4 ▶▶▶ Fine-map 🔊</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method sources.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method citations.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing sample size column (N).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using existing 'N' column.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method sources.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method citations.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method sources.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method citations.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="background-color: #FF55FF;">ABF</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">✅ All required columns present.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="background-color: #FF55FF;">FINEMAP</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">✅ All required columns present.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">✅ All optional columns present.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="background-color: #FF55FF;">SUSIE</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">✅ All required columns present.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">✅ All optional columns present.</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Fine-mapping using 3 tool(s): ABF, FINEMAP, SUSIE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> +++ Multi-finemap:: ABF +++</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using all default values for finemap_args$ABF</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing sample size column (N).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using existing 'N' column.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running ABF.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Credible Set SNPs identified = 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Merging ABF results with multi-finemap data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> +++ Multi-finemap:: FINEMAP +++</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using all default values for finemap_args$FINEMAP</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing sample size column (N).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using existing 'N' column.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Subsetting LD matrix and dat to common SNPs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing unnamed rows/cols</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Replacing NAs with 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + LD_matrix = 17 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + dat = 17 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 17 SNPs in common.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting obj to sparseMatrix.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing master file.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing data.z file.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing data.ld file.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> FINEMAP path: /github/home/.cache/R/echofinemap/FINEMAP/finemap_v1.4.1_x86_64/finemap_v1.4.1_x86_64</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferred FINEMAP version: 1.4.1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running FINEMAP.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> cd .../BST1 &amp;&amp;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     .../finemap_v1.4.1_x86_64</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     --sss</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     --in-files .../master</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     --log</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     --n-causal-snps 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |--------------------------------------|</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | Welcome to FINEMAP v1.4.1            |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |                                      |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | (c) 2015-2022 University of Helsinki |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |                                      |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | Help :                               |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | - ./finemap --help                   |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | - www.finemap.me                     |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | - www.christianbenner.com            |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |                                      |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | Contact :                            |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | - finemap@christianbenner.com        |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | - matti.pirinen@helsinki.fi          |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |--------------------------------------|</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SETTINGS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - dataset            : all</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - corr-config        : 0.95</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - n-causal-snps      : 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - n-configs-top      : 50000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - n-conv-sss         : 100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - n-iter             : 100000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - n-threads          : 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - prior-k0           : 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - prior-std          : 0.05 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - prob-conv-sss-tol  : 0.001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - prob-cred-set      : 0.95</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FINE-MAPPING (1/1)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - GWAS summary stats               : FINEMAP/data.z</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - SNP correlations                 : FINEMAP/data.ld</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Causal SNP stats                 : FINEMAP/data.snp</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Causal configurations            : FINEMAP/data.config</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Credible sets                    : FINEMAP/data.cred</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Log file                         : FINEMAP/data.log_sss</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Reading summary statistics       : +- Reading summary statistics       : -- Reading summary statistics       : +- Reading summary stats            : done!   - Reading SNP correlations         : 0.000%- Reading SNP correlations         : done!   - Reading input                    : done!   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Number of GWAS samples           : 1474097</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Number of SNPs                   : 17</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Prior-Pr(# of causal SNPs is k)  : </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   (0 -&gt; 0)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1 -&gt; 0.59</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2 -&gt; 0.295</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    3 -&gt; 0.0921</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    4 -&gt; 0.0202</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    5 -&gt; 0.00327</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0 configurations evaluated (0.000/100%)- 17 configurations evaluated (0.000/100%)- 31 configurations evaluated (0.000/100%)- 262 configurations evaluated (0.050/100%)- 322 configurations evaluated (0.100/100%)- 520 configurations evaluated (0.150/100%)- 604 configurations evaluated (0.200/100%)- 622 configurations evaluated (0.250/100%)- 653 configurations evaluated (0.299/100%) : converged after 299 iterations</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Computing causal SNP statistics  : 0.153%- Computing causal SNP statistics  : 7.810%- Computing causal SNP statistics  : 15.467%- Computing causal SNP statistics  : 23.124%- Computing causal SNP statistics  : 30.781%- Computing causal SNP statistics  : 38.438%- Computing causal SNP statistics  : 46.095%- Computing causal SNP statistics  : 53.752%- Computing causal SNP statistics  : 61.409%- Computing causal SNP statistics  : 69.066%- Computing causal SNP statistics  : 76.723%- Computing causal SNP statistics  : 84.380%- Computing causal SNP statistics  : 92.037%- Computing causal SNP statistics  : 99.694%- Computing causal SNP statistics  : done!   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Regional SNP heritability        : 9.81e-05 (SD: 1.61e-05 ; 95% CI: [6.96e-05,0.000131])</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Log10-BF of &gt;= one causal SNP    : 62.4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Post-expected # of causal SNPs   : 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Post-Pr(# of causal SNPs is k)   : </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   (0 -&gt; 0)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1 -&gt; 1.08e-39</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2 -&gt; 3.5e-38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    3 -&gt; 1.87e-38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    4 -&gt; 3.1e-38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    5 -&gt; 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Computing credible sets          : - Writing causal configurations    : 0.153%- Writing causal configurations    : 7.810%- Writing causal configurations    : 15.467%- Writing causal configurations    : 23.124%- Writing causal configurations    : 30.781%- Writing causal configurations    : 38.438%- Writing causal configurations    : 46.095%- Writing causal configurations    : 53.752%- Writing causal configurations    : 61.409%- Writing causal configurations    : 69.066%- Writing causal configurations    : 76.723%- Writing causal configurations    : 84.380%- Writing causal configurations    : 92.037%- Writing causal configurations    : 99.694%- Writing causal SNP statistics    : 5.882%- Writing causal SNP statistics    : 64.706%- Writing output                   : done!   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Run time                         : 0 hours, 0 minutes, 0 seconds</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 data.cred* file(s) found in the same subfolder.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selected file based on postPr_k: data.cred5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Importing conditional probabilities (.cred file).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Importing marginal probabilities (.snp file).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Importing configuration probabilities (.config file).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Credible Set SNPs identified = 12</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Merging FINEMAP results with multi-finemap data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> +++ Multi-finemap:: SUSIE +++</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using all default values for finemap_args$SUSIE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required namespace: Rfast</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing sample size column (N).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using existing 'N' column.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + SUSIE:: sample_size=1,474,097</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Subsetting LD matrix and dat to common SNPs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing unnamed rows/cols</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Replacing NAs with 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + LD_matrix = 17 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + dat = 17 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 17 SNPs in common.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting obj to sparseMatrix.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + SUSIE:: Using `susie_rss()` from susieR v0.12.27</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + SUSIE:: Extracting Credible Sets.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Credible Set SNPs identified = 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Merging SUSIE results with multi-finemap data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identifying Consensus SNPs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + support_thresh = 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Calculating mean Posterior Probability (mean.PP)...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 3 fine-mapping methods used.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 12 Credible Set SNPs identified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 0 Consensus SNPs identified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving merged finemapping results ==&gt; /tmp/RtmpZmsFJq/results/GWAS/Nalls23andMe_2019/BST1/Multi-finemap/1KGphase3_LD.Multi-finemap.tsv.gz</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Fine-mapping with 'ABF, FINEMAP, SUSIE' completed:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 5 ▶▶▶ Plot 📈</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> +-------- Locus Plot:  BST1 --------+</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + support_thresh = 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Calculating mean Posterior Probability (mean.PP)...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 3 fine-mapping methods used.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 12 Credible Set SNPs identified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 0 Consensus SNPs identified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Filling NAs in CS cols with 0.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Filling NAs in PP cols with 0.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> LD_matrix detected. Coloring SNPs by LD with lead SNP.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ echoplot:: GWAS full window track</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ echoplot:: GWAS track</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ echoplot:: Merged fine-mapping track</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Melting PP and CS from 4 fine-mapping methods.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ echoplot:: Adding Gene model track.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting dat to GRanges object.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required namespace: EnsDb.Hsapiens.v75</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> max_transcripts= 1 . </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1  transcripts from  1  genes returned.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Fetching data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Parsing exons...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Defining introns...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Defining UTRs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Defining CDS...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> aggregating...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing graphics...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Adding vertical lines to highlight SNP groups...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> +&gt;+&gt;+&gt;+&gt;+ zoom =  1x  +&lt;+&lt;+&lt;+&lt;+</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + echoplot:: Get window suffix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + echoplot:: Removing GWAS full window track @ zoom=1x</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Removing subplot margins...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Reordering tracks...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "+ Ensuring last track shows genomic units..."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Aligning xlimits for each subplot...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Checking track heights...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + echoplot:: Saving plot ==&gt; /tmp/RtmpZmsFJq/results/GWAS/Nalls23andMe_2019/BST1/multiview.BST1.1KGphase3.1x.png</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "simpleUnit" in cache; using the first, from namespace 'hexbin'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by 'ggbio'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Recording all `finemap_locus` arguments.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Formatting locus results.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ┌─────────────────────────────────────────┐</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> │                                         │</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> │   <span style="color: #FFFFFF; font-weight: 100;">)))&gt;</span><span style="color: #55FFFF;"> 🦇 MEX3C [locus 2 / 2] 🦇 </span><span style="color: #FFFFFF; font-weight: 100;">&lt;(((</span>   │</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> │                                         │</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> └─────────────────────────────────────────┘</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 1 ▶▶▶ Query 🔎</span> <span style="color: #00BBBB;">─────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Query Method: tabix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing GRanges query using min/max ranges within a single chromosome.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> query_dat is already a GRanges object. Returning directly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========= echotabix::convert =========</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting full summary stats file to tabix format for fast querying.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferred format: 'table'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Explicit format: 'table'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferring comment_char from tabular header: 'SNP'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Determining chrom type from file header.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Chromosome format: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detecting column delimiter.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified column separator: \t</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sorting rows by coordinates via bash.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Searching for header row with grep.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ( grep ^'SNP' .../nalls2019.fullSS_subset.tsv; grep</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     -v ^'SNP' .../nalls2019.fullSS_subset.tsv | sort</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     -k2,2n</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     -k3,3n ) &gt; .../file1eda6a1e6c55_sorted.tsv</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing outputs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using existing bgzipped file: /tmp/RtmpZmsFJq/nalls2019.fullSS_subset.tsv.bgz </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set force_new=TRUE to override this.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Tabix-indexing file using: Rsamtools</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Data successfully converted to bgzip-compressed, tabix-indexed format.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========= echotabix::query =========</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> query_dat is already a GRanges object. Returning directly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferred format: 'table'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying tabular tabix file using: Rsamtools.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking query chromosome style is correct.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Chromosome format: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieving data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting query results to data.table.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing query: 18:48682589-48684589</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding 'query' column to results.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved data with 7 rows</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving query ==&gt; /tmp/RtmpZmsFJq/results/GWAS/Nalls23andMe_2019/MEX3C/MEX3C_Nalls23andMe_2019_subset.tsv.gz</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Query: 7 SNPs x 11 columns.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Standardizing summary statistics subset.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Standardizing main column names.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Preparing A1,A1 cols</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Preparing MAF,Freq cols.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Could not infer MAF from Freq.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Preparing N_cases,N_controls cols.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Preparing proportion_cases col.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Calculating proportion_cases from N_cases and N_controls.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing sample size column (N).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing effective sample size using the LDSC method:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Neff = (N_CAS+N_CON) * (N_CAS/(N_CAS+N_CON)) / mean((N_CAS/(N_CAS+N_CON))[(N_CAS+N_CON)==max(N_CAS+N_CON)]))</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Mapping colnames from MungeSumstats ==&gt; echolocatoR</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Imputing t-statistic from Effect and StdErr.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + leadSNP missing. Assigning new one by min p-value.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Ensuring Effect,StdErr,P are numeric.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Ensuring 1 SNP per row and per genomic coordinate.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Removing extra whitespace</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Standardized query: 7 SNPs x 15 columns.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Saving standardized query ==&gt; /tmp/RtmpZmsFJq/results/GWAS/Nalls23andMe_2019/MEX3C/MEX3C_Nalls23andMe_2019_subset.tsv.gz</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 2 ▶▶▶ Extract Linkage Disequilibrium 🔗</span> <span style="color: #00BBBB;">────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> LD_reference identified as: 1kg.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 1000Genomes as LD reference panel.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing GRanges query using min/max ranges across one or more chromosomes.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + as_blocks=TRUE: Will query a single range per chromosome that covers all regions requested (plus anything in between).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> LD Reference Panel = 1KGphase3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying 1KG remote server.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selecting 503 EUR individuals from 1kgphase3.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========= echotabix::query =========</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> query_dat is already a GRanges object. Returning directly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Explicit format: 'vcf'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying VCF tabix file.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying VCF file using: VariantAnnotation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking query chromosome style is correct.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Chromosome format: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering query to 503 samples and returning ScanVcfParam object.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieving data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 1.3 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing 7 / 38 non-overlapping SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving VCF subset ==&gt; /tmp/RtmpZmsFJq/VCF/RtmpZmsFJq.chr18-48682662-48684103.ALL.chr18.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.bgz</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>converting NULL pointer to R NULL</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 0.2 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieved data with 7 rows across 503 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> echoLD::snpStats:: Filling `MAF` column with MAF from LD panel.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> echoLD:snpStats:: Computing pairwise LD between 7 SNPs across 503 individuals (stats = R).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 7 x 7 LD_matrix (sparse)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting obj to sparseMatrix.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving sparse LD matrix ==&gt; /tmp/RtmpZmsFJq/results/GWAS/Nalls23andMe_2019/MEX3C/LD/MEX3C.1KGphase3_LD.RDS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing 1 temp files.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + FILTER:: Filtering by LD features.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 3 ▶▶▶ Filter SNPs 🚰</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> FILTER:: Filtering by SNP features.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + FILTER:: Post-filtered data: 5 x 16</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Subsetting LD matrix and dat to common SNPs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing unnamed rows/cols</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Replacing NAs with 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + LD_matrix = 7 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + dat = 5 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 5 SNPs in common.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting obj to sparseMatrix.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 4 ▶▶▶ Fine-map 🔊</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method sources.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method citations.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing sample size column (N).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using existing 'N' column.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method sources.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method citations.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method sources.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gathering method citations.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="background-color: #FF55FF;">ABF</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">✅ All required columns present.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="background-color: #FF55FF;">FINEMAP</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">✅ All required columns present.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">✅ All optional columns present.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="background-color: #FF55FF;">SUSIE</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">✅ All required columns present.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">✅ All optional columns present.</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Fine-mapping using 3 tool(s): ABF, FINEMAP, SUSIE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> +++ Multi-finemap:: ABF +++</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using all default values for finemap_args$ABF</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing sample size column (N).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using existing 'N' column.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running ABF.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Credible Set SNPs identified = 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Merging ABF results with multi-finemap data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> +++ Multi-finemap:: FINEMAP +++</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using all default values for finemap_args$FINEMAP</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing sample size column (N).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using existing 'N' column.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Subsetting LD matrix and dat to common SNPs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing unnamed rows/cols</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Replacing NAs with 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + LD_matrix = 5 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + dat = 5 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 5 SNPs in common.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting obj to sparseMatrix.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing master file.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing data.z file.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing data.ld file.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> FINEMAP path: /github/home/.cache/R/echofinemap/FINEMAP/finemap_v1.4.1_x86_64/finemap_v1.4.1_x86_64</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferred FINEMAP version: 1.4.1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running FINEMAP.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> cd .../MEX3C &amp;&amp;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     .../finemap_v1.4.1_x86_64</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     --sss</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     --in-files .../master</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     --log</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     --n-causal-snps 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |--------------------------------------|</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | Welcome to FINEMAP v1.4.1            |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |                                      |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | (c) 2015-2022 University of Helsinki |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |                                      |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | Help :                               |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | - ./finemap --help                   |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | - www.finemap.me                     |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | - www.christianbenner.com            |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |                                      |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | Contact :                            |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | - finemap@christianbenner.com        |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> | - matti.pirinen@helsinki.fi          |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |--------------------------------------|</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SETTINGS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - dataset            : all</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - corr-config        : 0.95</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - n-causal-snps      : 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - n-configs-top      : 50000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - n-conv-sss         : 100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - n-iter             : 100000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - n-threads          : 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - prior-k0           : 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - prior-std          : 0.05 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - prob-conv-sss-tol  : 0.001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - prob-cred-set      : 0.95</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FINE-MAPPING (1/1)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - GWAS summary stats               : FINEMAP/data.z</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - SNP correlations                 : FINEMAP/data.ld</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Causal SNP stats                 : FINEMAP/data.snp</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Causal configurations            : FINEMAP/data.config</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Credible sets                    : FINEMAP/data.cred</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Log file                         : FINEMAP/data.log_sss</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Reading summary statistics       : +- Reading summary stats            : done!   - Reading SNP correlations         : 0.000%- Reading SNP correlations         : done!   - Reading input                    : done!   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Number of GWAS samples           : 1460059</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Number of SNPs                   : 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Prior-Pr(# of causal SNPs is k)  : </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   (0 -&gt; 0)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1 -&gt; 0.609</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2 -&gt; 0.305</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    3 -&gt; 0.0762</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    4 -&gt; 0.00952</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    5 -&gt; 0.000476</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0 configurations evaluated (0.000/100%)- 5 configurations evaluated (0.000/100%)- 7 configurations evaluated (0.000/100%)- 12 configurations evaluated (0.050/100%)- 12 configurations evaluated (0.100/100%)- 12 configurations evaluated (0.106/100%) : converged after 106 iterations</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Computing causal SNP statistics  : 8.333%- Computing causal SNP statistics  : done!   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Regional SNP heritability        : 1.93e-05 (SD: 7.4e-06 ; 95% CI: [7.12e-06,3.53e-05])</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Log10-BF of &gt;= one causal SNP    : 3.56</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Post-expected # of causal SNPs   : 1.09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Post-Pr(# of causal SNPs is k)   : </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   (0 -&gt; 0)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1 -&gt; 0.915</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2 -&gt; 0.0853</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    3 -&gt; 0.000204</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    4 -&gt; 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    5 -&gt; 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Computing credible sets          : - Writing causal configurations    : 8.333%- Writing causal SNP statistics    : 20.000%- Writing output                   : done!   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - Run time                         : 0 hours, 0 minutes, 0 seconds</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2 data.cred* file(s) found in the same subfolder.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selected file based on postPr_k: data.cred1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Importing conditional probabilities (.cred file).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No configurations were causal at PP&gt;=0.95.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Importing marginal probabilities (.snp file).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Importing configuration probabilities (.config file).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> FINEMAP was unable to identify any credible sets at PP&gt;=0.95.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Credible Set SNPs identified = 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Merging FINEMAP results with multi-finemap data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> +++ Multi-finemap:: SUSIE +++</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using all default values for finemap_args$SUSIE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing sample size column (N).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using existing 'N' column.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + SUSIE:: sample_size=1,460,059</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Subsetting LD matrix and dat to common SNPs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing unnamed rows/cols</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Replacing NAs with 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + LD_matrix = 5 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + dat = 5 SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 5 SNPs in common.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting obj to sparseMatrix.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + SUSIE:: Using `susie_rss()` from susieR v0.12.27</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + SUSIE:: Extracting Credible Sets.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Credible Set SNPs identified = 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ Merging SUSIE results with multi-finemap data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identifying Consensus SNPs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + support_thresh = 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Calculating mean Posterior Probability (mean.PP)...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 3 fine-mapping methods used.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 0 Credible Set SNPs identified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 0 Consensus SNPs identified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving merged finemapping results ==&gt; /tmp/RtmpZmsFJq/results/GWAS/Nalls23andMe_2019/MEX3C/Multi-finemap/1KGphase3_LD.Multi-finemap.tsv.gz</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Fine-mapping with 'ABF, FINEMAP, SUSIE' completed:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 5 ▶▶▶ Plot 📈</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> +-------- Locus Plot:  MEX3C --------+</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + support_thresh = 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Calculating mean Posterior Probability (mean.PP)...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 3 fine-mapping methods used.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 0 Credible Set SNPs identified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + 0 Consensus SNPs identified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Filling NAs in CS cols with 0.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Filling NAs in PP cols with 0.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> LD_matrix detected. Coloring SNPs by LD with lead SNP.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ echoplot:: GWAS full window track</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ echoplot:: GWAS track</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ echoplot:: Merged fine-mapping track</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Melting PP and CS from 4 fine-mapping methods.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ++ echoplot:: Adding Gene model track.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting dat to GRanges object.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> max_transcripts= 1 . </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0  transcripts from  0  genes returned.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Returning NULL for gene track.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Adding vertical lines to highlight SNP groups...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> +&gt;+&gt;+&gt;+&gt;+ zoom =  1x  +&lt;+&lt;+&lt;+&lt;+</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + echoplot:: Get window suffix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + echoplot:: Removing GWAS full window track @ zoom=1x</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Removing subplot margins...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Reordering tracks...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "+ Ensuring last track shows genomic units..."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Aligning xlimits for each subplot...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + Checking track heights...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> + echoplot:: Saving plot ==&gt; /tmp/RtmpZmsFJq/results/GWAS/Nalls23andMe_2019/MEX3C/multiview.MEX3C.1KGphase3.1x.png</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 1 rows containing missing values (geom_text).</span>
<span class="r-plt img"><img src="finemap_loci-1.png" alt="" width="700" height="433"></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 1 rows containing missing values (geom_text).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Recording all `finemap_locus` arguments.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Formatting locus results.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 6 ▶▶▶ Postprocess data 🎁</span> <span style="color: #00BBBB;">──────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Returning results as nested list.</span>
<span class="r-plt img"><img src="finemap_loci-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #55FFFF;">All loci done in: </span><span style="color: #FFFFFF;">0.27 min</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

