<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting started • echolocatoR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting started">
<meta property="og:description" content="echolocatoR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">echolocatoR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/echolocatoR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/QTLs.html">QTLs</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="../articles/finemapping_portal.html">echolocatoR Fine-mapping Portal</a>
    </li>
    <li>
      <a href="../articles/plot_locus.html">Plot locus</a>
    </li>
    <li>
      <a href="../articles/summarise.html">Summarise</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RajLabMSSM/echolocatoR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting started</h1>
                        <h4 data-toc-skip class="author"><h4> ¶ Author:
<i>Brian M. Schilder</i> ¶ </h4></h4>
            
            <h4 data-toc-skip class="date"><h4> ¶ Updated:
<i>Nov-07-2022</i> ¶ </h4></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/RajLabMSSM/echolocatoR/blob/HEAD/vignettes/echolocatoR.Rmd" class="external-link"><code>vignettes/echolocatoR.Rmd</code></a></small>
      <div class="hidden name"><code>echolocatoR.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RajLabMSSM/echolocatoR" class="external-link">echolocatoR</a></span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">##   method from   </span></span>
<span><span class="co">##   +.gg   ggplot2</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">⠊⠉⠡⣀⣀⠊⠉⠡⣀⣀⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⢁⡐⠉⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠉⢂⡈⠑⣀⠉⢄⡈⠡⣀</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⡈⡐⢂⢁⠒⡈⡐⢂⢁⠒⡈⡐⢂⢁⠑⡈⡈⢄⢁⠡⠌⡈⠤⢁⠡⠌⡈⠤⢁⠡⠌⡈⡠⢁⢁⠊⡈⡐⢂</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="color: #55FFFF; font-weight: bold;">🦇  🦇  🦇 e c h o l o c a t o R 🦇  🦇  🦇</span> <span style="color: #00BBBB;">─────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="color: #55FFFF; font-weight: bold;">v2.0.3</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">⠌⡈⡐⢂⢁⠒⡈⡐⢂⢁⠒⡈⡐⢂⢁⠑⡈⡈⢄⢁⠡⠌⡈⠤⢁⠡⠌⡈⠤⢁⠡⠌⡈⡠⢁⢁⠊⡈⡐⢂</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⢁⡐⠉⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠉⢂⡈⠑⣀⠉⢄⡈⠡⣀</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠊⠉⠡⣀⣀⠊⠉⠡⣀⣀⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠</span>                                    </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ If you use </span><span style="color: #00BBBB; font-weight: bold;">echolocatoR</span><span style="color: #00BBBB;"> or any of the </span><span style="color: #00BBBB; font-style: italic;">echoverse</span><span style="color: #00BBBB;"> subpackages, please cite:</span>      </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶ Brian M Schilder, Jack Humphrey, Towfique</span>                                </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     Raj (2021) echolocatoR: an automated</span>                                       </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     end-to-end statistical and functional</span>                                      </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     genomic fine-mapping pipeline,</span>                                             </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     Bioinformatics; btab658,</span>                                                   </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://doi.org/10.1093/bioinformatics/btab658</span>                             </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ Please report any bugs/feature requests on GitHub:</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://github.com/RajLabMSSM/echolocatoR/issues</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ Contributions are welcome!:</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://github.com/RajLabMSSM/echolocatoR/pulls</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<div class="section level2">
<h2 id="full-pipeline">Full pipeline<a class="anchor" aria-label="anchor" href="#full-pipeline"></a>
</h2>
<p>All examples below use data from the Parkinson’s disease GWAS by
Nalls et al. (2019).</p>
<div class="section level3">
<h3 id="prepare-top_snps-data-frame">Prepare <code>top_SNPs</code> data.frame<a class="anchor" aria-label="anchor" href="#prepare-top_snps-data-frame"></a>
</h3>
<ul>
<li>To enable rapid fine-mapping of many loci, you can create a
<code>top_SNPs</code> data.frame<br>
which contains the position of the lead/index SNP within each
locus.</li>
<li>
<code><a href="../reference/finemap_loci.html">finemap_loci()</a></code> (see next step) will then use this info
to extract subsets of the<br>
full GWAS/QTL summary statistics using windows centered on each
lead/index SNP.</li>
<li>The <code>topSS</code> argument can either be a data.frame, or a
path to a topSS file saved somehwere. Most common tabular data formats
(e.g. .tsv, .csv, .xlsx) are accepted.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#### Load example top SNPs (pre-formatted) ####</span></span>
<span><span class="va">topSS</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/echodata/man/topSNPs_Nalls2019_raw.html" class="external-link">topSNPs_Nalls2019_raw</a></span></span>
<span><span class="co">#### construct a column mapping object ####</span></span>
<span><span class="va">colmap</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/construct_colmap.html" class="external-link">construct_colmap</a></span><span class="op">(</span>P <span class="op">=</span> <span class="st">"P, all studies"</span>,</span>
<span>                                     Effect <span class="op">=</span> <span class="st">"Beta, all studies"</span>,</span>
<span>                                     Locus <span class="op">=</span> <span class="st">"Nearest Gene"</span>,</span>
<span>                                     Gene <span class="op">=</span> <span class="st">"QTL Nominated Gene (nearest QTL)"</span><span class="op">)</span></span>
<span><span class="co">#### Import top SNPs ####</span></span>
<span><span class="va">topSNPs</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/import_topSNPs.html" class="external-link">import_topSNPs</a></span><span class="op">(</span></span>
<span>    topSS <span class="op">=</span> <span class="fu">echodata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/echodata/man/topSNPs_Nalls2019_raw.html" class="external-link">topSNPs_Nalls2019_raw</a></span>,</span>
<span>    colmap <span class="op">=</span> <span class="va">colmap</span>,</span>
<span>    grouping_vars <span class="op">=</span> <span class="st">"Locus Number"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required namespace: MungeSumstats</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: P, all studies ==&gt; P</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: Beta, all studies ==&gt; Effect</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: Nearest Gene ==&gt; Locus</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: QTL Nominated Gene (nearest QTL) ==&gt; Gene</span></span></code></pre>
<pre><code><span><span class="co">## [1] "+ Assigning Gene and Locus independently."</span></span></code></pre>
<pre><code><span><span class="co">## Standardising column headers.</span></span></code></pre>
<pre><code><span><span class="co">## First line of summary statistics file:</span></span></code></pre>
<pre><code><span><span class="co">## SNP  CHR BP  Locus   Gene    Effect allele   Other allele    Effect allele frequency Effect  SE, all studies P   P, COJO, all studies    P, random effects, all studies  P, Conditional 23AndMe only P, 23AndMe only I2, all studies Freq1, previous studies Beta, previous studies  StdErr, previous studies    P, previous studies I2, previous studies    Freq1, new studies  Beta, new studies   StdErr, new studies P, new studies  I2, new studies Passes pooled 23andMe QC    Known GWAS locus within 1MB Failed final filtering and QC   Locus within 250KB  Locus Number    </span></span></code></pre>
<pre><code><span><span class="co">## Returning unmapped column names without making them uppercase.</span></span></code></pre>
<pre><code><span><span class="co">## + Mapping colnames from MungeSumstats ==&gt; echolocatoR</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">topSNPs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           SNP CHR       POS   Locus    Gene A2 A1 Effect allele frequency</span></span>
<span><span class="co">## 1:  rs1941685  18  31304318   ASXL3    &lt;NA&gt;  t  g                  0.4983</span></span>
<span><span class="co">## 2:  rs2280104   8  22525980    BIN3    BIN3  t  c                  0.3604</span></span>
<span><span class="co">## 3: rs61169879  17  59917366   BRIP1   MED13  t  c                  0.1641</span></span>
<span><span class="co">## 4:  rs4698412   4  15737348    BST1    BST1  a  g                  0.5529</span></span>
<span><span class="co">## 5: rs11950533   5 134199105 C5orf24 TXNDC15  a  c                  0.1020</span></span>
<span><span class="co">## 6:  rs9568188  13  49927732  CAB39L  CAB39L  t  c                  0.7397</span></span>
<span><span class="co">##     Effect SE, all studies        P P, COJO, all studies</span></span>
<span><span class="co">## 1:  0.0531          0.0094 1.69e-08             1.61e-08</span></span>
<span><span class="co">## 2:  0.0556          0.0098 1.16e-08             1.40e-08</span></span>
<span><span class="co">## 3:  0.0820          0.0134 9.28e-10             9.40e-10</span></span>
<span><span class="co">## 4:  0.1035          0.0094 2.06e-28             9.73e-36</span></span>
<span><span class="co">## 5: -0.0916          0.0158 7.16e-09             6.73e-09</span></span>
<span><span class="co">## 6:  0.0617          0.0108 1.15e-08             1.11e-08</span></span>
<span><span class="co">##    P, random effects, all studies P, Conditional 23AndMe only P, 23AndMe only</span></span>
<span><span class="co">## 1:                      1.690e-08                    1.64e-08        1.60e-08</span></span>
<span><span class="co">## 2:                      1.860e-06                    4.35e-02        4.94e-02</span></span>
<span><span class="co">## 3:                      6.210e-06                    9.07e-07        2.05e-06</span></span>
<span><span class="co">## 4:                      1.680e-19                    1.05e-07        1.15e-07</span></span>
<span><span class="co">## 5:                      2.680e-08                    5.08e-04        4.22e-04</span></span>
<span><span class="co">## 6:                      2.458e-04                    4.29e-06        4.41e-06</span></span>
<span><span class="co">##    I2, all studies Freq1, previous studies Beta, previous studies</span></span>
<span><span class="co">## 1:             0.0                  0.4978                 0.0507</span></span>
<span><span class="co">## 2:             8.9                  0.3595                 0.0647</span></span>
<span><span class="co">## 3:            16.4                  0.1641                 0.0849</span></span>
<span><span class="co">## 4:            13.9                  0.5547                 0.1023</span></span>
<span><span class="co">## 5:             1.9                  0.1008                -0.0986</span></span>
<span><span class="co">## 6:            21.4                  0.7405                 0.0657</span></span>
<span><span class="co">##    StdErr, previous studies P, previous studies I2, previous studies</span></span>
<span><span class="co">## 1:                   0.0113            6.82e-06                  0.0</span></span>
<span><span class="co">## 2:                   0.0117            3.23e-08                 15.0</span></span>
<span><span class="co">## 3:                   0.0163            1.90e-07                  0.0</span></span>
<span><span class="co">## 4:                   0.0112            7.32e-20                 49.3</span></span>
<span><span class="co">## 5:                   0.0190            2.05e-07                 43.8</span></span>
<span><span class="co">## 6:                   0.0130            4.00e-07                 39.8</span></span>
<span><span class="co">##    Freq1, new studies Beta, new studies StdErr, new studies P, new studies</span></span>
<span><span class="co">## 1:             0.4995            0.0586              0.0171      6.147e-04</span></span>
<span><span class="co">## 2:             0.3626            0.0350              0.0177      4.724e-02</span></span>
<span><span class="co">## 3:             0.1641            0.0761              0.0236      1.242e-03</span></span>
<span><span class="co">## 4:             0.5485            0.1062              0.0170      4.160e-10</span></span>
<span><span class="co">## 5:             0.1048           -0.0756              0.0287      8.388e-03</span></span>
<span><span class="co">## 6:             0.7380            0.0526              0.0196      7.353e-03</span></span>
<span><span class="co">##    I2, new studies Passes pooled 23andMe QC Known GWAS locus within 1MB</span></span>
<span><span class="co">## 1:            14.3                        T                           0</span></span>
<span><span class="co">## 2:             1.9                        T                           1</span></span>
<span><span class="co">## 3:            23.7                        T                           0</span></span>
<span><span class="co">## 4:            10.9                        T                           1</span></span>
<span><span class="co">## 5:             0.0                        T                           0</span></span>
<span><span class="co">## 6:            22.3                        T                           0</span></span>
<span><span class="co">##    Failed final filtering and QC Locus within 250KB Locus Number</span></span>
<span><span class="co">## 1:                             0                  0           73</span></span>
<span><span class="co">## 2:                             0                  0           39</span></span>
<span><span class="co">## 3:                             0                  0           71</span></span>
<span><span class="co">## 4:                             0                  0           20</span></span>
<span><span class="co">## 5:                             0                  0           28</span></span>
<span><span class="co">## 6:                             0                  0           53</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="path-to-full-summary-stats-file">Path to full summary stats file<a class="anchor" aria-label="anchor" href="#path-to-full-summary-stats-file"></a>
</h3>
<ul>
<li><p>Since a full GWAS summary stats file would be too large to
include within <em>echolocatoR</em>,<br>
we instead provide an example subset of the full summary stats.</p></li>
<li><p>To simulate how you’d actually use your own full summary stats
file, we will save our example dataset to your computer (you can change
the path to wherever you like).</p></li>
<li><p>We highly recommend munging your full summary stats using the
Bioconductor package <a href="https://github.com/neurogenomics/MungeSumstats" class="external-link"><code>MungeSumstats</code></a>
first. It’s easy to use and very robust. It also means you don’t have to
provide most column mapping arguments in <code>finemap_loci</code> when
<code>munged=TRUE</code>.</p></li>
</ul>
<p>Here’s an example of how to munge your full summary stats file:</p>
<pre><code><span><span class="va">fullSS_path</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/example_fullSS.html" class="external-link">example_fullSS</a></span><span class="op">(</span>munged <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">fullSS_path</span> <span class="op">&lt;-</span> <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MungeSumstats/man/format_sumstats.html" class="external-link">format_sumstats</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">fullSS_path</span>, ref_genome <span class="op">=</span> <span class="st">"GRCH37"</span><span class="op">)</span></span></code></pre>
<p>We have already munged the following example summary stats for
you.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fullSS_path</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/example_fullSS.html" class="external-link">example_fullSS</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"Nalls2019"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Writing file to ==&gt; /tmp/RtmpOPlFzv/nalls2019.fullSS_subset.tsv</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-fine-mapping-pipeline">Run fine-mapping pipeline<a class="anchor" aria-label="anchor" href="#run-fine-mapping-pipeline"></a>
</h3>
<p>For a full description of all arguments, see
<code><a href="../reference/finemap_loci.html">?finemap_loci</a></code>.</p>
<p>Here are some key arguments:</p>
<ul>
<li>
<em>results_dir</em>: Where you want to store all of your
results.<br>
</li>
<li>
<em>finemap_methods</em>: Which fine-mapping methods you want to
run. For a full list of currently supported methods, run the function
<code><a href="https://rdrr.io/pkg/echofinemap/man/lfm.html" class="external-link">echofinemap::lfm()</a></code>.<br>
</li>
<li>
<em>bp_distance</em>: Controls window size. Specifically,
<code>bp_distance</code> is the number of basepairs upstream/downstream
you want to extract for each locus. For example, if you want a 2Mb
window (+/- 1Mb from the lead/index SNP in <code>top_SNPs</code>), set
<code>bp_distance=1e+06</code>.<br>
</li>
<li>
<em>plot_zoom</em>: Zoom in/out from the center of each locus when
producing the multiview plot.<br>
You can adjust this separately from <code>bp_distance</code> so that you
don’t have rerun the whole pipeline each time (locus subsets, LD
matrices, and fine-mapping results are all automatically saved in
locus-specific folders).</li>
</ul>
<p><strong>Note</strong>: Please use the full absolute paths (instead of
relative paths) wherever possible (e.g. <code>results_dir</code>). This
is especially important for the tool <em>FINEMAP</em>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">echolocatoR</span><span class="fu">::</span><span class="fu"><a href="../reference/finemap_loci.html">finemap_loci</a></span><span class="op">(</span></span>
<span> fullSS_path <span class="op">=</span> <span class="va">fullSS_path</span>,</span>
<span> topSNPs <span class="op">=</span> <span class="va">topSNPs</span>,</span>
<span> loci <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BST1"</span>,<span class="st">"MEX3C"</span><span class="op">)</span>,</span>
<span> LD_reference <span class="op">=</span> <span class="st">"1KGphase3"</span>,</span>
<span> dataset_name <span class="op">=</span> <span class="st">"Nalls23andMe_2019"</span>,</span>
<span> fullSS_genome_build <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span> bp_distance <span class="op">=</span> <span class="fl">1000</span>,</span>
<span> finemap_methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ABF"</span>,<span class="st">"SUSIE"</span>,<span class="st">"FINEMAP"</span><span class="op">)</span>,</span>
<span> munged <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "+ Assigning Gene and Locus independently."</span></span></code></pre>
<pre><code><span><span class="co">## Standardising column headers.</span></span></code></pre>
<pre><code><span><span class="co">## First line of summary statistics file:</span></span></code></pre>
<pre><code><span><span class="co">## SNP  CHR POS Locus   Gene    A2  A1  Effect allele frequency Effect  SE, all studies P   P, COJO, all studies    P, random effects, all studies  P, Conditional 23AndMe only P, 23AndMe only I2, all studies Freq1, previous studies Beta, previous studies  StdErr, previous studies    P, previous studies I2, previous studies    Freq1, new studies  Beta, new studies   StdErr, new studies P, new studies  I2, new studies Passes pooled 23andMe QC    Known GWAS locus within 1MB Failed final filtering and QC   Locus within 250KB  Locus Number    </span></span></code></pre>
<pre><code><span><span class="co">## Returning unmapped column names without making them uppercase.</span></span></code></pre>
<pre><code><span><span class="co">## + Mapping colnames from MungeSumstats ==&gt; echolocatoR</span></span></code></pre>
<pre><code><span><span class="co">## ┌────────────────────────────────────────┐</span></span>
<span><span class="co">## │                                        │</span></span>
<span><span class="co">## │   <span style="color: #FFFFFF; font-weight: 100;">)))&gt;</span><span style="color: #55FFFF;"> 🦇 BST1 [locus 1 / 2] 🦇 </span><span style="color: #FFFFFF; font-weight: 100;">&lt;(((</span>   │</span></span>
<span><span class="co">## │                                        │</span></span>
<span><span class="co">## └────────────────────────────────────────┘</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 1 ▶▶▶ Query 🔎</span> <span style="color: #00BBBB;">─────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## + Query Method: tabix</span></span></code></pre>
<pre><code><span><span class="co">## Constructing GRanges query using min/max ranges within a single chromosome.</span></span></code></pre>
<pre><code><span><span class="co">## query_dat is already a GRanges object. Returning directly.</span></span></code></pre>
<pre><code><span><span class="co">## ========= echotabix::convert =========</span></span></code></pre>
<pre><code><span><span class="co">## Converting full summary stats file to tabix format for fast querying.</span></span></code></pre>
<pre><code><span><span class="co">## Inferred format: 'table'</span></span></code></pre>
<pre><code><span><span class="co">## Explicit format: 'table'</span></span></code></pre>
<pre><code><span><span class="co">## Inferring comment_char from tabular header: 'SNP'</span></span></code></pre>
<pre><code><span><span class="co">## Determining chrom type from file header.</span></span></code></pre>
<pre><code><span><span class="co">## Assuming fullSS_path summary stats have already been processed with MungeSumstats.</span></span></code></pre>
<pre><code><span><span class="co">## Chromosome format: 1</span></span></code></pre>
<pre><code><span><span class="co">## Detecting column delimiter.</span></span></code></pre>
<pre><code><span><span class="co">## Identified column separator: \t</span></span></code></pre>
<pre><code><span><span class="co">## Sorting rows by coordinates via bash.</span></span></code></pre>
<pre><code><span><span class="co">## Searching for header row with grep.</span></span></code></pre>
<pre><code><span><span class="co">## ( grep ^'SNP' .../nalls2019.fullSS_subset.tsv; grep</span></span>
<span><span class="co">##     -v ^'SNP' .../nalls2019.fullSS_subset.tsv | sort</span></span>
<span><span class="co">##     -k2,2n</span></span>
<span><span class="co">##     -k3,3n ) &gt; .../file212e52bc52e8_sorted.tsv</span></span></code></pre>
<pre><code><span><span class="co">## Constructing outputs</span></span></code></pre>
<pre><code><span><span class="co">## bgzipping file with Rsamtools.</span></span></code></pre>
<pre><code><span><span class="co">## Reading bgzipped file using: data.table</span></span></code></pre>
<pre><code><span><span class="co">## Header preview:</span></span></code></pre>
<pre><code><span><span class="co">##           SNP CHR       BP A1 A2   FREQ    BETA     SE      P N_CAS   N_CON</span></span>
<span><span class="co">## 1:   rs741214   4 13737637  C  A 0.2341  0.0091 0.0119 0.4460 49053 1411006</span></span>
<span><span class="co">## 2: rs79385021   4 13737659  T  C 0.9109  0.0009 0.0198 0.9642 30435  974587</span></span>
<span><span class="co">## 3:   rs741215   4 13737767  C  T 0.0733  0.0190 0.0194 0.3284 49053 1411006</span></span>
<span><span class="co">## 4:   rs759261   4 13737915  G  A 0.4511  0.0075 0.0100 0.4543 49053 1411006</span></span>
<span><span class="co">## 5: rs35006360   4 13738014  G  A 0.1148 -0.0195 0.0160 0.2208 49053 1411006</span></span></code></pre>
<pre><code><span><span class="co">## Tabix-indexing file using: Rsamtools</span></span></code></pre>
<pre><code><span><span class="co">## Data successfully converted to bgzip-compressed, tabix-indexed format.</span></span></code></pre>
<pre><code><span><span class="co">## ========= echotabix::query =========</span></span></code></pre>
<pre><code><span><span class="co">## query_dat is already a GRanges object. Returning directly.</span></span></code></pre>
<pre><code><span><span class="co">## Inferred format: 'table'</span></span></code></pre>
<pre><code><span><span class="co">## Querying tabular tabix file using: Rsamtools.</span></span></code></pre>
<pre><code><span><span class="co">## Checking query chromosome style is correct.</span></span></code></pre>
<pre><code><span><span class="co">## Chromosome format: 1</span></span></code></pre>
<pre><code><span><span class="co">## Retrieving data.</span></span></code></pre>
<pre><code><span><span class="co">## Converting query results to data.table.</span></span></code></pre>
<pre><code><span><span class="co">## Processing query: 4:15736348-15738348</span></span></code></pre>
<pre><code><span><span class="co">## Adding 'query' column to results.</span></span></code></pre>
<pre><code><span><span class="co">## Retrieved data with 17 rows</span></span></code></pre>
<pre><code><span><span class="co">## Saving query ==&gt; /tmp/RtmpOPlFzv/results/GWAS/Nalls23andMe_2019/BST1/BST1_Nalls23andMe_2019_subset.tsv.gz</span></span></code></pre>
<pre><code><span><span class="co">## + Query: 17 SNPs x 11 columns.</span></span></code></pre>
<pre><code><span><span class="co">## Standardizing summary statistics subset.</span></span></code></pre>
<pre><code><span><span class="co">## Standardizing main column names.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Preparing A1,A1 cols</span></span></code></pre>
<pre><code><span><span class="co">## ++ Preparing MAF,Freq cols.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Could not infer MAF from Freq.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Preparing N_cases,N_controls cols.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Preparing proportion_cases col.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Calculating proportion_cases from N_cases and N_controls.</span></span></code></pre>
<pre><code><span><span class="co">## Preparing sample size column (N).</span></span></code></pre>
<pre><code><span><span class="co">## Computing effective sample size using the LDSC method:</span></span>
<span><span class="co">##  Neff = (N_CAS+N_CON) * (N_CAS/(N_CAS+N_CON)) / mean((N_CAS/(N_CAS+N_CON))[(N_CAS+N_CON)==max(N_CAS+N_CON)]))</span></span></code></pre>
<pre><code><span><span class="co">## + Mapping colnames from MungeSumstats ==&gt; echolocatoR</span></span></code></pre>
<pre><code><span><span class="co">## + Imputing t-statistic from Effect and StdErr.</span></span></code></pre>
<pre><code><span><span class="co">## + leadSNP missing. Assigning new one by min p-value.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Ensuring Effect,StdErr,P are numeric.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Ensuring 1 SNP per row and per genomic coordinate.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Removing extra whitespace</span></span></code></pre>
<pre><code><span><span class="co">## + Standardized query: 17 SNPs x 15 columns.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Saving standardized query ==&gt; /tmp/RtmpOPlFzv/results/GWAS/Nalls23andMe_2019/BST1/BST1_Nalls23andMe_2019_subset.tsv.gz</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 2 ▶▶▶ Extract Linkage Disequilibrium 🔗</span> <span style="color: #00BBBB;">────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## LD_reference identified as: 1kg.</span></span></code></pre>
<pre><code><span><span class="co">## Using 1000Genomes as LD reference panel.</span></span></code></pre>
<pre><code><span><span class="co">## Constructing GRanges query using min/max ranges across one or more chromosomes.</span></span></code></pre>
<pre><code><span><span class="co">## + as_blocks=TRUE: Will query a single range per chromosome that covers all regions requested (plus anything in between).</span></span></code></pre>
<pre><code><span><span class="co">## LD Reference Panel = 1KGphase3</span></span></code></pre>
<pre><code><span><span class="co">## Querying 1KG remote server.</span></span></code></pre>
<pre><code><span><span class="co">## Selecting 503 EUR individuals from 1kgphase3.</span></span></code></pre>
<pre><code><span><span class="co">## ========= echotabix::query =========</span></span></code></pre>
<pre><code><span><span class="co">## query_dat is already a GRanges object. Returning directly.</span></span></code></pre>
<pre><code><span><span class="co">## Explicit format: 'vcf'</span></span></code></pre>
<pre><code><span><span class="co">## Querying VCF tabix file.</span></span></code></pre>
<pre><code><span><span class="co">## Querying VCF file using: VariantAnnotation</span></span></code></pre>
<pre><code><span><span class="co">## Checking query chromosome style is correct.</span></span></code></pre>
<pre><code><span><span class="co">## Chromosome format: 1</span></span></code></pre>
<pre><code><span><span class="co">## Filtering query to 503 samples and returning ScanVcfParam object.</span></span></code></pre>
<pre><code><span><span class="co">## Retrieving data.</span></span></code></pre>
<pre><code><span><span class="co">## Time difference of 1.9 secs</span></span></code></pre>
<pre><code><span><span class="co">## Removing 17 / 66 non-overlapping SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## Saving VCF subset ==&gt; /tmp/RtmpOPlFzv/VCF/RtmpOPlFzv.chr4-15736493-15738253.ALL.chr4.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.bgz</span></span></code></pre>
<pre><code><span><span class="co">## Warning in .Call(.make_vcf_geno, filename, fixed, names(geno), as.list(geno), :</span></span>
<span><span class="co">## converting NULL pointer to R NULL</span></span></code></pre>
<pre><code><span><span class="co">## Time difference of 0.3 secs</span></span></code></pre>
<pre><code><span><span class="co">## Retrieved data with 17 rows across 503 samples.</span></span></code></pre>
<pre><code><span><span class="co">## echoLD::snpStats:: Filling `MAF` column with MAF from LD panel.</span></span></code></pre>
<pre><code><span><span class="co">## echoLD:snpStats:: Computing pairwise LD between 17 SNPs across 503 individuals (stats = R).</span></span></code></pre>
<pre><code><span><span class="co">## Time difference of 0 secs</span></span></code></pre>
<pre><code><span><span class="co">## 17 x 17 LD_matrix (sparse)</span></span></code></pre>
<pre><code><span><span class="co">## Converting obj to sparseMatrix.</span></span></code></pre>
<pre><code><span><span class="co">## Saving sparse LD matrix ==&gt; /tmp/RtmpOPlFzv/results/GWAS/Nalls23andMe_2019/BST1/LD/BST1.1KGphase3_LD.RDS</span></span></code></pre>
<pre><code><span><span class="co">## Removing 1 temp files.</span></span></code></pre>
<pre><code><span><span class="co">## + FILTER:: Filtering by LD features.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 3 ▶▶▶ Filter SNPs 🚰</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## FILTER:: Filtering by SNP features.</span></span></code></pre>
<pre><code><span><span class="co">## + FILTER:: Post-filtered data: 17 x 16</span></span></code></pre>
<pre><code><span><span class="co">## + Subsetting LD matrix and dat to common SNPs...</span></span></code></pre>
<pre><code><span><span class="co">## Removing unnamed rows/cols</span></span></code></pre>
<pre><code><span><span class="co">## Replacing NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## + LD_matrix = 17 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + dat = 17 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + 17 SNPs in common.</span></span></code></pre>
<pre><code><span><span class="co">## Converting obj to sparseMatrix.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 4 ▶▶▶ Fine-map 🔊</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## Gathering method sources.</span></span></code></pre>
<pre><code><span><span class="co">## Gathering method citations.</span></span></code></pre>
<pre><code><span><span class="co">## Preparing sample size column (N).</span></span></code></pre>
<pre><code><span><span class="co">## Using existing 'N' column.</span></span></code></pre>
<pre><code><span><span class="co">## Gathering method sources.</span></span></code></pre>
<pre><code><span><span class="co">## Gathering method citations.</span></span></code></pre>
<pre><code><span><span class="co">## Gathering method sources.</span></span></code></pre>
<pre><code><span><span class="co">## Gathering method citations.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="background-color: #FF55FF;">ABF</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All required columns present.</span></span></span>
<span><span class="co">## <span style="background-color: #FF55FF;">SUSIE</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All required columns present.</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All optional columns present.</span></span></span>
<span><span class="co">## <span style="background-color: #FF55FF;">FINEMAP</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All required columns present.</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All optional columns present.</span></span></span></code></pre>
<pre><code><span><span class="co">## ++ Fine-mapping using 3 tool(s): ABF, SUSIE, FINEMAP</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## +++ Multi-finemap:: ABF +++</span></span></code></pre>
<pre><code><span><span class="co">## Using all default values for finemap_args$ABF</span></span></code></pre>
<pre><code><span><span class="co">## Preparing sample size column (N).</span></span></code></pre>
<pre><code><span><span class="co">## Using existing 'N' column.</span></span></code></pre>
<pre><code><span><span class="co">## Running ABF.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Credible Set SNPs identified = 0</span></span></code></pre>
<pre><code><span><span class="co">## ++ Merging ABF results with multi-finemap data.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## +++ Multi-finemap:: SUSIE +++</span></span></code></pre>
<pre><code><span><span class="co">## Using all default values for finemap_args$SUSIE</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: Rfast</span></span></code></pre>
<pre><code><span><span class="co">## Preparing sample size column (N).</span></span></code></pre>
<pre><code><span><span class="co">## Using existing 'N' column.</span></span></code></pre>
<pre><code><span><span class="co">## + SUSIE:: sample_size=1,474,097</span></span></code></pre>
<pre><code><span><span class="co">## + Subsetting LD matrix and dat to common SNPs...</span></span></code></pre>
<pre><code><span><span class="co">## Removing unnamed rows/cols</span></span></code></pre>
<pre><code><span><span class="co">## Replacing NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## + LD_matrix = 17 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + dat = 17 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + 17 SNPs in common.</span></span></code></pre>
<pre><code><span><span class="co">## Converting obj to sparseMatrix.</span></span></code></pre>
<pre><code><span><span class="co">## + SUSIE:: Using `susie_rss()` from susieR v0.12.27</span></span></code></pre>
<pre><code><span><span class="co">## + SUSIE:: Extracting Credible Sets.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Credible Set SNPs identified = 0</span></span></code></pre>
<pre><code><span><span class="co">## ++ Merging SUSIE results with multi-finemap data.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## +++ Multi-finemap:: FINEMAP +++</span></span></code></pre>
<pre><code><span><span class="co">## Using all default values for finemap_args$FINEMAP</span></span></code></pre>
<pre><code><span><span class="co">## Preparing sample size column (N).</span></span></code></pre>
<pre><code><span><span class="co">## Using existing 'N' column.</span></span></code></pre>
<pre><code><span><span class="co">## + Subsetting LD matrix and dat to common SNPs...</span></span></code></pre>
<pre><code><span><span class="co">## Removing unnamed rows/cols</span></span></code></pre>
<pre><code><span><span class="co">## Replacing NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## + LD_matrix = 17 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + dat = 17 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + 17 SNPs in common.</span></span></code></pre>
<pre><code><span><span class="co">## Converting obj to sparseMatrix.</span></span></code></pre>
<pre><code><span><span class="co">## Constructing master file.</span></span></code></pre>
<pre><code><span><span class="co">## Constructing data.z file.</span></span></code></pre>
<pre><code><span><span class="co">## Constructing data.ld file.</span></span></code></pre>
<pre><code><span><span class="co">## FINEMAP path: /github/home/.cache/R/echofinemap/FINEMAP/finemap_v1.4.1_x86_64/finemap_v1.4.1_x86_64</span></span></code></pre>
<pre><code><span><span class="co">## Inferred FINEMAP version: 1.4.1</span></span></code></pre>
<pre><code><span><span class="co">## Running FINEMAP.</span></span></code></pre>
<pre><code>## cd .../BST1 &amp;&amp;
##     .../finemap_v1.4.1_x86_64
##    
##     --sss
##    
##     --in-files .../master
##    
##     --log
##    
##     --n-causal-snps 5
## 
## |--------------------------------------|
## | Welcome to FINEMAP v1.4.1            |
## |                                      |
## | (c) 2015-2022 University of Helsinki |
## |                                      |
## | Help :                               |
## | - ./finemap --help                   |
## | - www.finemap.me                     |
## | - www.christianbenner.com            |
## |                                      |
## | Contact :                            |
## | - finemap@christianbenner.com        |
## | - matti.pirinen@helsinki.fi          |
## |--------------------------------------|
## 
## --------
## SETTINGS
## --------
## - dataset            : all
## - corr-config        : 0.95
## - n-causal-snps      : 5
## - n-configs-top      : 50000
## - n-conv-sss         : 100
## - n-iter             : 100000
## - n-threads          : 1
## - prior-k0           : 0
## - prior-std          : 0.05 
## - prob-conv-sss-tol  : 0.001
## - prob-cred-set      : 0.95
## 
## ------------
## FINE-MAPPING (1/1)
## ------------
## - GWAS summary stats               : FINEMAP/data.z
## - SNP correlations                 : FINEMAP/data.ld
## - Causal SNP stats                 : FINEMAP/data.snp
## - Causal configurations            : FINEMAP/data.config
## - Credible sets                    : FINEMAP/data.cred
## - Log file                         : FINEMAP/data.log_sss
## 
- Reading summary statistics       : +
- Reading summary statistics       : -
- Reading summary statistics       : +
- Reading summary stats            : done!   
- Reading SNP correlations         : 0.000%
- Reading SNP correlations         : done!   
- Reading input                    : done!   
## 
## - Number of GWAS samples           : 1474097
## - Number of SNPs                   : 17
## - Prior-Pr(# of causal SNPs is k)  : 
##   (0 -&gt; 0)
##    1 -&gt; 0.59
##    2 -&gt; 0.295
##    3 -&gt; 0.0921
##    4 -&gt; 0.0202
##    5 -&gt; 0.00327
## 0 configurations evaluated (0.000/100%)
- 17 configurations evaluated (0.000/100%)
- 32 configurations evaluated (0.000/100%)
- 371 configurations evaluated (0.050/100%)
- 452 configurations evaluated (0.100/100%)
- 532 configurations evaluated (0.150/100%)
- 585 configurations evaluated (0.200/100%)
- 637 configurations evaluated (0.250/100%)
- 637 configurations evaluated (0.300/100%)
- 637 configurations evaluated (0.339/100%) : converged after 339 iterations
## 
- Computing causal SNP statistics  : 0.157%
- Computing causal SNP statistics  : 8.006%
- Computing causal SNP statistics  : 15.856%
- Computing causal SNP statistics  : 23.705%
- Computing causal SNP statistics  : 31.554%
- Computing causal SNP statistics  : 39.403%
- Computing causal SNP statistics  : 47.253%
- Computing causal SNP statistics  : 55.102%
- Computing causal SNP statistics  : 62.951%
- Computing causal SNP statistics  : 70.801%
- Computing causal SNP statistics  : 78.650%
- Computing causal SNP statistics  : 86.499%
- Computing causal SNP statistics  : 94.349%
- Computing causal SNP statistics  : done!   
## - Regional SNP heritability        : 9.7e-05 (SD: 1.55e-05 ; 95% CI: [6.82e-05,0.00013])
## - Log10-BF of &gt;= one causal SNP    : 62.1
## - Post-expected # of causal SNPs   : 5
## - Post-Pr(# of causal SNPs is k)   : 
##   (0 -&gt; 0)
##    1 -&gt; 2.28e-39
##    2 -&gt; 7.42e-38
##    3 -&gt; 3.98e-38
##    4 -&gt; 6.59e-38
##    5 -&gt; 1
## - Computing credible sets          : 
- Writing causal configurations    : 0.157%
- Writing causal configurations    : 8.006%
- Writing causal configurations    : 15.856%
- Writing causal configurations    : 23.705%
- Writing causal configurations    : 31.554%
- Writing causal configurations    : 39.403%
- Writing causal configurations    : 47.253%
- Writing causal configurations    : 55.102%
- Writing causal configurations    : 62.951%
- Writing causal configurations    : 70.801%
- Writing causal configurations    : 78.650%
- Writing causal configurations    : 86.499%
- Writing causal configurations    : 94.349%
- Writing causal SNP statistics    : 5.882%
- Writing causal SNP statistics    : 64.706%
- Writing output                   : done!   
## - Run time                         : 0 hours, 0 minutes, 0 seconds</code></pre>
<pre><code><span><span class="co">## 1 data.cred* file(s) found in the same subfolder.</span></span></code></pre>
<pre><code><span><span class="co">## Selected file based on postPr_k: data.cred5</span></span></code></pre>
<pre><code><span><span class="co">## Importing conditional probabilities (.cred file).</span></span></code></pre>
<pre><code><span><span class="co">## Importing marginal probabilities (.snp file).</span></span></code></pre>
<pre><code><span><span class="co">## Importing configuration probabilities (.config file).</span></span></code></pre>
<pre><code><span><span class="co">## ++ Credible Set SNPs identified = 12</span></span></code></pre>
<pre><code><span><span class="co">## ++ Merging FINEMAP results with multi-finemap data.</span></span></code></pre>
<pre><code><span><span class="co">## Identifying Consensus SNPs...</span></span></code></pre>
<pre><code><span><span class="co">## + support_thresh = 2</span></span></code></pre>
<pre><code><span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span></code></pre>
<pre><code><span><span class="co">## + 3 fine-mapping methods used.</span></span></code></pre>
<pre><code><span><span class="co">## + 12 Credible Set SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + 0 Consensus SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## Saving merged finemapping results ==&gt; /tmp/RtmpOPlFzv/results/GWAS/Nalls23andMe_2019/BST1/Multi-finemap/1KGphase3_LD.Multi-finemap.tsv.gz</span></span></code></pre>
<pre><code><span><span class="co">## + Fine-mapping with 'ABF, SUSIE, FINEMAP' completed:</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 5 ▶▶▶ Plot 📈</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## +-------- Locus Plot:  BST1 --------+</span></span></code></pre>
<pre><code><span><span class="co">## + support_thresh = 2</span></span></code></pre>
<pre><code><span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span></code></pre>
<pre><code><span><span class="co">## + 3 fine-mapping methods used.</span></span></code></pre>
<pre><code><span><span class="co">## + 12 Credible Set SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + 0 Consensus SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in CS cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in PP cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## LD_matrix detected. Coloring SNPs by LD with lead SNP.</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: GWAS full window track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: GWAS track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Merged fine-mapping track</span></span></code></pre>
<pre><code><span><span class="co">## Melting PP and CS from 4 fine-mapping methods.</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Adding Gene model track.</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: EnsDb.Hsapiens.v75</span></span></code></pre>
<pre><code><span><span class="co">## max_transcripts= 1 .</span></span></code></pre>
<pre><code><span><span class="co">## 1  transcripts from  1  genes returned.</span></span></code></pre>
<pre><code><span><span class="co">## Fetching data...OK</span></span>
<span><span class="co">## Parsing exons...OK</span></span>
<span><span class="co">## Defining introns...OK</span></span>
<span><span class="co">## Defining UTRs...OK</span></span>
<span><span class="co">## Defining CDS...OK</span></span>
<span><span class="co">## aggregating...</span></span>
<span><span class="co">## Done</span></span>
<span><span class="co">## Constructing graphics...</span></span>
<span><span class="co">## + Adding vertical lines to highlight SNP groups...</span></span>
<span><span class="co">## +&gt;+&gt;+&gt;+&gt;+ zoom =  1x  +&lt;+&lt;+&lt;+&lt;+</span></span>
<span><span class="co">## + echoplot:: Get window suffix...</span></span>
<span><span class="co">## + echoplot:: Removing GWAS full window track @ zoom=1x</span></span>
<span><span class="co">## + Removing subplot margins...</span></span>
<span><span class="co">## + Reordering tracks...</span></span></code></pre>
<pre><code><span><span class="co">## [1] "+ Ensuring last track shows genomic units..."</span></span></code></pre>
<pre><code><span><span class="co">## + Aligning xlimits for each subplot...</span></span>
<span><span class="co">## + Checking track heights...</span></span>
<span><span class="co">## + echoplot:: Saving plot ==&gt; /tmp/RtmpOPlFzv/results/GWAS/Nalls23andMe_2019/BST1/multiview.BST1.1KGphase3.1x.png</span></span>
<span><span class="co">## Found more than one class "simpleUnit" in cache; using the first, from namespace 'hexbin'</span></span>
<span><span class="co">## Also defined by 'ggbio'</span></span>
<span><span class="co">## Recording all `finemap_locus` arguments.</span></span>
<span><span class="co">## Formatting locus results.</span></span></code></pre>
<pre><code><span><span class="co">## ┌─────────────────────────────────────────┐</span></span>
<span><span class="co">## │                                         │</span></span>
<span><span class="co">## │   <span style="color: #FFFFFF; font-weight: 100;">)))&gt;</span><span style="color: #55FFFF;"> 🦇 MEX3C [locus 2 / 2] 🦇 </span><span style="color: #FFFFFF; font-weight: 100;">&lt;(((</span>   │</span></span>
<span><span class="co">## │                                         │</span></span>
<span><span class="co">## └─────────────────────────────────────────┘</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 1 ▶▶▶ Query 🔎</span> <span style="color: #00BBBB;">─────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## + Query Method: tabix</span></span>
<span><span class="co">## Constructing GRanges query using min/max ranges within a single chromosome.</span></span>
<span><span class="co">## query_dat is already a GRanges object. Returning directly.</span></span>
<span><span class="co">## ========= echotabix::convert =========</span></span>
<span><span class="co">## Converting full summary stats file to tabix format for fast querying.</span></span>
<span><span class="co">## Inferred format: 'table'</span></span>
<span><span class="co">## Explicit format: 'table'</span></span>
<span><span class="co">## Inferring comment_char from tabular header: 'SNP'</span></span>
<span><span class="co">## Determining chrom type from file header.</span></span>
<span><span class="co">## Chromosome format: 1</span></span>
<span><span class="co">## Detecting column delimiter.</span></span>
<span><span class="co">## Identified column separator: \t</span></span>
<span><span class="co">## Sorting rows by coordinates via bash.</span></span>
<span><span class="co">## Searching for header row with grep.</span></span></code></pre>
<pre><code><span><span class="co">## ( grep ^'SNP' .../nalls2019.fullSS_subset.tsv; grep</span></span>
<span><span class="co">##     -v ^'SNP' .../nalls2019.fullSS_subset.tsv | sort</span></span>
<span><span class="co">##     -k2,2n</span></span>
<span><span class="co">##     -k3,3n ) &gt; .../file212e3bd3e0f2_sorted.tsv</span></span></code></pre>
<pre><code><span><span class="co">## Constructing outputs</span></span>
<span><span class="co">## Using existing bgzipped file: /tmp/RtmpOPlFzv/nalls2019.fullSS_subset.tsv.bgz </span></span>
<span><span class="co">## Set force_new=TRUE to override this.</span></span>
<span><span class="co">## Tabix-indexing file using: Rsamtools</span></span>
<span><span class="co">## Data successfully converted to bgzip-compressed, tabix-indexed format.</span></span>
<span><span class="co">## ========= echotabix::query =========</span></span>
<span><span class="co">## query_dat is already a GRanges object. Returning directly.</span></span>
<span><span class="co">## Inferred format: 'table'</span></span>
<span><span class="co">## Querying tabular tabix file using: Rsamtools.</span></span>
<span><span class="co">## Checking query chromosome style is correct.</span></span>
<span><span class="co">## Chromosome format: 1</span></span>
<span><span class="co">## Retrieving data.</span></span>
<span><span class="co">## Converting query results to data.table.</span></span>
<span><span class="co">## Processing query: 18:48682589-48684589</span></span>
<span><span class="co">## Adding 'query' column to results.</span></span>
<span><span class="co">## Retrieved data with 7 rows</span></span>
<span><span class="co">## Saving query ==&gt; /tmp/RtmpOPlFzv/results/GWAS/Nalls23andMe_2019/MEX3C/MEX3C_Nalls23andMe_2019_subset.tsv.gz</span></span>
<span><span class="co">## + Query: 7 SNPs x 11 columns.</span></span>
<span><span class="co">## Standardizing summary statistics subset.</span></span>
<span><span class="co">## Standardizing main column names.</span></span>
<span><span class="co">## ++ Preparing A1,A1 cols</span></span>
<span><span class="co">## ++ Preparing MAF,Freq cols.</span></span>
<span><span class="co">## ++ Could not infer MAF from Freq.</span></span>
<span><span class="co">## ++ Preparing N_cases,N_controls cols.</span></span>
<span><span class="co">## ++ Preparing proportion_cases col.</span></span>
<span><span class="co">## ++ Calculating proportion_cases from N_cases and N_controls.</span></span>
<span><span class="co">## Preparing sample size column (N).</span></span>
<span><span class="co">## Computing effective sample size using the LDSC method:</span></span>
<span><span class="co">##  Neff = (N_CAS+N_CON) * (N_CAS/(N_CAS+N_CON)) / mean((N_CAS/(N_CAS+N_CON))[(N_CAS+N_CON)==max(N_CAS+N_CON)]))</span></span>
<span><span class="co">## + Mapping colnames from MungeSumstats ==&gt; echolocatoR</span></span>
<span><span class="co">## + Imputing t-statistic from Effect and StdErr.</span></span>
<span><span class="co">## + leadSNP missing. Assigning new one by min p-value.</span></span>
<span><span class="co">## ++ Ensuring Effect,StdErr,P are numeric.</span></span>
<span><span class="co">## ++ Ensuring 1 SNP per row and per genomic coordinate.</span></span>
<span><span class="co">## ++ Removing extra whitespace</span></span>
<span><span class="co">## + Standardized query: 7 SNPs x 15 columns.</span></span>
<span><span class="co">## ++ Saving standardized query ==&gt; /tmp/RtmpOPlFzv/results/GWAS/Nalls23andMe_2019/MEX3C/MEX3C_Nalls23andMe_2019_subset.tsv.gz</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 2 ▶▶▶ Extract Linkage Disequilibrium 🔗</span> <span style="color: #00BBBB;">────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## LD_reference identified as: 1kg.</span></span>
<span><span class="co">## Using 1000Genomes as LD reference panel.</span></span>
<span><span class="co">## Constructing GRanges query using min/max ranges across one or more chromosomes.</span></span>
<span><span class="co">## + as_blocks=TRUE: Will query a single range per chromosome that covers all regions requested (plus anything in between).</span></span>
<span><span class="co">## LD Reference Panel = 1KGphase3</span></span>
<span><span class="co">## Querying 1KG remote server.</span></span>
<span><span class="co">## Selecting 503 EUR individuals from 1kgphase3.</span></span>
<span><span class="co">## ========= echotabix::query =========</span></span>
<span><span class="co">## query_dat is already a GRanges object. Returning directly.</span></span>
<span><span class="co">## Explicit format: 'vcf'</span></span>
<span><span class="co">## Querying VCF tabix file.</span></span>
<span><span class="co">## Querying VCF file using: VariantAnnotation</span></span>
<span><span class="co">## Checking query chromosome style is correct.</span></span>
<span><span class="co">## Chromosome format: 1</span></span>
<span><span class="co">## Filtering query to 503 samples and returning ScanVcfParam object.</span></span>
<span><span class="co">## Retrieving data.</span></span>
<span><span class="co">## Time difference of 2.6 secs</span></span>
<span><span class="co">## Removing 7 / 38 non-overlapping SNPs.</span></span>
<span><span class="co">## Saving VCF subset ==&gt; /tmp/RtmpOPlFzv/VCF/RtmpOPlFzv.chr18-48682662-48684103.ALL.chr18.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.bgz</span></span></code></pre>
<pre><code><span><span class="co">## Warning in .Call(.make_vcf_geno, filename, fixed, names(geno), as.list(geno), :</span></span>
<span><span class="co">## converting NULL pointer to R NULL</span></span></code></pre>
<pre><code><span><span class="co">## Time difference of 0.2 secs</span></span>
<span><span class="co">## Retrieved data with 7 rows across 503 samples.</span></span>
<span><span class="co">## echoLD::snpStats:: Filling `MAF` column with MAF from LD panel.</span></span>
<span><span class="co">## echoLD:snpStats:: Computing pairwise LD between 7 SNPs across 503 individuals (stats = R).</span></span>
<span><span class="co">## Time difference of 0 secs</span></span>
<span><span class="co">## 7 x 7 LD_matrix (sparse)</span></span>
<span><span class="co">## Converting obj to sparseMatrix.</span></span>
<span><span class="co">## Saving sparse LD matrix ==&gt; /tmp/RtmpOPlFzv/results/GWAS/Nalls23andMe_2019/MEX3C/LD/MEX3C.1KGphase3_LD.RDS</span></span>
<span><span class="co">## Removing 1 temp files.</span></span>
<span><span class="co">## + FILTER:: Filtering by LD features.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 3 ▶▶▶ Filter SNPs 🚰</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## FILTER:: Filtering by SNP features.</span></span>
<span><span class="co">## + FILTER:: Post-filtered data: 5 x 16</span></span>
<span><span class="co">## + Subsetting LD matrix and dat to common SNPs...</span></span>
<span><span class="co">## Removing unnamed rows/cols</span></span>
<span><span class="co">## Replacing NAs with 0</span></span>
<span><span class="co">## + LD_matrix = 7 SNPs.</span></span>
<span><span class="co">## + dat = 5 SNPs.</span></span>
<span><span class="co">## + 5 SNPs in common.</span></span>
<span><span class="co">## Converting obj to sparseMatrix.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 4 ▶▶▶ Fine-map 🔊</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## Gathering method sources.</span></span>
<span><span class="co">## Gathering method citations.</span></span>
<span><span class="co">## Preparing sample size column (N).</span></span>
<span><span class="co">## Using existing 'N' column.</span></span>
<span><span class="co">## Gathering method sources.</span></span>
<span><span class="co">## Gathering method citations.</span></span>
<span><span class="co">## Gathering method sources.</span></span>
<span><span class="co">## Gathering method citations.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="background-color: #FF55FF;">ABF</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All required columns present.</span></span></span>
<span><span class="co">## <span style="background-color: #FF55FF;">SUSIE</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All required columns present.</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All optional columns present.</span></span></span>
<span><span class="co">## <span style="background-color: #FF55FF;">FINEMAP</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All required columns present.</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All optional columns present.</span></span></span></code></pre>
<pre><code><span><span class="co">## ++ Fine-mapping using 3 tool(s): ABF, SUSIE, FINEMAP</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## +++ Multi-finemap:: ABF +++</span></span>
<span><span class="co">## Using all default values for finemap_args$ABF</span></span>
<span><span class="co">## Preparing sample size column (N).</span></span>
<span><span class="co">## Using existing 'N' column.</span></span>
<span><span class="co">## Running ABF.</span></span>
<span><span class="co">## ++ Credible Set SNPs identified = 0</span></span>
<span><span class="co">## ++ Merging ABF results with multi-finemap data.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## +++ Multi-finemap:: SUSIE +++</span></span>
<span><span class="co">## Using all default values for finemap_args$SUSIE</span></span>
<span><span class="co">## Preparing sample size column (N).</span></span>
<span><span class="co">## Using existing 'N' column.</span></span>
<span><span class="co">## + SUSIE:: sample_size=1,460,059</span></span>
<span><span class="co">## + Subsetting LD matrix and dat to common SNPs...</span></span>
<span><span class="co">## Removing unnamed rows/cols</span></span>
<span><span class="co">## Replacing NAs with 0</span></span>
<span><span class="co">## + LD_matrix = 5 SNPs.</span></span>
<span><span class="co">## + dat = 5 SNPs.</span></span>
<span><span class="co">## + 5 SNPs in common.</span></span>
<span><span class="co">## Converting obj to sparseMatrix.</span></span>
<span><span class="co">## + SUSIE:: Using `susie_rss()` from susieR v0.12.27</span></span>
<span><span class="co">## + SUSIE:: Extracting Credible Sets.</span></span>
<span><span class="co">## ++ Credible Set SNPs identified = 0</span></span>
<span><span class="co">## ++ Merging SUSIE results with multi-finemap data.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## +++ Multi-finemap:: FINEMAP +++</span></span>
<span><span class="co">## Using all default values for finemap_args$FINEMAP</span></span>
<span><span class="co">## Preparing sample size column (N).</span></span>
<span><span class="co">## Using existing 'N' column.</span></span>
<span><span class="co">## + Subsetting LD matrix and dat to common SNPs...</span></span>
<span><span class="co">## Removing unnamed rows/cols</span></span>
<span><span class="co">## Replacing NAs with 0</span></span>
<span><span class="co">## + LD_matrix = 5 SNPs.</span></span>
<span><span class="co">## + dat = 5 SNPs.</span></span>
<span><span class="co">## + 5 SNPs in common.</span></span>
<span><span class="co">## Converting obj to sparseMatrix.</span></span>
<span><span class="co">## Constructing master file.</span></span>
<span><span class="co">## Constructing data.z file.</span></span>
<span><span class="co">## Constructing data.ld file.</span></span>
<span><span class="co">## FINEMAP path: /github/home/.cache/R/echofinemap/FINEMAP/finemap_v1.4.1_x86_64/finemap_v1.4.1_x86_64</span></span>
<span><span class="co">## Inferred FINEMAP version: 1.4.1</span></span>
<span><span class="co">## Running FINEMAP.</span></span></code></pre>
<pre><code>## cd .../MEX3C &amp;&amp;
##     .../finemap_v1.4.1_x86_64
##    
##     --sss
##    
##     --in-files .../master
##    
##     --log
##    
##     --n-causal-snps 5
## 
## |--------------------------------------|
## | Welcome to FINEMAP v1.4.1            |
## |                                      |
## | (c) 2015-2022 University of Helsinki |
## |                                      |
## | Help :                               |
## | - ./finemap --help                   |
## | - www.finemap.me                     |
## | - www.christianbenner.com            |
## |                                      |
## | Contact :                            |
## | - finemap@christianbenner.com        |
## | - matti.pirinen@helsinki.fi          |
## |--------------------------------------|
## 
## --------
## SETTINGS
## --------
## - dataset            : all
## - corr-config        : 0.95
## - n-causal-snps      : 5
## - n-configs-top      : 50000
## - n-conv-sss         : 100
## - n-iter             : 100000
## - n-threads          : 1
## - prior-k0           : 0
## - prior-std          : 0.05 
## - prob-conv-sss-tol  : 0.001
## - prob-cred-set      : 0.95
## 
## ------------
## FINE-MAPPING (1/1)
## ------------
## - GWAS summary stats               : FINEMAP/data.z
## - SNP correlations                 : FINEMAP/data.ld
## - Causal SNP stats                 : FINEMAP/data.snp
## - Causal configurations            : FINEMAP/data.config
## - Credible sets                    : FINEMAP/data.cred
## - Log file                         : FINEMAP/data.log_sss
## 
- Reading summary statistics       : +
- Reading summary stats            : done!   
- Reading SNP correlations         : 0.000%
- Reading SNP correlations         : done!   
- Reading input                    : done!   
## 
## - Number of GWAS samples           : 1460059
## - Number of SNPs                   : 5
## - Prior-Pr(# of causal SNPs is k)  : 
##   (0 -&gt; 0)
##    1 -&gt; 0.609
##    2 -&gt; 0.305
##    3 -&gt; 0.0762
##    4 -&gt; 0.00952
##    5 -&gt; 0.000476
## 0 configurations evaluated (0.000/100%)
- 5 configurations evaluated (0.000/100%)
- 7 configurations evaluated (0.000/100%)
- 13 configurations evaluated (0.050/100%)
- 13 configurations evaluated (0.100/100%)
- 13 configurations evaluated (0.103/100%) : converged after 103 iterations
## 
- Computing causal SNP statistics  : 7.692%
- Computing causal SNP statistics  : done!   
## - Regional SNP heritability        : 1.93e-05 (SD: 7.14e-06 ; 95% CI: [7.32e-06,3.47e-05])
## - Log10-BF of &gt;= one causal SNP    : 3.56
## - Post-expected # of causal SNPs   : 1.09
## - Post-Pr(# of causal SNPs is k)   : 
##   (0 -&gt; 0)
##    1 -&gt; 0.914
##    2 -&gt; 0.0853
##    3 -&gt; 0.000355
##    4 -&gt; 0
##    5 -&gt; 0
## - Computing credible sets          : 
- Writing causal configurations    : 7.692%
- Writing causal SNP statistics    : 20.000%
- Writing output                   : done!   
## - Run time                         : 0 hours, 0 minutes, 0 seconds</code></pre>
<pre><code><span><span class="co">## 2 data.cred* file(s) found in the same subfolder.</span></span>
<span><span class="co">## Selected file based on postPr_k: data.cred1</span></span>
<span><span class="co">## Importing conditional probabilities (.cred file).</span></span>
<span><span class="co">## No configurations were causal at PP&gt;=0.95.</span></span>
<span><span class="co">## Importing marginal probabilities (.snp file).</span></span>
<span><span class="co">## Importing configuration probabilities (.config file).</span></span>
<span><span class="co">## FINEMAP was unable to identify any credible sets at PP&gt;=0.95.</span></span>
<span><span class="co">## ++ Credible Set SNPs identified = 0</span></span>
<span><span class="co">## ++ Merging FINEMAP results with multi-finemap data.</span></span>
<span><span class="co">## Identifying Consensus SNPs...</span></span>
<span><span class="co">## + support_thresh = 2</span></span>
<span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span>
<span><span class="co">## + 3 fine-mapping methods used.</span></span>
<span><span class="co">## + 0 Credible Set SNPs identified.</span></span>
<span><span class="co">## + 0 Consensus SNPs identified.</span></span>
<span><span class="co">## Saving merged finemapping results ==&gt; /tmp/RtmpOPlFzv/results/GWAS/Nalls23andMe_2019/MEX3C/Multi-finemap/1KGphase3_LD.Multi-finemap.tsv.gz</span></span>
<span><span class="co">## + Fine-mapping with 'ABF, SUSIE, FINEMAP' completed:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 5 ▶▶▶ Plot 📈</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## +-------- Locus Plot:  MEX3C --------+</span></span>
<span><span class="co">## + support_thresh = 2</span></span>
<span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span>
<span><span class="co">## + 3 fine-mapping methods used.</span></span>
<span><span class="co">## + 0 Credible Set SNPs identified.</span></span>
<span><span class="co">## + 0 Consensus SNPs identified.</span></span>
<span><span class="co">## + Filling NAs in CS cols with 0.</span></span>
<span><span class="co">## + Filling NAs in PP cols with 0.</span></span>
<span><span class="co">## LD_matrix detected. Coloring SNPs by LD with lead SNP.</span></span>
<span><span class="co">## ++ echoplot:: GWAS full window track</span></span>
<span><span class="co">## ++ echoplot:: GWAS track</span></span>
<span><span class="co">## ++ echoplot:: Merged fine-mapping track</span></span>
<span><span class="co">## Melting PP and CS from 4 fine-mapping methods.</span></span>
<span><span class="co">## ++ echoplot:: Adding Gene model track.</span></span>
<span><span class="co">## Converting dat to GRanges object.</span></span>
<span><span class="co">## max_transcripts= 1 . </span></span>
<span><span class="co">## 0  transcripts from  0  genes returned.</span></span>
<span><span class="co">## Returning NULL for gene track.</span></span>
<span><span class="co">## + Adding vertical lines to highlight SNP groups...</span></span>
<span><span class="co">## +&gt;+&gt;+&gt;+&gt;+ zoom =  1x  +&lt;+&lt;+&lt;+&lt;+</span></span>
<span><span class="co">## + echoplot:: Get window suffix...</span></span>
<span><span class="co">## + echoplot:: Removing GWAS full window track @ zoom=1x</span></span>
<span><span class="co">## + Removing subplot margins...</span></span>
<span><span class="co">## + Reordering tracks...</span></span></code></pre>
<pre><code><span><span class="co">## [1] "+ Ensuring last track shows genomic units..."</span></span></code></pre>
<pre><code><span><span class="co">## + Aligning xlimits for each subplot...</span></span>
<span><span class="co">## + Checking track heights...</span></span>
<span><span class="co">## + echoplot:: Saving plot ==&gt; /tmp/RtmpOPlFzv/results/GWAS/Nalls23andMe_2019/MEX3C/multiview.MEX3C.1KGphase3.1x.png</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 1 rows containing missing values (geom_text).</span></span></code></pre>
<p><img src="echolocatoR_files/figure-html/Run%20fine-mapping%20pipeline-1.png" width="700"></p>
<pre><code><span><span class="co">## Warning: Removed 1 rows containing missing values (geom_text).</span></span></code></pre>
<pre><code><span><span class="co">## Recording all `finemap_locus` arguments.</span></span>
<span><span class="co">## Formatting locus results.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 6 ▶▶▶ Postprocess data 🎁</span> <span style="color: #00BBBB;">──────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## Returning results as nested list.</span></span></code></pre>
<p><img src="echolocatoR_files/figure-html/Run%20fine-mapping%20pipeline-2.png" width="700"></p>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">All loci done in: </span><span style="color: #FFFFFF;">0.3 min</span></span></span>
<span><span class="co"><span style="color: #55FFFF;">## </span></span></span></code></pre>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><div class="sourceCode" id="cb263"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre style="max-height: 200px;"><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.5 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] snpStats_1.47.1   Matrix_1.5-1      survival_3.4-0    echolocatoR_2.0.3</span></span>
<span><span class="co">## [5] BiocStyle_2.26.0 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rappdirs_0.3.3              rtracklayer_1.58.0         </span></span>
<span><span class="co">##   [3] GGally_2.1.2                R.methodsS3_1.8.2          </span></span>
<span><span class="co">##   [5] ragg_1.2.4                  tidyr_1.2.1                </span></span>
<span><span class="co">##   [7] echoLD_0.99.8               ggplot2_3.3.6              </span></span>
<span><span class="co">##   [9] bit64_4.0.5                 knitr_1.40                 </span></span>
<span><span class="co">##  [11] irlba_2.3.5.1               DelayedArray_0.24.0        </span></span>
<span><span class="co">##  [13] R.utils_2.12.1              data.table_1.14.4          </span></span>
<span><span class="co">##  [15] rpart_4.1.19                KEGGREST_1.38.0            </span></span>
<span><span class="co">##  [17] RCurl_1.98-1.9              AnnotationFilter_1.22.0    </span></span>
<span><span class="co">##  [19] generics_0.1.3              BiocGenerics_0.44.0        </span></span>
<span><span class="co">##  [21] GenomicFeatures_1.50.2      RSQLite_2.2.18             </span></span>
<span><span class="co">##  [23] proxy_0.4-27                bit_4.0.4                  </span></span>
<span><span class="co">##  [25] tzdb_0.3.0                  xml2_1.3.3                 </span></span>
<span><span class="co">##  [27] SummarizedExperiment_1.28.0 assertthat_0.2.1           </span></span>
<span><span class="co">##  [29] gargle_1.2.1                viridis_0.6.2              </span></span>
<span><span class="co">##  [31] xfun_0.34                   hms_1.1.2                  </span></span>
<span><span class="co">##  [33] jquerylib_0.1.4             evaluate_0.17              </span></span>
<span><span class="co">##  [35] fansi_1.0.3                 restfulr_0.0.15            </span></span>
<span><span class="co">##  [37] progress_1.2.2              dbplyr_2.2.1               </span></span>
<span><span class="co">##  [39] readxl_1.4.1                Rgraphviz_2.41.2           </span></span>
<span><span class="co">##  [41] igraph_1.3.5                DBI_1.1.3                  </span></span>
<span><span class="co">##  [43] htmlwidgets_1.5.4           reshape_0.8.9              </span></span>
<span><span class="co">##  [45] downloadR_0.99.5            stats4_4.2.1               </span></span>
<span><span class="co">##  [47] purrr_0.3.5                 ellipsis_0.3.2             </span></span>
<span><span class="co">##  [49] dplyr_1.0.10                backports_1.4.1            </span></span>
<span><span class="co">##  [51] bookdown_0.29               biomaRt_2.54.0             </span></span>
<span><span class="co">##  [53] deldir_1.0-6                MatrixGenerics_1.10.0      </span></span>
<span><span class="co">##  [55] MungeSumstats_1.6.0         vctrs_0.5.0                </span></span>
<span><span class="co">##  [57] Biobase_2.58.0              ensembldb_2.22.0           </span></span>
<span><span class="co">##  [59] withr_2.5.0                 cachem_1.0.6               </span></span>
<span><span class="co">##  [61] BSgenome_1.66.1             checkmate_2.1.0            </span></span>
<span><span class="co">##  [63] GenomicAlignments_1.34.0    prettyunits_1.1.1          </span></span>
<span><span class="co">##  [65] cluster_2.1.4               ape_5.6-2                  </span></span>
<span><span class="co">##  [67] dir.expiry_1.5.1            lazyeval_0.2.2             </span></span>
<span><span class="co">##  [69] crayon_1.5.2                basilisk.utils_1.9.4       </span></span>
<span><span class="co">##  [71] crul_1.3                    labeling_0.4.2             </span></span>
<span><span class="co">##  [73] pkgconfig_2.0.3             GenomeInfoDb_1.34.1        </span></span>
<span><span class="co">##  [75] nlme_3.1-160                ProtGenerics_1.30.0        </span></span>
<span><span class="co">##  [77] XGR_1.1.8                   nnet_7.3-18                </span></span>
<span><span class="co">##  [79] pals_1.7                    rlang_1.0.6                </span></span>
<span><span class="co">##  [81] lifecycle_1.0.3             filelock_1.0.2             </span></span>
<span><span class="co">##  [83] httpcode_0.3.0              BiocFileCache_2.6.0        </span></span>
<span><span class="co">##  [85] echotabix_0.99.8            dichromat_2.0-0.1          </span></span>
<span><span class="co">##  [87] cellranger_1.1.0            coloc_5.1.0.1              </span></span>
<span><span class="co">##  [89] rprojroot_2.0.3             matrixStats_0.62.0         </span></span>
<span><span class="co">##  [91] graph_1.76.0                osfr_0.2.9                 </span></span>
<span><span class="co">##  [93] boot_1.3-28                 base64enc_0.1-3            </span></span>
<span><span class="co">##  [95] png_0.1-7                   viridisLite_0.4.1          </span></span>
<span><span class="co">##  [97] rjson_0.2.21                rootSolve_1.8.2.3          </span></span>
<span><span class="co">##  [99] bitops_1.0-7                R.oo_1.25.0                </span></span>
<span><span class="co">## [101] ggnetwork_0.5.10            Biostrings_2.66.0          </span></span>
<span><span class="co">## [103] blob_1.2.3                  mixsqp_0.3-43              </span></span>
<span><span class="co">## [105] stringr_1.4.1               echoplot_0.99.5            </span></span>
<span><span class="co">## [107] dnet_1.1.7                  readr_2.1.3                </span></span>
<span><span class="co">## [109] jpeg_0.1-9                  S4Vectors_0.36.0           </span></span>
<span><span class="co">## [111] echodata_0.99.15            scales_1.2.1               </span></span>
<span><span class="co">## [113] memoise_2.0.1               magrittr_2.0.3             </span></span>
<span><span class="co">## [115] plyr_1.8.7                  hexbin_1.28.2              </span></span>
<span><span class="co">## [117] zlibbioc_1.44.0             compiler_4.2.1             </span></span>
<span><span class="co">## [119] echoconda_0.99.8            BiocIO_1.8.0               </span></span>
<span><span class="co">## [121] RColorBrewer_1.1-3          catalogueR_1.0.0           </span></span>
<span><span class="co">## [123] EnsDb.Hsapiens.v75_2.99.0   Rsamtools_2.14.0           </span></span>
<span><span class="co">## [125] cli_3.4.1                   XVector_0.38.0             </span></span>
<span><span class="co">## [127] echoannot_0.99.10           patchwork_1.1.2            </span></span>
<span><span class="co">## [129] htmlTable_2.4.1             Formula_1.2-4              </span></span>
<span><span class="co">## [131] MASS_7.3-58.1               tidyselect_1.2.0           </span></span>
<span><span class="co">## [133] stringi_1.7.8               textshaping_0.3.6          </span></span>
<span><span class="co">## [135] highr_0.9                   yaml_2.3.6                 </span></span>
<span><span class="co">## [137] supraHex_1.35.0             latticeExtra_0.6-30        </span></span>
<span><span class="co">## [139] ggrepel_0.9.1               grid_4.2.1                 </span></span>
<span><span class="co">## [141] sass_0.4.2                  VariantAnnotation_1.44.0   </span></span>
<span><span class="co">## [143] tools_4.2.1                 lmom_2.9                   </span></span>
<span><span class="co">## [145] parallel_4.2.1              rstudioapi_0.14            </span></span>
<span><span class="co">## [147] foreign_0.8-83              piggyback_0.1.4            </span></span>
<span><span class="co">## [149] gridExtra_2.3               gld_2.6.6                  </span></span>
<span><span class="co">## [151] farver_2.1.1                RcppZiggurat_0.1.6         </span></span>
<span><span class="co">## [153] digest_0.6.30               BiocManager_1.30.19        </span></span>
<span><span class="co">## [155] Rcpp_1.0.9                  GenomicRanges_1.50.0       </span></span>
<span><span class="co">## [157] OrganismDbi_1.40.0          httr_1.4.4                 </span></span>
<span><span class="co">## [159] AnnotationDbi_1.60.0        RCircos_1.2.2              </span></span>
<span><span class="co">## [161] ggbio_1.46.0                biovizBase_1.46.0          </span></span>
<span><span class="co">## [163] colorspace_2.0-3            XML_3.99-0.12              </span></span>
<span><span class="co">## [165] fs_1.5.2                    reticulate_1.26            </span></span>
<span><span class="co">## [167] IRanges_2.32.0              splines_4.2.1              </span></span>
<span><span class="co">## [169] RBGL_1.74.0                 expm_0.999-6               </span></span>
<span><span class="co">## [171] pkgdown_2.0.6               echofinemap_0.99.4         </span></span>
<span><span class="co">## [173] basilisk_1.9.12             Exact_3.2                  </span></span>
<span><span class="co">## [175] mapproj_1.2.9               systemfonts_1.0.4          </span></span>
<span><span class="co">## [177] jsonlite_1.8.3              Rfast_2.0.6                </span></span>
<span><span class="co">## [179] susieR_0.12.27              R6_2.5.1                   </span></span>
<span><span class="co">## [181] Hmisc_4.7-1                 pillar_1.8.1               </span></span>
<span><span class="co">## [183] htmltools_0.5.3             glue_1.6.2                 </span></span>
<span><span class="co">## [185] fastmap_1.1.0               DT_0.26                    </span></span>
<span><span class="co">## [187] BiocParallel_1.32.0         class_7.3-20               </span></span>
<span><span class="co">## [189] codetools_0.2-18            maps_3.4.1                 </span></span>
<span><span class="co">## [191] mvtnorm_1.1-3               utf8_1.2.2                 </span></span>
<span><span class="co">## [193] lattice_0.20-45             bslib_0.4.0                </span></span>
<span><span class="co">## [195] tibble_3.1.8                curl_4.3.3                 </span></span>
<span><span class="co">## [197] DescTools_0.99.47           zip_2.2.2                  </span></span>
<span><span class="co">## [199] openxlsx_4.2.5.1            interp_1.1-3               </span></span>
<span><span class="co">## [201] rmarkdown_2.17              googleAuthR_2.0.0          </span></span>
<span><span class="co">## [203] desc_1.4.2                  munsell_0.5.0              </span></span>
<span><span class="co">## [205] e1071_1.7-12                GenomeInfoDbData_1.2.9     </span></span>
<span><span class="co">## [207] reshape2_1.4.4              gtable_0.3.1</span></span></code></pre>
</details><p><br></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
