<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>echolocatoR Fine-mapping Portal • echolocatoR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="echolocatoR Fine-mapping Portal">
<meta property="og:description" content="echolocatoR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">echolocatoR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/echolocatoR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/QTLs.html">QTLs</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="../articles/finemapping_portal.html">echolocatoR Fine-mapping Portal</a>
    </li>
    <li>
      <a href="../articles/plot_locus.html">Plot locus</a>
    </li>
    <li>
      <a href="../articles/summarise.html">Summarise</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RajLabMSSM/echolocatoR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>echolocatoR Fine-mapping Portal</h1>
                        <h4 data-toc-skip class="author"><h4> ¶ Author:
<i>Brian M. Schilder</i> ¶ </h4></h4>
            
            <h4 data-toc-skip class="date"><h4> ¶ Updated:
<i>Nov-07-2022</i> ¶ </h4></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/RajLabMSSM/echolocatoR/blob/HEAD/vignettes/finemapping_portal.Rmd" class="external-link"><code>vignettes/finemapping_portal.Rmd</code></a></small>
      <div class="hidden name"><code>finemapping_portal.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RajLabMSSM/echolocatoR" class="external-link">echolocatoR</a></span><span class="op">)</span>   </span></code></pre></div>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">##   method from   </span></span>
<span><span class="co">##   +.gg   ggplot2</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">⠊⠉⠡⣀⣀⠊⠉⠡⣀⣀⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⢁⡐⠉⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠉⢂⡈⠑⣀⠉⢄⡈⠡⣀</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⡈⡐⢂⢁⠒⡈⡐⢂⢁⠒⡈⡐⢂⢁⠑⡈⡈⢄⢁⠡⠌⡈⠤⢁⠡⠌⡈⠤⢁⠡⠌⡈⡠⢁⢁⠊⡈⡐⢂</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="color: #55FFFF; font-weight: bold;">🦇  🦇  🦇 e c h o l o c a t o R 🦇  🦇  🦇</span> <span style="color: #00BBBB;">─────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="color: #55FFFF; font-weight: bold;">v2.0.3</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">⠌⡈⡐⢂⢁⠒⡈⡐⢂⢁⠒⡈⡐⢂⢁⠑⡈⡈⢄⢁⠡⠌⡈⠤⢁⠡⠌⡈⠤⢁⠡⠌⡈⡠⢁⢁⠊⡈⡐⢂</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⢁⡐⠉⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠉⢂⡈⠑⣀⠉⢄⡈⠡⣀</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠊⠉⠡⣀⣀⠊⠉⠡⣀⣀⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠</span>                                    </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ If you use </span><span style="color: #00BBBB; font-weight: bold;">echolocatoR</span><span style="color: #00BBBB;"> or any of the </span><span style="color: #00BBBB; font-style: italic;">echoverse</span><span style="color: #00BBBB;"> subpackages, please cite:</span>      </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶ Brian M Schilder, Jack Humphrey, Towfique</span>                                </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     Raj (2021) echolocatoR: an automated</span>                                       </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     end-to-end statistical and functional</span>                                      </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     genomic fine-mapping pipeline,</span>                                             </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     Bioinformatics; btab658,</span>                                                   </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://doi.org/10.1093/bioinformatics/btab658</span>                             </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ Please report any bugs/feature requests on GitHub:</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://github.com/RajLabMSSM/echolocatoR/issues</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ Contributions are welcome!:</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://github.com/RajLabMSSM/echolocatoR/pulls</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<p>Here we take advantage of the fine-mapping results files already
available on the <a href="https://rajlab.shinyapps.io/Fine_Mapping_Shiny" class="external-link">echolocatoR
Fine-mapping Portal</a>.</p>
<div class="section level2">
<h2 id="download-data">Download data<a class="anchor" aria-label="anchor" href="#download-data"></a>
</h2>
<div class="section level3">
<h3 id="select-loci">Select loci<a class="anchor" aria-label="anchor" href="#select-loci"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Limit the number of loci for demo purposes</span></span>
<span><span class="va">loci</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/echodata/man/topSNPs_Nalls2019.html" class="external-link">topSNPs_Nalls2019</a></span><span class="op">$</span><span class="va">Locus</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="portal-query">Portal query<a class="anchor" aria-label="anchor" href="#portal-query"></a>
</h3>
<p>We can query multiple studies, loci, and even data types at once.
When <code>as_datatable=TRUE</code>, all file paths and metadata are
conveniently organized into a <code>data.table.</code></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">local_files</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/portal_query.html" class="external-link">portal_query</a></span><span class="op">(</span>dataset_types<span class="op">=</span><span class="st">"GWAS"</span>,</span>
<span>                                      phenotypes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"parkinson"</span><span class="op">)</span>,</span>
<span>                                      file_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"multi_finemap"</span>,<span class="st">"LD"</span><span class="op">)</span>, </span>
<span>                                      loci <span class="op">=</span> <span class="va">loci</span>,</span>
<span>                                      LD_panels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UKB"</span><span class="op">)</span>,</span>
<span>                                      results_dir <span class="op">=</span> <span class="va">results_dir</span>, </span>
<span>                                      as_datatable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Fetching echolocatoR Fine-mapping Portal study metadata.</span></span></code></pre>
<pre><code><span><span class="co">## + 1 dataset(s) remain after filtering.</span></span></code></pre>
<pre><code><span><span class="co">## + Searching for multi_finemap files...</span></span></code></pre>
<pre><code><span><span class="co">## OK (HTTP 200).+ Searching for LD files...</span></span>
<span><span class="co">## OK (HTTP 200).+ 6 unique files identified.</span></span>
<span><span class="co">## + Downloading 4 files...</span></span>
<span><span class="co">## + Returning table with local file paths.</span></span></code></pre>
<div class="section level4">
<h4 id="merge-fine-mapping-results">Merge fine-mapping results<a class="anchor" aria-label="anchor" href="#merge-fine-mapping-results"></a>
</h4>
<p>Next, we can gather all of the fine-mapping results generated by
<code><a href="../reference/finemap_loci.html">finemap_loci()</a></code> previously.<br><code>merge_finemapping_results</code> recursively searches for the
correct files within a hierarchical folder structure and imports only
the multi-finemap files.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">merged_DT</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/merge_finemapping_results.html" class="external-link">merge_finemapping_results</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="va">results_dir</span>,   </span>
<span>                                                 minimum_support <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                                                 include_leadSNPs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                                 consensus_thresh <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## + Gathering all fine-mapping results from storage...</span></span></code></pre>
<pre><code><span><span class="co">## + 2 multi-finemap files found.</span></span></code></pre>
<pre><code><span><span class="co">## + Importing results... ASXL3</span></span></code></pre>
<pre><code><span><span class="co">## + Importing results... BIN3</span></span></code></pre>
<pre><code><span><span class="co">## Identifying Consensus SNPs...</span></span></code></pre>
<pre><code><span><span class="co">## + support_thresh = 2</span></span></code></pre>
<pre><code><span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span></code></pre>
<pre><code><span><span class="co">## + 4 fine-mapping methods used.</span></span></code></pre>
<pre><code><span><span class="co">## + 17 Credible Set SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + 6 Consensus SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + Saving merged results ==&gt; /tmp/RtmpJQtSZ8/file217973ffe623merged_results.csv.gz</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/results_report.html" class="external-link">results_report</a></span><span class="op">(</span><span class="va">merged_DT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## echolocatoR results report (all loci):</span></span></code></pre>
<pre><code><span><span class="co">## + Overall report:</span></span></code></pre>
<pre><code><span><span class="co">## ++ 2 Loci.</span></span></code></pre>
<pre><code><span><span class="co">## ++ 10605 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + Lead SNP report:</span></span></code></pre>
<pre><code><span><span class="co">## ++ 2 lead SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Lead SNP mean PP = 0.5</span></span></code></pre>
<pre><code><span><span class="co">## + Union Credible Set report:</span></span></code></pre>
<pre><code><span><span class="co">## ++ 17 UCS SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## ++ UCS mean PP = 0.35</span></span></code></pre>
<pre><code><span><span class="co">## ++ 2 UCS SNPs that are also lead SNPs</span></span></code></pre>
<pre><code><span><span class="co">## + Consensus SNP report:</span></span></code></pre>
<pre><code><span><span class="co">## ++ 6 Consensus SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Consensus SNP mean PP = 0.54</span></span></code></pre>
<pre><code><span><span class="co">## ++ 1 Consensus SNPs that are also lead SNPs</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="import-ld">Import LD<a class="anchor" aria-label="anchor" href="#import-ld"></a>
</h3>
<p>Next, we import the a subset of the LD matrices for only the lead
SNP.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_files</span> <span class="op">&lt;-</span> <span class="va">local_files</span><span class="op">[</span><span class="va">file_type</span><span class="op">==</span><span class="st">"LD"</span>,<span class="op">]</span></span>
<span><span class="va">ld_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">ld_files</span><span class="op">$</span><span class="va">local_file</span>, </span>
<span>                                      <span class="va">ld_files</span><span class="op">$</span><span class="va">locus</span><span class="op">)</span>,</span>
<span>                      <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ld_matrices</span><span class="op">$</span><span class="va">ASXL3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">SNP</th>
<th align="right">rs1941685</th>
<th align="right">rs1941685.1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">rs12968480</td>
<td align="right">0.0396723</td>
<td align="right">0.0396723</td>
</tr>
<tr class="even">
<td align="left">rs12967667</td>
<td align="right">0.0424208</td>
<td align="right">0.0424208</td>
</tr>
<tr class="odd">
<td align="left">rs9945156</td>
<td align="right">0.0424096</td>
<td align="right">0.0424096</td>
</tr>
<tr class="even">
<td align="left">rs1851700</td>
<td align="right">0.0419115</td>
<td align="right">0.0419115</td>
</tr>
<tr class="odd">
<td align="left">rs117840441</td>
<td align="right">0.0664300</td>
<td align="right">0.0664300</td>
</tr>
<tr class="even">
<td align="left">rs1523592</td>
<td align="right">0.0424137</td>
<td align="right">0.0424137</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="plot-locus">Plot locus<a class="anchor" aria-label="anchor" href="#plot-locus"></a>
</h2>
<p>Now let’s plot one locus as an example.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">locus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">merged_DT</span><span class="op">$</span><span class="va">Locus</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="co"># Pick the first locus</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">merged_DT</span><span class="op">[</span><span class="va">Locus</span><span class="op">==</span><span class="va">locus</span>,<span class="op">]</span> </span>
<span><span class="va">LD_matrix</span> <span class="op">=</span> <span class="va">ld_matrices</span><span class="op">[[</span><span class="va">locus</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">locus_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,<span class="va">locus</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu">echoplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">plot_locus</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">dat</span>,   </span>
<span>                    LD_matrix <span class="op">=</span> <span class="va">LD_matrix</span>,</span>
<span>                    LD_reference <span class="op">=</span> <span class="st">"UKB"</span>,</span>
<span>                    locus_dir <span class="op">=</span> <span class="va">locus_dir</span>, </span>
<span>                      </span>
<span>                    nott_epigenome <span class="op">=</span> <span class="cn">TRUE</span>,   </span>
<span>                    nott_regulatory_rects <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                    nott_show_placseq <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    zoom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"20x"</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre style="max-height: 400px;"><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.5 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] echolocatoR_2.0.3 BiocStyle_2.26.0 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rappdirs_0.3.3              rtracklayer_1.58.0         </span></span>
<span><span class="co">##   [3] GGally_2.1.2                R.methodsS3_1.8.2          </span></span>
<span><span class="co">##   [5] ragg_1.2.4                  tidyr_1.2.1                </span></span>
<span><span class="co">##   [7] echoLD_0.99.8               ggplot2_3.3.6              </span></span>
<span><span class="co">##   [9] bit64_4.0.5                 knitr_1.40                 </span></span>
<span><span class="co">##  [11] irlba_2.3.5.1               DelayedArray_0.24.0        </span></span>
<span><span class="co">##  [13] R.utils_2.12.1              data.table_1.14.4          </span></span>
<span><span class="co">##  [15] rpart_4.1.19                KEGGREST_1.38.0            </span></span>
<span><span class="co">##  [17] RCurl_1.98-1.9              AnnotationFilter_1.22.0    </span></span>
<span><span class="co">##  [19] generics_0.1.3              BiocGenerics_0.44.0        </span></span>
<span><span class="co">##  [21] GenomicFeatures_1.50.2      RSQLite_2.2.18             </span></span>
<span><span class="co">##  [23] proxy_0.4-27                bit_4.0.4                  </span></span>
<span><span class="co">##  [25] tzdb_0.3.0                  xml2_1.3.3                 </span></span>
<span><span class="co">##  [27] SummarizedExperiment_1.28.0 assertthat_0.2.1           </span></span>
<span><span class="co">##  [29] viridis_0.6.2               xfun_0.34                  </span></span>
<span><span class="co">##  [31] hms_1.1.2                   jquerylib_0.1.4            </span></span>
<span><span class="co">##  [33] evaluate_0.17               fansi_1.0.3                </span></span>
<span><span class="co">##  [35] restfulr_0.0.15             progress_1.2.2             </span></span>
<span><span class="co">##  [37] dbplyr_2.2.1                readxl_1.4.1               </span></span>
<span><span class="co">##  [39] Rgraphviz_2.41.2            igraph_1.3.5               </span></span>
<span><span class="co">##  [41] DBI_1.1.3                   htmlwidgets_1.5.4          </span></span>
<span><span class="co">##  [43] reshape_0.8.9               downloadR_0.99.5           </span></span>
<span><span class="co">##  [45] stats4_4.2.1                purrr_0.3.5                </span></span>
<span><span class="co">##  [47] ellipsis_0.3.2              dplyr_1.0.10               </span></span>
<span><span class="co">##  [49] backports_1.4.1             bookdown_0.29              </span></span>
<span><span class="co">##  [51] biomaRt_2.54.0              deldir_1.0-6               </span></span>
<span><span class="co">##  [53] MatrixGenerics_1.10.0       vctrs_0.5.0                </span></span>
<span><span class="co">##  [55] Biobase_2.58.0              ensembldb_2.22.0           </span></span>
<span><span class="co">##  [57] cachem_1.0.6                withr_2.5.0                </span></span>
<span><span class="co">##  [59] BSgenome_1.66.1             checkmate_2.1.0            </span></span>
<span><span class="co">##  [61] GenomicAlignments_1.34.0    prettyunits_1.1.1          </span></span>
<span><span class="co">##  [63] cluster_2.1.4               ape_5.6-2                  </span></span>
<span><span class="co">##  [65] dir.expiry_1.5.1            lazyeval_0.2.2             </span></span>
<span><span class="co">##  [67] crayon_1.5.2                basilisk.utils_1.9.4       </span></span>
<span><span class="co">##  [69] crul_1.3                    pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [71] GenomeInfoDb_1.34.1         nlme_3.1-160               </span></span>
<span><span class="co">##  [73] ProtGenerics_1.30.0         XGR_1.1.8                  </span></span>
<span><span class="co">##  [75] nnet_7.3-18                 pals_1.7                   </span></span>
<span><span class="co">##  [77] rlang_1.0.6                 lifecycle_1.0.3            </span></span>
<span><span class="co">##  [79] filelock_1.0.2              httpcode_0.3.0             </span></span>
<span><span class="co">##  [81] BiocFileCache_2.6.0         echotabix_0.99.8           </span></span>
<span><span class="co">##  [83] dichromat_2.0-0.1           cellranger_1.1.0           </span></span>
<span><span class="co">##  [85] coloc_5.1.0.1               rprojroot_2.0.3            </span></span>
<span><span class="co">##  [87] matrixStats_0.62.0          graph_1.76.0               </span></span>
<span><span class="co">##  [89] Matrix_1.5-1                osfr_0.2.9                 </span></span>
<span><span class="co">##  [91] boot_1.3-28                 base64enc_0.1-3            </span></span>
<span><span class="co">##  [93] png_0.1-7                   viridisLite_0.4.1          </span></span>
<span><span class="co">##  [95] rjson_0.2.21                rootSolve_1.8.2.3          </span></span>
<span><span class="co">##  [97] bitops_1.0-7                R.oo_1.25.0                </span></span>
<span><span class="co">##  [99] ggnetwork_0.5.10            Biostrings_2.66.0          </span></span>
<span><span class="co">## [101] blob_1.2.3                  mixsqp_0.3-43              </span></span>
<span><span class="co">## [103] stringr_1.4.1               echoplot_0.99.5            </span></span>
<span><span class="co">## [105] dnet_1.1.7                  readr_2.1.3                </span></span>
<span><span class="co">## [107] jpeg_0.1-9                  S4Vectors_0.36.0           </span></span>
<span><span class="co">## [109] echodata_0.99.15            scales_1.2.1               </span></span>
<span><span class="co">## [111] memoise_2.0.1               magrittr_2.0.3             </span></span>
<span><span class="co">## [113] plyr_1.8.7                  hexbin_1.28.2              </span></span>
<span><span class="co">## [115] zlibbioc_1.44.0             compiler_4.2.1             </span></span>
<span><span class="co">## [117] echoconda_0.99.8            BiocIO_1.8.0               </span></span>
<span><span class="co">## [119] RColorBrewer_1.1-3          catalogueR_1.0.0           </span></span>
<span><span class="co">## [121] Rsamtools_2.14.0            cli_3.4.1                  </span></span>
<span><span class="co">## [123] XVector_0.38.0              echoannot_0.99.10          </span></span>
<span><span class="co">## [125] patchwork_1.1.2             htmlTable_2.4.1            </span></span>
<span><span class="co">## [127] Formula_1.2-4               MASS_7.3-58.1              </span></span>
<span><span class="co">## [129] tidyselect_1.2.0            stringi_1.7.8              </span></span>
<span><span class="co">## [131] textshaping_0.3.6           highr_0.9                  </span></span>
<span><span class="co">## [133] yaml_2.3.6                  supraHex_1.35.0            </span></span>
<span><span class="co">## [135] latticeExtra_0.6-30         ggrepel_0.9.1              </span></span>
<span><span class="co">## [137] grid_4.2.1                  sass_0.4.2                 </span></span>
<span><span class="co">## [139] VariantAnnotation_1.44.0    tools_4.2.1                </span></span>
<span><span class="co">## [141] lmom_2.9                    parallel_4.2.1             </span></span>
<span><span class="co">## [143] rstudioapi_0.14             foreign_0.8-83             </span></span>
<span><span class="co">## [145] piggyback_0.1.4             gridExtra_2.3              </span></span>
<span><span class="co">## [147] gld_2.6.6                   digest_0.6.30              </span></span>
<span><span class="co">## [149] snpStats_1.47.1             BiocManager_1.30.19        </span></span>
<span><span class="co">## [151] Rcpp_1.0.9                  GenomicRanges_1.50.0       </span></span>
<span><span class="co">## [153] OrganismDbi_1.40.0          httr_1.4.4                 </span></span>
<span><span class="co">## [155] AnnotationDbi_1.60.0        RCircos_1.2.2              </span></span>
<span><span class="co">## [157] ggbio_1.46.0                biovizBase_1.46.0          </span></span>
<span><span class="co">## [159] colorspace_2.0-3            XML_3.99-0.12              </span></span>
<span><span class="co">## [161] fs_1.5.2                    reticulate_1.26            </span></span>
<span><span class="co">## [163] IRanges_2.32.0              splines_4.2.1              </span></span>
<span><span class="co">## [165] RBGL_1.74.0                 expm_0.999-6               </span></span>
<span><span class="co">## [167] pkgdown_2.0.6               echofinemap_0.99.4         </span></span>
<span><span class="co">## [169] basilisk_1.9.12             Exact_3.2                  </span></span>
<span><span class="co">## [171] mapproj_1.2.9               systemfonts_1.0.4          </span></span>
<span><span class="co">## [173] jsonlite_1.8.3              susieR_0.12.27             </span></span>
<span><span class="co">## [175] R6_2.5.1                    Hmisc_4.7-1                </span></span>
<span><span class="co">## [177] pillar_1.8.1                htmltools_0.5.3            </span></span>
<span><span class="co">## [179] glue_1.6.2                  fastmap_1.1.0              </span></span>
<span><span class="co">## [181] DT_0.26                     BiocParallel_1.32.0        </span></span>
<span><span class="co">## [183] class_7.3-20                codetools_0.2-18           </span></span>
<span><span class="co">## [185] maps_3.4.1                  mvtnorm_1.1-3              </span></span>
<span><span class="co">## [187] utf8_1.2.2                  lattice_0.20-45            </span></span>
<span><span class="co">## [189] bslib_0.4.0                 tibble_3.1.8               </span></span>
<span><span class="co">## [191] curl_4.3.3                  DescTools_0.99.47          </span></span>
<span><span class="co">## [193] zip_2.2.2                   openxlsx_4.2.5.1           </span></span>
<span><span class="co">## [195] interp_1.1-3                survival_3.4-0             </span></span>
<span><span class="co">## [197] rmarkdown_2.17              desc_1.4.2                 </span></span>
<span><span class="co">## [199] munsell_0.5.0               e1071_1.7-12               </span></span>
<span><span class="co">## [201] GenomeInfoDbData_1.2.9      reshape2_1.4.4             </span></span>
<span><span class="co">## [203] gtable_0.3.1</span></span></code></pre>
</details><p><br></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
