<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Plot locus ‚Ä¢ echolocatoR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Plot locus">
<meta property="og:description" content="echolocatoR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">echolocatoR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/echolocatoR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/QTLs.html">QTLs</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="../articles/finemapping_portal.html">echolocatoR Fine-mapping Portal</a>
    </li>
    <li>
      <a href="../articles/plot_locus.html">Plot locus</a>
    </li>
    <li>
      <a href="../articles/summarise.html">Summarise</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RajLabMSSM/echolocatoR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Plot locus</h1>
                        <h4 data-toc-skip class="author"><h4> ¬∂ Author:
<i>Brian M. Schilder</i> ¬∂ </h4></h4>
            
            <h4 data-toc-skip class="date"><h4> ¬∂ Updated:
<i>Nov-06-2022</i> ¬∂ </h4></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/RajLabMSSM/echolocatoR/blob/HEAD/vignettes/plot_locus.Rmd" class="external-link"><code>vignettes/plot_locus.Rmd</code></a></small>
      <div class="hidden name"><code>plot_locus.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RajLabMSSM/echolocatoR" class="external-link">echolocatoR</a></span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">##   method from   </span></span>
<span><span class="co">##   +.gg   ggplot2</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">‚†ä‚†â‚†°‚£Ä‚£Ä‚†ä‚†â‚†°‚£Ä‚£Ä‚†ä‚†â‚†¢‚£Ä‚°†‚†ä‚†â‚†¢‚£Ä‚°†‚†ä‚†â‚†¢‚£Ä‚°†‚†ä‚†â‚†¢‚£Ä‚°†‚†ä‚†â‚†¢‚£Ä‚°†‚†ä‚†â‚†¢‚£Ä‚°†</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">‚†å‚¢Å‚°ê‚†â‚£Ä‚†ä‚¢Ç‚°ê‚†ë‚£Ä‚†ä‚¢Ç‚°ê‚†ë‚£Ä‚†ä‚¢Ç‚°ê‚†ë‚£Ä‚†ä‚¢Ç‚°ê‚†ë‚£Ä‚†ä‚¢Ç‚°ê‚†ë‚£Ä‚†â‚¢Ç‚°à‚†ë‚£Ä‚†â‚¢Ñ‚°à‚†°‚£Ä</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">‚†å‚°à‚°ê‚¢Ç‚¢Å‚†í‚°à‚°ê‚¢Ç‚¢Å‚†í‚°à‚°ê‚¢Ç‚¢Å‚†ë‚°à‚°à‚¢Ñ‚¢Å‚†°‚†å‚°à‚†§‚¢Å‚†°‚†å‚°à‚†§‚¢Å‚†°‚†å‚°à‚°†‚¢Å‚¢Å‚†ä‚°à‚°ê‚¢Ç</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">‚îÄ‚îÄ</span> <span style="color: #55FFFF; font-weight: bold;">ü¶á  ü¶á  ü¶á e c h o l o c a t o R ü¶á  ü¶á  ü¶á</span> <span style="color: #00BBBB;">‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">‚îÄ‚îÄ</span> <span style="color: #55FFFF; font-weight: bold;">v2.0.2</span> <span style="color: #00BBBB;">‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">‚†å‚°à‚°ê‚¢Ç‚¢Å‚†í‚°à‚°ê‚¢Ç‚¢Å‚†í‚°à‚°ê‚¢Ç‚¢Å‚†ë‚°à‚°à‚¢Ñ‚¢Å‚†°‚†å‚°à‚†§‚¢Å‚†°‚†å‚°à‚†§‚¢Å‚†°‚†å‚°à‚°†‚¢Å‚¢Å‚†ä‚°à‚°ê‚¢Ç</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">‚†å‚¢Å‚°ê‚†â‚£Ä‚†ä‚¢Ç‚°ê‚†ë‚£Ä‚†ä‚¢Ç‚°ê‚†ë‚£Ä‚†ä‚¢Ç‚°ê‚†ë‚£Ä‚†ä‚¢Ç‚°ê‚†ë‚£Ä‚†ä‚¢Ç‚°ê‚†ë‚£Ä‚†â‚¢Ç‚°à‚†ë‚£Ä‚†â‚¢Ñ‚°à‚†°‚£Ä</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">‚†ä‚†â‚†°‚£Ä‚£Ä‚†ä‚†â‚†°‚£Ä‚£Ä‚†ä‚†â‚†¢‚£Ä‚°†‚†ä‚†â‚†¢‚£Ä‚°†‚†ä‚†â‚†¢‚£Ä‚°†‚†ä‚†â‚†¢‚£Ä‚°†‚†ä‚†â‚†¢‚£Ä‚°†‚†ä‚†â‚†¢‚£Ä‚°†</span>                                    </span></span>
<span><span class="co">## <span style="color: #00BBBB;">‚ìû If you use </span><span style="color: #00BBBB; font-weight: bold;">echolocatoR</span><span style="color: #00BBBB;"> or any of the </span><span style="color: #00BBBB; font-style: italic;">echoverse</span><span style="color: #00BBBB;"> subpackages, please cite:</span>      </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ‚ñ∂ Brian M Schilder, Jack Humphrey, Towfique</span>                                </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     Raj (2021) echolocatoR: an automated</span>                                       </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     end-to-end statistical and functional</span>                                      </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     genomic fine-mapping pipeline,</span>                                             </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     Bioinformatics; btab658,</span>                                                   </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://doi.org/10.1093/bioinformatics/btab658</span>                             </span></span>
<span><span class="co">## <span style="color: #00BBBB;">‚ìû Please report any bugs/feature requests on GitHub:</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ‚ñ∂</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://github.com/RajLabMSSM/echolocatoR/issues</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">‚ìû Contributions are welcome!:</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ‚ñ∂</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://github.com/RajLabMSSM/echolocatoR/pulls</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span></span></code></pre>
<div class="section level2">
<h2 id="plotting-loci-with-echolocator">Plotting loci with <em>echolocatoR</em><a class="anchor" aria-label="anchor" href="#plotting-loci-with-echolocator"></a>
</h2>
<p><em>echolocatoR</em> contains various functions that can be used
separately<br>
from the comprehensive <code><a href="../reference/finemap_loci.html">finemap_loci()</a></code> pipeline.</p>
<p>Generate a multi-view plot of a given locus using
<code><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">echoplot::plot_locus()</a></code>.</p>
<ul>
<li>You can mix and match different tracks and annotations using the
different arguments (see <code><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">?echoplot::plot_locus</a></code> for
details).</li>
</ul>
<p>The plot is centered on the lead/index SNP. If a list is supplied to
plot.xoom * <code><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">echoplot::plot_locus()</a></code> returns a series of
<code>ggplot</code> objects bound together with <a href="https://patchwork.data-imaginist.com" class="external-link"><code>patchwork</code></a>.
One can further modify this object using <code>ggplot2</code> functions
like <code>+ theme()</code>. + The modifications will be applied to all
tracks at once.</p>
<ul>
<li>Save a high-resolution versions the plot by setting
<code>save_plot=T</code>.
<ul>
<li>Further increase resolution by adjusting the <code>dpi</code>
argument (<em>default=300</em>).</li>
<li>Files are saved in <em>jpg</em> format by default, but users can
specify their preferred file format
(e.g.¬†<code>file_format="png"</code>)</li>
<li>Adjust the <code>height</code> and <code>width</code> of the saved
plot using these respective arguments.</li>
<li>The plot will be automatically saved in the locus-specific directory
as:<br>
*multiview_<locus>_&lt;plot.zoom&gt;.jpg*.</locus>
</li>
</ul>
</li>
</ul>
<div class="section level3">
<h3 id="load-example-data">Load example data<a class="anchor" aria-label="anchor" href="#load-example-data"></a>
</h3>
<p>Load example dataset of the results from fine-mapping the BST1 locus
with <code><a href="../reference/finemap_loci.html">finemap_loci()</a></code>. Original data comes from the recent
Nalls et al.¬†(2019) Parkinson‚Äôs disease GWAS (see <code><a href="https://rdrr.io/pkg/echodata/man/BST1.html" class="external-link">?BST1</a></code> for
details).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/echodata/man/BST1.html" class="external-link">BST1</a></span></span>
<span><span class="va">LD_matrix</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/echodata/man/BST1_LD_matrix.html" class="external-link">BST1_LD_matrix</a></span></span>
<span><span class="va">locus_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu">echodata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/echodata/man/locus_dir.html" class="external-link">locus_dir</a></span><span class="op">)</span></span>
<span><span class="va">LD_reference</span> <span class="op">&lt;-</span> <span class="st">"UKB"</span> <span class="co"># Used for naming saved plots </span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="full-window">Full window<a class="anchor" aria-label="anchor" href="#full-window"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trk_plot</span> <span class="op">&lt;-</span> <span class="fu">echoplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">plot_locus</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">dat</span>, </span>
<span>                                 LD_matrix <span class="op">=</span> <span class="va">LD_matrix</span>, </span>
<span>                                 LD_reference <span class="op">=</span> <span class="va">LD_reference</span>,</span>
<span>                                 locus_dir <span class="op">=</span> <span class="va">locus_dir</span>,</span>
<span>                                 zoom <span class="op">=</span> <span class="st">"10x"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## +-------- Locus Plot:  BST1 --------+</span></span></code></pre>
<pre><code><span><span class="co">## + support_thresh = 2</span></span></code></pre>
<pre><code><span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span></code></pre>
<pre><code><span><span class="co">## + 4 fine-mapping methods used.</span></span></code></pre>
<pre><code><span><span class="co">## + 7 Credible Set SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + 3 Consensus SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in CS cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in PP cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## LD_matrix detected. Coloring SNPs by LD with lead SNP.</span></span></code></pre>
<pre><code><span><span class="co">## Filling r/r2 NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: GWAS full window track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: GWAS track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Merged fine-mapping track</span></span></code></pre>
<pre><code><span><span class="co">## Melting PP and CS from 5 fine-mapping methods.</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Adding Gene model track.</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: EnsDb.Hsapiens.v75</span></span></code></pre>
<pre><code><span><span class="co">## max_transcripts= 1 .</span></span></code></pre>
<pre><code><span><span class="co">## 16  transcripts from  16  genes returned.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: pals</span></span></code></pre>
<pre><code><span><span class="co">## Fetching data...OK</span></span>
<span><span class="co">## Parsing exons...OK</span></span>
<span><span class="co">## Defining introns...OK</span></span>
<span><span class="co">## Defining UTRs...OK</span></span>
<span><span class="co">## Defining CDS...OK</span></span>
<span><span class="co">## aggregating...</span></span>
<span><span class="co">## Done</span></span>
<span><span class="co">## Constructing graphics...</span></span>
<span><span class="co">## Loading required namespace: pals</span></span>
<span><span class="co">## + Adding vertical lines to highlight SNP groups...</span></span>
<span><span class="co">## +&gt;+&gt;+&gt;+&gt;+ zoom =  10x  +&lt;+&lt;+&lt;+&lt;+</span></span>
<span><span class="co">## + echoplot:: Get window suffix...</span></span>
<span><span class="co">## + Constructing zoom polygon...</span></span>
<span><span class="co">## + Highlighting zoom origin...</span></span>
<span><span class="co">## + Removing subplot margins...</span></span>
<span><span class="co">## + Reordering tracks...</span></span></code></pre>
<pre><code><span><span class="co">## [1] "+ Ensuring last track shows genomic units..."</span></span></code></pre>
<pre><code><span><span class="co">## + Aligning xlimits for each subplot...</span></span>
<span><span class="co">## + Checking track heights...</span></span>
<span><span class="co">## Found more than one class "simpleUnit" in cache; using the first, from namespace 'hexbin'</span></span>
<span><span class="co">## Also defined by 'ggbio'</span></span></code></pre>
<p><img src="plot_locus_files/figure-html/trk_plot-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="at-multiple-zooms">At multiple zooms<a class="anchor" aria-label="anchor" href="#at-multiple-zooms"></a>
</h3>
<ul>
<li>You can easily generate the same locus plot at multiple zoomed in
views by supplying a list to <code>plot.zoom</code>.<br>
</li>
<li>This list can be composed of zoom multipliers
(e.g.¬†<code>c("1x", "2x")</code>), window widths in units of basepairs
(e.g.¬†<code>c(5000, 1500)</code>), or a mixture of both
(e.g.¬†<code>c("1x","4x", 5000, 2000)</code>).</li>
<li>Each zoom view will be saved individually with its respective scale
as the suffix (e.g.¬†<code>multiview.BST1.UKB.4x.jpg</code>).<br>
</li>
<li>Each zoom view is stored as a named item within the returned
list.</li>
</ul>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trk_plot_zooms</span> <span class="op">&lt;-</span> <span class="fu">echoplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">plot_locus</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">dat</span>, </span>
<span>                                       LD_matrix <span class="op">=</span> <span class="va">LD_matrix</span>, </span>
<span>                                       LD_reference <span class="op">=</span> <span class="va">LD_reference</span>,</span>
<span>                                       locus_dir <span class="op">=</span> <span class="va">locus_dir</span>,</span>
<span>                                       zoom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1x"</span>,<span class="st">"5x"</span>,<span class="st">"10x"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">trk_plot_zooms</span><span class="op">)</span> <span class="co"># Get zoom view names</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="return-as-list">Return as list<a class="anchor" aria-label="anchor" href="#return-as-list"></a>
</h3>
<ul>
<li>For even further control over each track of the multi-view plot,
specify <code>echoplot::plot_locus(..., return_list=T)</code> to instead
return a named list (nested within each zoom view list item) of
<code>ggplot</code> objects which can each be modified
individually.</li>
<li>Once you‚Äôve made your modifications, you can then bind this list of
plots back together with
<code>patchwork::wrap_plots(tracks_list, ncol = 1)</code>.</li>
</ul>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trk_plot_list</span> <span class="op">&lt;-</span> <span class="fu">echoplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">plot_locus</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">dat</span>, </span>
<span>                                      LD_matrix <span class="op">=</span> <span class="va">LD_matrix</span>, </span>
<span>                                      LD_reference <span class="op">=</span> <span class="va">LD_reference</span>,</span>
<span>                                      locus_dir <span class="op">=</span> <span class="va">locus_dir</span>,</span>
<span>                                      zoom <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>                                      show_plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                      return_list <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  </span></code></pre></div>
<pre><code><span><span class="co">## +-------- Locus Plot:  BST1 --------+</span></span></code></pre>
<pre><code><span><span class="co">## + support_thresh = 2</span></span></code></pre>
<pre><code><span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span></code></pre>
<pre><code><span><span class="co">## + 4 fine-mapping methods used.</span></span></code></pre>
<pre><code><span><span class="co">## + 7 Credible Set SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + 3 Consensus SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in CS cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in PP cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## LD_matrix detected. Coloring SNPs by LD with lead SNP.</span></span></code></pre>
<pre><code><span><span class="co">## Filling r/r2 NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: GWAS full window track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: GWAS track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Merged fine-mapping track</span></span></code></pre>
<pre><code><span><span class="co">## Melting PP and CS from 5 fine-mapping methods.</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Adding Gene model track.</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## max_transcripts= 1 .</span></span></code></pre>
<pre><code><span><span class="co">## 16  transcripts from  16  genes returned.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: pals</span></span></code></pre>
<pre><code><span><span class="co">## Fetching data...OK</span></span>
<span><span class="co">## Parsing exons...OK</span></span>
<span><span class="co">## Defining introns...OK</span></span>
<span><span class="co">## Defining UTRs...OK</span></span>
<span><span class="co">## Defining CDS...OK</span></span>
<span><span class="co">## aggregating...</span></span>
<span><span class="co">## Done</span></span>
<span><span class="co">## Constructing graphics...</span></span>
<span><span class="co">## Loading required namespace: pals</span></span>
<span><span class="co">## + Adding vertical lines to highlight SNP groups...</span></span>
<span><span class="co">## +&gt;+&gt;+&gt;+&gt;+ zoom =  10x  +&lt;+&lt;+&lt;+&lt;+</span></span>
<span><span class="co">## + echoplot:: Get window suffix...</span></span>
<span><span class="co">## + Constructing zoom polygon...</span></span>
<span><span class="co">## + Highlighting zoom origin...</span></span>
<span><span class="co">## + Removing subplot margins...</span></span>
<span><span class="co">## + Reordering tracks...</span></span></code></pre>
<pre><code><span><span class="co">## [1] "+ Ensuring last track shows genomic units..."</span></span></code></pre>
<pre><code><span><span class="co">## + Aligning xlimits for each subplot...</span></span>
<span><span class="co">## + Checking track heights...</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">view1_list</span> <span class="op">&lt;-</span> <span class="va">trk_plot_list</span><span class="op">[[</span><span class="st">"10x"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">view1_list</span><span class="op">)</span> <span class="co"># Get track names from a particular zoom view</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GWAS full window" "zoom_polygon"     "Genes"            "GWAS"            </span></span>
<span><span class="co">## [5] "Fine-mapping"</span></span></code></pre>
<p>Modify a specific tracks within a view.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="co"># Modify your selected track</span></span>
<span><span class="va">modified_track</span> <span class="op">&lt;-</span> <span class="va">view1_list</span><span class="op">$</span><span class="va">GWAS</span> <span class="op">+</span> </span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Modified GWAS"</span><span class="op">)</span> <span class="op">+</span>  </span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"purple"</span>, </span>
<span>                                   mid <span class="op">=</span> <span class="st">"magenta"</span>,</span>
<span>                                   high <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>                                   midpoint <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span>,</span>
<span>                   panel.background <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Scale for 'colour' is already present. Adding another scale for 'colour',</span></span>
<span><span class="co">## which will replace the existing scale.</span></span></code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Put it back into your track list</span></span>
<span><span class="va">view1_list</span><span class="op">[[</span><span class="st">"GWAS"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">modified_track</span></span>
<span><span class="co"># Remove a plot you don't want</span></span>
<span><span class="va">view1_list</span><span class="op">[[</span><span class="st">"Genes"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="co"># Specify the relative heights of each track (make sure it matches your new # of plots!)</span></span>
<span><span class="va">track_heights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.3</span>,<span class="fl">.1</span>,<span class="fl">.3</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Bind them back together and plot</span></span>
<span><span class="va">fused_plot</span> <span class="op">&lt;-</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">view1_list</span>, </span>
<span>                                    heights <span class="op">=</span> <span class="va">track_heights</span>,</span>
<span>                                    ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">fused_plot</span><span class="op">)</span></span></code></pre></div>
<p><img src="plot_locus_files/figure-html/modify%20track-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="using-xgr-annotations">Using XGR annotations<a class="anchor" aria-label="anchor" href="#using-xgr-annotations"></a>
</h3>
<ul>
<li>Whenever you use annotation arguments
(e.g.¬†<code>xgr_libnames</code>,<code>roadmap</code>,<code>nott_epigenome</code>)
the annotations that overlap with your locus will automatically be saved
as <code>GRanges</code> objects in a locus-specific subdirectory:<br><em>results/<dataset_type>/<dataset_name>/<locus>/annotation</locus></dataset_name></dataset_type></em>
</li>
<li>If a selected annotation has previously been downloaded and stored
for that locus, <code><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">echoplot::plot_locus()</a></code> will automatically
detect and import it to save time.</li>
</ul>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trk_plot.xgr</span> <span class="op">&lt;-</span> <span class="fu">echoplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">plot_locus</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">dat</span>, </span>
<span>                                     LD_matrix <span class="op">=</span> <span class="va">LD_matrix</span>, </span>
<span>                                     LD_reference <span class="op">=</span> <span class="va">LD_reference</span>,</span>
<span>                                     locus_dir <span class="op">=</span> <span class="va">locus_dir</span>,  </span>
<span>                                     xgr_libnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENCODE_TFBS_ClusteredV3_CellTypes"</span><span class="op">)</span>, </span>
<span>                                     zoom <span class="op">=</span> <span class="st">"10x"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## +-------- Locus Plot:  BST1 --------+</span></span></code></pre>
<pre><code><span><span class="co">## + support_thresh = 2</span></span></code></pre>
<pre><code><span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span></code></pre>
<pre><code><span><span class="co">## + 4 fine-mapping methods used.</span></span></code></pre>
<pre><code><span><span class="co">## + 7 Credible Set SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + 3 Consensus SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in CS cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in PP cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## LD_matrix detected. Coloring SNPs by LD with lead SNP.</span></span></code></pre>
<pre><code><span><span class="co">## Filling r/r2 NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: GWAS full window track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: GWAS track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Merged fine-mapping track</span></span></code></pre>
<pre><code><span><span class="co">## Melting PP and CS from 5 fine-mapping methods.</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Adding Gene model track.</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## max_transcripts= 1 .</span></span></code></pre>
<pre><code><span><span class="co">## 16  transcripts from  16  genes returned.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: pals</span></span></code></pre>
<pre><code><span><span class="co">## Fetching data...OK</span></span>
<span><span class="co">## Parsing exons...OK</span></span>
<span><span class="co">## Defining introns...OK</span></span>
<span><span class="co">## Defining UTRs...OK</span></span>
<span><span class="co">## Defining CDS...OK</span></span>
<span><span class="co">## aggregating...</span></span>
<span><span class="co">## Done</span></span>
<span><span class="co">## Constructing graphics...</span></span>
<span><span class="co">## Loading required namespace: pals</span></span>
<span><span class="co">## echoannot:: Plotting XGR annotations.</span></span>
<span><span class="co">## Start at 2022-11-06 11:09:46</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## 'ENCODE_TFBS_ClusteredV3_CellTypes' (from http://galahad.well.ox.ac.uk/bigdata/ENCODE_TFBS_ClusteredV3_CellTypes.RData) has been loaded into the working environment (at 2022-11-06 11:09:56)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## End at 2022-11-06 11:09:56</span></span>
<span><span class="co">## Runtime in total is: 10 secs</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Converting dat to GRanges object.</span></span>
<span><span class="co">## 1,579 query SNP(s) detected with reference overlap.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Ignoring unknown parameters: facets</span></span></code></pre>
<pre><code><span><span class="co">## Warning in max(xlim): no non-missing arguments to max; returning -Inf</span></span></code></pre>
<pre><code><span><span class="co">## + Adding vertical lines to highlight SNP groups...</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Groups with fewer than two data points have been dropped.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Groups with fewer than two data points have been dropped.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 2 rows containing missing values (position_stack).</span></span></code></pre>
<pre><code><span><span class="co">## +&gt;+&gt;+&gt;+&gt;+ zoom =  10x  +&lt;+&lt;+&lt;+&lt;+</span></span>
<span><span class="co">## + echoplot:: Get window suffix...</span></span>
<span><span class="co">## + Constructing zoom polygon...</span></span>
<span><span class="co">## + Highlighting zoom origin...</span></span>
<span><span class="co">## + Removing subplot margins...</span></span>
<span><span class="co">## + Reordering tracks...</span></span></code></pre>
<pre><code><span><span class="co">## [1] "+ Ensuring last track shows genomic units..."</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Groups with fewer than two data points have been dropped.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Groups with fewer than two data points have been dropped.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 2 rows containing missing values (position_stack).</span></span></code></pre>
<pre><code><span><span class="co">## + Aligning xlimits for each subplot...</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Groups with fewer than two data points have been dropped.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Groups with fewer than two data points have been dropped.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 2 rows containing missing values (position_stack).</span></span></code></pre>
<pre><code><span><span class="co">## + Checking track heights...</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Groups with fewer than two data points have been dropped.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Groups with fewer than two data points have been dropped.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 2 rows containing missing values (position_stack).</span></span></code></pre>
<p><img src="plot_locus_files/figure-html/trk_plot.xgr-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="using-roadmap-annotations">Using <a href="http://www.roadmapepigenomics.org/data/tables/brain" class="external-link">Roadmap</a>
annotations<a class="anchor" aria-label="anchor" href="#using-roadmap-annotations"></a>
</h3>
<ul>
<li>Using the <code>Roadmap=T</code> and
<code>Roadmap_query="&lt;query&gt;"</code> arguments searches the
Roadmap for chromatin mark data across various cell-types, cell-lines
and tissues.<br>
</li>
<li>Note that Roadmap queries requires <code>tabix</code> to be
installed on your machine, or within a conda environment
(<code>conda_env = "echoR"</code>).</li>
<li>Parallelizing these queries across multiple threads speeds up this
process (<code>nThread=&lt;n_cores_available&gt;</code>), as does
reusing previously stored data which is automatically saved to the
locus-specific subfolder
(<code>&lt;dataset_type&gt;/&lt;dataset_name&gt;/&lt;locus&gt;/annotations/Roadmap.ChromatinMarks_CellTypes.RDS</code>).</li>
</ul>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trk_plot.roadmap</span> <span class="op">&lt;-</span> <span class="fu">echoplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">plot_locus</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">dat</span>, </span>
<span>                                         LD_matrix <span class="op">=</span> <span class="va">LD_matrix</span>, </span>
<span>                                         LD_reference <span class="op">=</span> <span class="va">LD_reference</span>,</span>
<span>                                         locus_dir <span class="op">=</span> <span class="va">locus_dir</span>,</span>
<span>                                         zoom <span class="op">=</span> <span class="st">"5x"</span>,</span>
<span>                                         </span>
<span>                                         roadmap <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                         roadmap_query <span class="op">=</span> <span class="st">"monocyte"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="using-nott_2019-annotations">Using <a href="https://science.sciencemag.org/content/366/6469/1134.abstract" class="external-link">Nott_2019</a>
annotations<a class="anchor" aria-label="anchor" href="#using-nott_2019-annotations"></a>
</h3>
<ul>
<li>Query and plot brain cell type-specific epigenomic assays from <a href="https://science.sciencemag.org/content/366/6469/1134.abstract" class="external-link">Nott
et al.¬†(Science, 2019)</a><br>
(see <code>?NOTT_2019.bigwig_metadata</code> for details).</li>
</ul>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trk_plot.nott_2019</span> <span class="op">&lt;-</span> <span class="fu">echoplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echoplot/man/plot_locus.html" class="external-link">plot_locus</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">dat</span>,  </span>
<span>                                 LD_matrix <span class="op">=</span> <span class="va">LD_matrix</span>, </span>
<span>                                 LD_reference <span class="op">=</span> <span class="va">LD_reference</span>,</span>
<span>                                 locus_dir <span class="op">=</span> <span class="va">locus_dir</span>, </span>
<span>                                 zoom <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>                                 </span>
<span>                                 nott_epigenome <span class="op">=</span> <span class="cn">TRUE</span>,  </span>
<span>                                 nott_binwidth <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                                 nott_regulatory_rects <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                 nott_show_placseq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="using-qtl-datasets">Using QTL datasets<a class="anchor" aria-label="anchor" href="#using-qtl-datasets"></a>
</h3>
<ul>
<li>Plot multiple QTL p-value columns (or really P-value columns from
any kind of datset).<br>
</li>
<li>Each QTL dataset will be plotted as a new track.</li>
</ul>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat1</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/copy.html" class="external-link">copy</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span>
<span><span class="va">dat2</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/copy.html" class="external-link">copy</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span>
<span><span class="co"># Make fake QTL P-values for the sake a demonstration</span></span>
<span><span class="va">dat1</span><span class="op">$</span><span class="va">P</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/jitter.html" class="external-link">jitter</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">$</span><span class="va">P</span>, amount <span class="op">=</span> <span class="fl">1e-15</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dat2</span><span class="op">$</span><span class="va">P</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/jitter.html" class="external-link">jitter</a></span><span class="op">(</span><span class="va">dat2</span><span class="op">$</span><span class="va">P</span>, amount <span class="op">=</span> <span class="fl">1e-16</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"fake_eQTL"</span><span class="op">=</span><span class="va">dat1</span>, </span>
<span>                 <span class="st">"fake_sQTL"</span><span class="op">=</span><span class="va">dat2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">trk_plot.qtl</span> <span class="op">&lt;-</span> <span class="fu">echoplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echoplot/man/plot_locus_multi.html" class="external-link">plot_locus_multi</a></span><span class="op">(</span>dat_ls <span class="op">=</span> <span class="va">dat_list</span>, </span>
<span>                                           LD_ls <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">LD_matrix</span>,<span class="va">LD_matrix</span><span class="op">)</span>,  </span>
<span>                                           locus_dir <span class="op">=</span> <span class="va">locus_dir</span>,</span>
<span>                                           zoom <span class="op">=</span> <span class="st">"10x"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## LD_matrix detected. Coloring SNPs by LD with lead SNP.</span></span></code></pre>
<pre><code><span><span class="co">## Filling r/r2 NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## LD_matrix detected. Coloring SNPs by LD with lead SNP.</span></span></code></pre>
<pre><code><span><span class="co">## Filling r/r2 NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## +-------- Locus Plot:  BST1 --------+</span></span></code></pre>
<pre><code><span><span class="co">## + support_thresh = 2</span></span></code></pre>
<pre><code><span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span></code></pre>
<pre><code><span><span class="co">## + 4 fine-mapping methods used.</span></span></code></pre>
<pre><code><span><span class="co">## + 14 Credible Set SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + 6 Consensus SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in CS cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in PP cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: GWAS full window track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: GWAS track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Merged fine-mapping track</span></span></code></pre>
<pre><code><span><span class="co">## Melting PP and CS from 5 fine-mapping methods.</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Adding Gene model track.</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## max_transcripts= 1 .</span></span></code></pre>
<pre><code><span><span class="co">## 16  transcripts from  16  genes returned.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: pals</span></span></code></pre>
<pre><code><span><span class="co">## Fetching data...OK</span></span>
<span><span class="co">## Parsing exons...OK</span></span>
<span><span class="co">## Defining introns...OK</span></span>
<span><span class="co">## Defining UTRs...OK</span></span>
<span><span class="co">## Defining CDS...OK</span></span>
<span><span class="co">## aggregating...</span></span>
<span><span class="co">## Done</span></span>
<span><span class="co">## Constructing graphics...</span></span>
<span><span class="co">## Loading required namespace: pals</span></span>
<span><span class="co">## + Adding vertical lines to highlight SNP groups...</span></span>
<span><span class="co">## +&gt;+&gt;+&gt;+&gt;+ zoom =  10x  +&lt;+&lt;+&lt;+&lt;+</span></span>
<span><span class="co">## + echoplot:: Get window suffix...</span></span>
<span><span class="co">## + Constructing zoom polygon...</span></span>
<span><span class="co">## + Highlighting zoom origin...</span></span>
<span><span class="co">## + Removing subplot margins...</span></span>
<span><span class="co">## + Reordering tracks...</span></span></code></pre>
<pre><code><span><span class="co">## [1] "+ Ensuring last track shows genomic units..."</span></span></code></pre>
<pre><code><span><span class="co">## + Aligning xlimits for each subplot...</span></span>
<span><span class="co">## + Checking track heights...</span></span></code></pre>
<p><img src="plot_locus_files/figure-html/trk_plot.QTL-1.png" width="700"></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre style="max-height: 400px;"><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.5 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] patchwork_1.1.2   ggplot2_3.3.6     echolocatoR_2.0.2 BiocStyle_2.26.0 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rappdirs_0.3.3              rtracklayer_1.58.0         </span></span>
<span><span class="co">##   [3] GGally_2.1.2                R.methodsS3_1.8.2          </span></span>
<span><span class="co">##   [5] ragg_1.2.4                  tidyr_1.2.1                </span></span>
<span><span class="co">##   [7] echoLD_0.99.8               bit64_4.0.5                </span></span>
<span><span class="co">##   [9] knitr_1.40                  irlba_2.3.5.1              </span></span>
<span><span class="co">##  [11] DelayedArray_0.24.0         R.utils_2.12.1             </span></span>
<span><span class="co">##  [13] data.table_1.14.4           rpart_4.1.19               </span></span>
<span><span class="co">##  [15] KEGGREST_1.38.0             RCurl_1.98-1.9             </span></span>
<span><span class="co">##  [17] AnnotationFilter_1.21.0     generics_0.1.3             </span></span>
<span><span class="co">##  [19] BiocGenerics_0.44.0         GenomicFeatures_1.50.2     </span></span>
<span><span class="co">##  [21] RSQLite_2.2.18              proxy_0.4-27               </span></span>
<span><span class="co">##  [23] bit_4.0.4                   tzdb_0.3.0                 </span></span>
<span><span class="co">##  [25] xml2_1.3.3                  SummarizedExperiment_1.28.0</span></span>
<span><span class="co">##  [27] assertthat_0.2.1            viridis_0.6.2              </span></span>
<span><span class="co">##  [29] xfun_0.34                   hms_1.1.2                  </span></span>
<span><span class="co">##  [31] jquerylib_0.1.4             evaluate_0.17              </span></span>
<span><span class="co">##  [33] fansi_1.0.3                 restfulr_0.0.15            </span></span>
<span><span class="co">##  [35] progress_1.2.2              dbplyr_2.2.1               </span></span>
<span><span class="co">##  [37] readxl_1.4.1                Rgraphviz_2.41.2           </span></span>
<span><span class="co">##  [39] igraph_1.3.5                DBI_1.1.3                  </span></span>
<span><span class="co">##  [41] htmlwidgets_1.5.4           reshape_0.8.9              </span></span>
<span><span class="co">##  [43] downloadR_0.99.5            stats4_4.2.1               </span></span>
<span><span class="co">##  [45] purrr_0.3.5                 ellipsis_0.3.2             </span></span>
<span><span class="co">##  [47] dplyr_1.0.10                backports_1.4.1            </span></span>
<span><span class="co">##  [49] bookdown_0.29               biomaRt_2.54.0             </span></span>
<span><span class="co">##  [51] deldir_1.0-6                MatrixGenerics_1.10.0      </span></span>
<span><span class="co">##  [53] vctrs_0.5.0                 Biobase_2.58.0             </span></span>
<span><span class="co">##  [55] ensembldb_2.21.5            cachem_1.0.6               </span></span>
<span><span class="co">##  [57] withr_2.5.0                 BSgenome_1.66.1            </span></span>
<span><span class="co">##  [59] checkmate_2.1.0             GenomicAlignments_1.34.0   </span></span>
<span><span class="co">##  [61] prettyunits_1.1.1           cluster_2.1.4              </span></span>
<span><span class="co">##  [63] ape_5.6-2                   dir.expiry_1.5.1           </span></span>
<span><span class="co">##  [65] lazyeval_0.2.2              crayon_1.5.2               </span></span>
<span><span class="co">##  [67] basilisk.utils_1.9.4        crul_1.3                   </span></span>
<span><span class="co">##  [69] pkgconfig_2.0.3             labeling_0.4.2             </span></span>
<span><span class="co">##  [71] GenomeInfoDb_1.34.1         nlme_3.1-160               </span></span>
<span><span class="co">##  [73] ProtGenerics_1.29.1         XGR_1.1.8                  </span></span>
<span><span class="co">##  [75] nnet_7.3-18                 rlang_1.0.6                </span></span>
<span><span class="co">##  [77] lifecycle_1.0.3             filelock_1.0.2             </span></span>
<span><span class="co">##  [79] httpcode_0.3.0              BiocFileCache_2.6.0        </span></span>
<span><span class="co">##  [81] echotabix_0.99.8            dichromat_2.0-0.1          </span></span>
<span><span class="co">##  [83] cellranger_1.1.0            coloc_5.1.0.1              </span></span>
<span><span class="co">##  [85] rprojroot_2.0.3             matrixStats_0.62.0         </span></span>
<span><span class="co">##  [87] graph_1.75.1                Matrix_1.5-1               </span></span>
<span><span class="co">##  [89] osfr_0.2.9                  boot_1.3-28                </span></span>
<span><span class="co">##  [91] base64enc_0.1-3             png_0.1-7                  </span></span>
<span><span class="co">##  [93] viridisLite_0.4.1           rjson_0.2.21               </span></span>
<span><span class="co">##  [95] rootSolve_1.8.2.3           bitops_1.0-7               </span></span>
<span><span class="co">##  [97] R.oo_1.25.0                 ggnetwork_0.5.10           </span></span>
<span><span class="co">##  [99] Biostrings_2.66.0           blob_1.2.3                 </span></span>
<span><span class="co">## [101] mixsqp_0.3-43               stringr_1.4.1              </span></span>
<span><span class="co">## [103] echoplot_0.99.5             dnet_1.1.7                 </span></span>
<span><span class="co">## [105] readr_2.1.3                 jpeg_0.1-9                 </span></span>
<span><span class="co">## [107] S4Vectors_0.36.0            echodata_0.99.15           </span></span>
<span><span class="co">## [109] scales_1.2.1                memoise_2.0.1              </span></span>
<span><span class="co">## [111] magrittr_2.0.3              plyr_1.8.7                 </span></span>
<span><span class="co">## [113] hexbin_1.28.2               zlibbioc_1.44.0            </span></span>
<span><span class="co">## [115] compiler_4.2.1              echoconda_0.99.8           </span></span>
<span><span class="co">## [117] BiocIO_1.8.0                RColorBrewer_1.1-3         </span></span>
<span><span class="co">## [119] catalogueR_1.0.0            EnsDb.Hsapiens.v75_2.99.0  </span></span>
<span><span class="co">## [121] Rsamtools_2.14.0            cli_3.4.1                  </span></span>
<span><span class="co">## [123] XVector_0.38.0              echoannot_0.99.10          </span></span>
<span><span class="co">## [125] htmlTable_2.4.1             Formula_1.2-4              </span></span>
<span><span class="co">## [127] MASS_7.3-58.1               tidyselect_1.2.0           </span></span>
<span><span class="co">## [129] stringi_1.7.8               textshaping_0.3.6          </span></span>
<span><span class="co">## [131] highr_0.9                   yaml_2.3.6                 </span></span>
<span><span class="co">## [133] supraHex_1.35.0             latticeExtra_0.6-30        </span></span>
<span><span class="co">## [135] ggrepel_0.9.1               grid_4.2.1                 </span></span>
<span><span class="co">## [137] sass_0.4.2                  VariantAnnotation_1.44.0   </span></span>
<span><span class="co">## [139] tools_4.2.1                 lmom_2.9                   </span></span>
<span><span class="co">## [141] parallel_4.2.1              rstudioapi_0.14            </span></span>
<span><span class="co">## [143] foreign_0.8-83              piggyback_0.1.4            </span></span>
<span><span class="co">## [145] gridExtra_2.3               gld_2.6.6                  </span></span>
<span><span class="co">## [147] farver_2.1.1                digest_0.6.30              </span></span>
<span><span class="co">## [149] snpStats_1.47.1             BiocManager_1.30.19        </span></span>
<span><span class="co">## [151] Rcpp_1.0.9                  GenomicRanges_1.50.0       </span></span>
<span><span class="co">## [153] OrganismDbi_1.39.1          httr_1.4.4                 </span></span>
<span><span class="co">## [155] AnnotationDbi_1.60.0        RCircos_1.2.2              </span></span>
<span><span class="co">## [157] ggbio_1.45.0                biovizBase_1.45.0          </span></span>
<span><span class="co">## [159] colorspace_2.0-3            XML_3.99-0.12              </span></span>
<span><span class="co">## [161] fs_1.5.2                    reticulate_1.26            </span></span>
<span><span class="co">## [163] IRanges_2.32.0              splines_4.2.1              </span></span>
<span><span class="co">## [165] RBGL_1.73.0                 expm_0.999-6               </span></span>
<span><span class="co">## [167] pkgdown_2.0.6.9000          echofinemap_0.99.4         </span></span>
<span><span class="co">## [169] basilisk_1.9.12             Exact_3.2                  </span></span>
<span><span class="co">## [171] systemfonts_1.0.4           jsonlite_1.8.3             </span></span>
<span><span class="co">## [173] susieR_0.12.27              R6_2.5.1                   </span></span>
<span><span class="co">## [175] Hmisc_4.7-1                 pillar_1.8.1               </span></span>
<span><span class="co">## [177] htmltools_0.5.3             glue_1.6.2                 </span></span>
<span><span class="co">## [179] fastmap_1.1.0               DT_0.26                    </span></span>
<span><span class="co">## [181] BiocParallel_1.32.0         class_7.3-20               </span></span>
<span><span class="co">## [183] codetools_0.2-18            mvtnorm_1.1-3              </span></span>
<span><span class="co">## [185] utf8_1.2.2                  lattice_0.20-45            </span></span>
<span><span class="co">## [187] bslib_0.4.0                 tibble_3.1.8               </span></span>
<span><span class="co">## [189] curl_4.3.3                  DescTools_0.99.47          </span></span>
<span><span class="co">## [191] zip_2.2.2                   openxlsx_4.2.5.1           </span></span>
<span><span class="co">## [193] interp_1.1-3                survival_3.4-0             </span></span>
<span><span class="co">## [195] rmarkdown_2.17              desc_1.4.2                 </span></span>
<span><span class="co">## [197] munsell_0.5.0               e1071_1.7-12               </span></span>
<span><span class="co">## [199] GenomeInfoDbData_1.2.9      reshape2_1.4.4             </span></span>
<span><span class="co">## [201] gtable_0.3.1</span></span></code></pre>
<details><p><br></p>
</details></details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
