<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>QTLs • echolocatoR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="QTLs">
<meta property="og:description" content="echolocatoR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">echolocatoR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/echolocatoR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/QTLs.html">QTLs</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="../articles/summarise.html">Summarise</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RajLabMSSM/echolocatoR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>QTLs</h1>
                        <h4 data-toc-skip class="author"><h4> ¶ Author:
<i>Brian M. Schilder</i> ¶ </h4></h4>
            
            <h4 data-toc-skip class="date"><h4> ¶ Updated:
<i>Jan-20-2023</i> ¶ </h4></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/RajLabMSSM/echolocatoR/blob/HEAD/vignettes/QTLs.Rmd" class="external-link"><code>vignettes/QTLs.Rmd</code></a></small>
      <div class="hidden name"><code>QTLs.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RajLabMSSM/echolocatoR" class="external-link">echolocatoR</a></span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">##   method from   </span></span>
<span><span class="co">##   +.gg   ggplot2</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">⠊⠉⠡⣀⣀⠊⠉⠡⣀⣀⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⢁⡐⠉⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠉⢂⡈⠑⣀⠉⢄⡈⠡⣀</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⡈⡐⢂⢁⠒⡈⡐⢂⢁⠒⡈⡐⢂⢁⠑⡈⡈⢄⢁⠡⠌⡈⠤⢁⠡⠌⡈⠤⢁⠡⠌⡈⡠⢁⢁⠊⡈⡐⢂</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="color: #55FFFF; font-weight: bold;">🦇  🦇  🦇 e c h o l o c a t o R 🦇  🦇  🦇</span> <span style="color: #00BBBB;">─────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="color: #55FFFF; font-weight: bold;">v2.0.3</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">⠌⡈⡐⢂⢁⠒⡈⡐⢂⢁⠒⡈⡐⢂⢁⠑⡈⡈⢄⢁⠡⠌⡈⠤⢁⠡⠌⡈⠤⢁⠡⠌⡈⡠⢁⢁⠊⡈⡐⢂</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⢁⡐⠉⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠉⢂⡈⠑⣀⠉⢄⡈⠡⣀</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠊⠉⠡⣀⣀⠊⠉⠡⣀⣀⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠</span>                                    </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ If you use </span><span style="color: #00BBBB; font-weight: bold;">echolocatoR</span><span style="color: #00BBBB;"> or any of the </span><span style="color: #00BBBB; font-style: italic;">echoverse</span><span style="color: #00BBBB;"> subpackages, please cite:</span>      </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶ Brian M Schilder, Jack Humphrey, Towfique</span>                                </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     Raj (2021) echolocatoR: an automated</span>                                       </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     end-to-end statistical and functional</span>                                      </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     genomic fine-mapping pipeline,</span>                                             </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     Bioinformatics; btab658,</span>                                                   </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://doi.org/10.1093/bioinformatics/btab658</span>                             </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ Please report any bugs/feature requests on GitHub:</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://github.com/RajLabMSSM/echolocatoR/issues</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ Contributions are welcome!:</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://github.com/RajLabMSSM/echolocatoR/pulls</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<div class="section level2">
<h2 id="qtl-pipeline">QTL pipeline<a class="anchor" aria-label="anchor" href="#qtl-pipeline"></a>
</h2>
<ul>
<li>Here, we will use GWAS-eQTL colocalization results provided via the
<a href="https://github.com/RajLabMSSM/Fine_Mapping_Shiny" class="external-link"><em>echolocatoR
Fine-mapping Portal</em></a>, from the publication:</li>
</ul>
<blockquote>
<p>Lopes, K.d.P., Snijders, G.J.L., Humphrey, J. et al. Genetic analysis
of the human microglial transcriptome across brain regions, aging and
disease pathologies. Nat Genet 54, 4–17 (2022). <a href="https://doi.org/10.1038/s41588-021-00976-y" class="external-link">https://doi.org/10.1038/s41588-021-00976-y</a></p>
</blockquote>
<div class="section level3">
<h3 id="import-qtl-data">Import QTL data<a class="anchor" aria-label="anchor" href="#import-qtl-data"></a>
</h3>
<p>This data is actually merged GWAS-QTL colocalization results, but it
contains all of the necessary columns from the original eQTL summary
stats that we need to perform eQTL fine-mapping.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coloc_res</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/get_Kunkle2019_coloc.html" class="external-link">get_Kunkle2019_coloc</a></span><span class="op">(</span>return_path <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="prepare-colmap">Prepare <code>colmap</code><a class="anchor" aria-label="anchor" href="#prepare-colmap"></a>
</h3>
<p>Prepare a column mapping object for the summary statistics. We’ll
reuse this for both the <code>import_topSNPs</code> and
<code>finemap_loci</code> steps.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colmap</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/construct_colmap.html" class="external-link">construct_colmap</a></span><span class="op">(</span> </span>
<span>      CHR <span class="op">=</span> <span class="st">"chr"</span>,</span>
<span>      POS <span class="op">=</span> <span class="st">"pos"</span>,</span>
<span>      N <span class="op">=</span> <span class="st">"qtl.N"</span>,</span>
<span>      SNP <span class="op">=</span> <span class="st">"snp"</span>,</span>
<span>      P <span class="op">=</span> <span class="st">"qtl.pvalues"</span>,</span>
<span>      Effect <span class="op">=</span> <span class="st">"qtl.beta"</span>,</span>
<span>      StdErr <span class="op">=</span> <span class="st">"qtl.varbeta"</span>,</span>
<span>      MAF <span class="op">=</span> <span class="st">"qtl.MAF"</span>,</span>
<span>      Locus <span class="op">=</span> <span class="st">"Locus"</span>,</span>
<span>      Gene <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="prepare-top_snps-data-frame">Prepare <code>top_SNPs</code> data.frame<a class="anchor" aria-label="anchor" href="#prepare-top_snps-data-frame"></a>
</h3>
<ul>
<li>In this case, we don’t have a top SNPs file ready. So we’re just
going to make one directly from the full summary stats file itself
(<em>NOTE</em>: You can only use this approach if you can fit the entire
file in memory).<br>
</li>
<li>In this case, you’ll want to make sure to set
<code>grouping_vars=c("Locus","Gene")</code> so that you get top SNPs
for each eGene-locus pair (not just one SNP per locus).</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">topSNPs</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/import_topSNPs.html" class="external-link">import_topSNPs</a></span><span class="op">(</span></span>
<span>  topSS <span class="op">=</span> <span class="va">coloc_res</span><span class="op">$</span><span class="va">path</span>,</span>
<span>  colmap <span class="op">=</span> <span class="va">colmap</span>,</span>
<span>  <span class="co">## Important for QTLs: group by both Locus and Gene</span></span>
<span>  grouping_vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Locus"</span>,<span class="st">"Gene"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required namespace: MungeSumstats</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: chr ==&gt; CHR</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: pos ==&gt; POS</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: snp ==&gt; SNP</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: qtl.pvalues ==&gt; P</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: qtl.beta ==&gt; Effect</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: qtl.varbeta ==&gt; StdErr</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: qtl.MAF ==&gt; MAF</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: gene ==&gt; Gene</span></span></code></pre>
<pre><code><span><span class="co">## Renaming column: qtl.N ==&gt; N</span></span></code></pre>
<pre><code><span><span class="co">## [1] "+ Assigning Gene and Locus independently."</span></span></code></pre>
<pre><code><span><span class="co">## Standardising column headers.</span></span></code></pre>
<pre><code><span><span class="co">## First line of summary statistics file:</span></span></code></pre>
<pre><code><span><span class="co">## Locus    SNP V.df1   z.df1   r.df1   lABF.df1    V.df2   z.df2   r.df2   lABF.df2    internal.sum.lABF   SNP.PP.H4   gwas.pvalues    gwas.beta   gwas.varbeta    gwas.MAF    CHR POS A1  A2  gwas.N  gwas.type   s   Gene    P   Effect  StdErr  MAF QTL_chr QTL_pos N   qtl.type    N_cases N_controls  </span></span></code></pre>
<pre><code><span><span class="co">## Returning unmapped column names without making them uppercase.</span></span></code></pre>
<pre><code><span><span class="co">## + Mapping colnames from MungeSumstats ==&gt; echolocatoR</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">topSNPs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                     Locus        SNP      V.df1      z.df1     r.df1   lABF.df1</span></span>
<span><span class="co">## 1:  ABCA7_ENSG00000160953 rs76951864 0.00227529 -1.2138365 0.9461792 -0.7639979</span></span>
<span><span class="co">## 2:   BIN1_ENSG00000136731  rs4663110 0.00021904  2.0405405 0.9945538 -0.5358563</span></span>
<span><span class="co">## 3:    CR1_ENSG00000266094  rs1518110 0.00030276 -1.0977011 0.9924879 -1.8476694</span></span>
<span><span class="co">## 4: INPP5D_ENSG00000168918  rs1881492 0.00036100  0.1631579 0.9910557 -2.3451794</span></span>
<span><span class="co">## 5: MS4A6A_ENSG00000149476  rs6591611 0.00028900 -1.4176471 0.9928268 -1.4710500</span></span>
<span><span class="co">## 6:  PILRA_ENSG00000106366  rs2227631 0.00020736 -0.1041667 0.9948427 -2.6282771</span></span>
<span><span class="co">##          V.df2     z.df2     r.df2   lABF.df2 internal.sum.lABF    SNP.PP.H4</span></span>
<span><span class="co">## 1: 0.007802613 -5.826752 0.2374864  3.8958871         3.1318892 6.231637e-08</span></span>
<span><span class="co">## 2: 0.004149176  2.778507 0.4321435  1.3851521         0.8492958 1.013056e-25</span></span>
<span><span class="co">## 3: 0.002913938 -3.976592 0.4592553  3.3237628         1.4760934 4.495411e-11</span></span>
<span><span class="co">## 4: 0.012164546 -2.654756 0.1470877  0.4387679        -1.9064116 3.587352e-08</span></span>
<span><span class="co">## 5: 0.015193274  2.675112 0.2112492  0.6372210        -0.8338290 8.201862e-13</span></span>
<span><span class="co">## 6: 0.006858275  8.459513 0.5879401 20.5941901        17.9659130 9.907667e-01</span></span>
<span><span class="co">##    gwas.pvalues gwas.beta gwas.varbeta gwas.MAF CHR       POS A1 A2 gwas.N</span></span>
<span><span class="co">## 1:      0.22500   -0.0579   0.00227529   0.1054  19    374604  A  G  94437</span></span>
<span><span class="co">## 2:      0.04153    0.0302   0.00021904   0.6262   2 127922463  T  C  94437</span></span>
<span><span class="co">## 3:      0.27270   -0.0191   0.00030276   0.7803   1 206944861  A  C  94437</span></span>
<span><span class="co">## 4:      0.87100    0.0031   0.00036100   0.7763   2 233406998  T  G  94437</span></span>
<span><span class="co">## 5:      0.15760   -0.0241   0.00028900   0.2207  11  60405172  A  C  94437</span></span>
<span><span class="co">## 6:      0.91860   -0.0015   0.00020736   0.4056   7 100769538  A  G  94437</span></span>
<span><span class="co">##    gwas.type    s            Gene           P    Effect      StdErr    Freq</span></span>
<span><span class="co">## 1:        cc 0.37 ENSG00000160953 8.31563e-09 -0.514691 0.007802613 0.08770</span></span>
<span><span class="co">## 2:        cc 0.37 ENSG00000136731 7.93573e-05  0.178975 0.004149176 0.34385</span></span>
<span><span class="co">## 3:        cc 0.37 ENSG00000266094 1.00516e-04 -0.214660 0.002913938 0.25363</span></span>
<span><span class="co">## 4:        cc 0.37 ENSG00000168918 6.42575e-06 -0.292801 0.012164546 0.14447</span></span>
<span><span class="co">## 5:        cc 0.37 ENSG00000149476 2.20256e-04  0.329737 0.015193274 0.18503</span></span>
<span><span class="co">## 6:        cc 0.37 ENSG00000106366 5.45534e-17  0.700572 0.006858275 0.42796</span></span>
<span><span class="co">##    QTL_chr   QTL_pos  N qtl.type N_cases N_controls</span></span>
<span><span class="co">## 1:   chr19    374604 90    quant   34941      59496</span></span>
<span><span class="co">## 2:    chr2 127164887 90    quant   34941      59496</span></span>
<span><span class="co">## 3:    chr1 206771516 90    quant   34941      59496</span></span>
<span><span class="co">## 4:    chr2 232542288 90    quant   34941      59496</span></span>
<span><span class="co">## 5:   chr11  60637699 90    quant   34941      59496</span></span>
<span><span class="co">## 6:    chr7 101126257 90    quant   34941      59496</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-fine-mapping-pipeline">Run fine-mapping pipeline<a class="anchor" aria-label="anchor" href="#run-fine-mapping-pipeline"></a>
</h3>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">echolocatoR</span><span class="fu">::</span><span class="fu"><a href="../reference/finemap_loci.html">finemap_loci</a></span><span class="op">(</span>fullSS_path <span class="op">=</span> <span class="va">coloc_res</span><span class="op">$</span><span class="va">path</span>,</span>
<span>                                 topSNPs <span class="op">=</span> <span class="va">topSNPs</span>,</span>
<span>                                 <span class="co">## Let's just fine-map 1 locus for demo purposes</span></span>
<span>                                 loci <span class="op">=</span> <span class="va">topSNPs</span><span class="op">$</span><span class="va">Locus</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                                 dataset_name <span class="op">=</span> <span class="st">"Kunkle_2019.microgliaQTL"</span>,</span>
<span>                                 dataset_type <span class="op">=</span> <span class="st">"QTL"</span>,</span>
<span>                                 bp_distance <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>                                 colmap <span class="op">=</span> <span class="va">colmap</span>,</span>
<span>                                 show_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                 finemap_methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ABF"</span>,<span class="st">"FINEMAP"</span>,<span class="st">"SUSIE"</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "+ Assigning Gene and Locus independently."</span></span></code></pre>
<pre><code><span><span class="co">## Standardising column headers.</span></span></code></pre>
<pre><code><span><span class="co">## First line of summary statistics file:</span></span></code></pre>
<pre><code><span><span class="co">## Locus    SNP V.df1   z.df1   r.df1   lABF.df1    V.df2   z.df2   r.df2   lABF.df2    internal.sum.lABF   SNP.PP.H4   gwas.pvalues    gwas.beta   gwas.varbeta    gwas.MAF    CHR POS A1  A2  gwas.N  gwas.type   s   Gene    P   Effect  StdErr  Freq    QTL_chr QTL_pos N   qtl.type    N_cases N_controls  </span></span></code></pre>
<pre><code><span><span class="co">## Returning unmapped column names without making them uppercase.</span></span></code></pre>
<pre><code><span><span class="co">## + Mapping colnames from MungeSumstats ==&gt; echolocatoR</span></span></code></pre>
<pre><code><span><span class="co">## WARNING:: fullSS_genome_build not provided. Assuming 'GRCH37'.</span></span></code></pre>
<pre><code><span><span class="co">## ┌─────────────────────────────────────────────────────────┐</span></span>
<span><span class="co">## │                                                         │</span></span>
<span><span class="co">## │   <span style="color: #FFFFFF; font-weight: 100;">)))&gt;</span><span style="color: #55FFFF;"> 🦇 ABCA7_ENSG00000160953 [locus 1 / 1] 🦇 </span><span style="color: #FFFFFF; font-weight: 100;">&lt;(((</span>   │</span></span>
<span><span class="co">## │                                                         │</span></span>
<span><span class="co">## └─────────────────────────────────────────────────────────┘</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 1 ▶▶▶ Query 🔎</span> <span style="color: #00BBBB;">─────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## + Query Method: tabix</span></span></code></pre>
<pre><code><span><span class="co">## Constructing GRanges query using min/max ranges within a single chromosome.</span></span></code></pre>
<pre><code><span><span class="co">## query_dat is already a GRanges object. Returning directly.</span></span></code></pre>
<pre><code><span><span class="co">## ========= echotabix::convert =========</span></span></code></pre>
<pre><code><span><span class="co">## Converting full summary stats file to tabix format for fast querying.</span></span></code></pre>
<pre><code><span><span class="co">## Inferred format: 'table'</span></span></code></pre>
<pre><code><span><span class="co">## Explicit format: 'table'</span></span></code></pre>
<pre><code><span><span class="co">## Inferring comment_char from tabular header: 'Locus'</span></span></code></pre>
<pre><code><span><span class="co">## Determining chrom type from file header.</span></span></code></pre>
<pre><code><span><span class="co">## Chromosome format: 1</span></span></code></pre>
<pre><code><span><span class="co">## Detecting column delimiter.</span></span></code></pre>
<pre><code><span><span class="co">## Identified column separator: ,</span></span></code></pre>
<pre><code><span><span class="co">## WARNING: Columns must be tab-separated ('\t') in order to be sorted outside of R (which is more memory-efficient). </span></span>
<span><span class="co">## Will instead import full data into R to sort and rewrite to disk.</span></span></code></pre>
<pre><code><span><span class="co">## Sorting rows by coordinates via data.table.</span></span></code></pre>
<pre><code><span><span class="co">## Constructing outputs</span></span></code></pre>
<pre><code><span><span class="co">## Using existing bgzipped file: /github/home/.cache/R/echodata/Microglia_all_regions_Kunkle_2019_COLOC.snp-level_select.tsv.bgz </span></span>
<span><span class="co">## Set force_new=TRUE to override this.</span></span></code></pre>
<pre><code><span><span class="co">## Tabix-indexing file using: Rsamtools</span></span></code></pre>
<pre><code><span><span class="co">## Data successfully converted to bgzip-compressed, tabix-indexed format.</span></span></code></pre>
<pre><code><span><span class="co">## ========= echotabix::query =========</span></span></code></pre>
<pre><code><span><span class="co">## query_dat is already a GRanges object. Returning directly.</span></span></code></pre>
<pre><code><span><span class="co">## Inferred format: 'table'</span></span></code></pre>
<pre><code><span><span class="co">## Querying tabular tabix file using: Rsamtools.</span></span></code></pre>
<pre><code><span><span class="co">## Checking query chromosome style is correct.</span></span></code></pre>
<pre><code><span><span class="co">## Chromosome format: 1</span></span></code></pre>
<pre><code><span><span class="co">## Retrieving data.</span></span></code></pre>
<pre><code><span><span class="co">## Converting query results to data.table.</span></span></code></pre>
<pre><code><span><span class="co">## Processing query: 19:373604-375604</span></span></code></pre>
<pre><code><span><span class="co">## Adding 'query' column to results.</span></span></code></pre>
<pre><code><span><span class="co">## Retrieved data with 10 rows</span></span></code></pre>
<pre><code><span><span class="co">## Saving query ==&gt; /tmp/RtmpVZRnW7/results/QTL/Kunkle_2019.microgliaQTL/ABCA7_ENSG00000160953/ABCA7_ENSG00000160953_Kunkle_2019.microgliaQTL_subset.tsv.gz</span></span></code></pre>
<pre><code><span><span class="co">## + Query: 10 SNPs x 34 columns.</span></span></code></pre>
<pre><code><span><span class="co">## Standardizing summary statistics subset.</span></span></code></pre>
<pre><code><span><span class="co">## Standardizing main column names.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Preparing A1,A1 cols</span></span></code></pre>
<pre><code><span><span class="co">## ++ Preparing MAF,Freq cols.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Could not infer MAF from Freq.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Removing SNPs with MAF== 0 | NULL | NA or &gt;1.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Preparing N_cases,N_controls cols.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Preparing proportion_cases col.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Calculating proportion_cases from N_cases and N_controls.</span></span></code></pre>
<pre><code><span><span class="co">## Preparing sample size column (N).</span></span></code></pre>
<pre><code><span><span class="co">## Using existing 'N' column.</span></span></code></pre>
<pre><code><span><span class="co">## + Imputing t-statistic from Effect and StdErr.</span></span></code></pre>
<pre><code><span><span class="co">## + leadSNP missing. Assigning new one by min p-value.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Ensuring Effect,StdErr,P are numeric.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Ensuring 1 SNP per row and per genomic coordinate.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Removing extra whitespace</span></span></code></pre>
<pre><code><span><span class="co">## + Standardized query: 10 SNPs x 37 columns.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Saving standardized query ==&gt; /tmp/RtmpVZRnW7/results/QTL/Kunkle_2019.microgliaQTL/ABCA7_ENSG00000160953/ABCA7_ENSG00000160953_Kunkle_2019.microgliaQTL_subset.tsv.gz</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 2 ▶▶▶ Extract Linkage Disequilibrium 🔗</span> <span style="color: #00BBBB;">────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## LD_reference identified as: 1kg.</span></span></code></pre>
<pre><code><span><span class="co">## Using 1000Genomes as LD reference panel.</span></span></code></pre>
<pre><code><span><span class="co">## Constructing GRanges query using min/max ranges across one or more chromosomes.</span></span></code></pre>
<pre><code><span><span class="co">## + as_blocks=TRUE: Will query a single range per chromosome that covers all regions requested (plus anything in between).</span></span></code></pre>
<pre><code><span><span class="co">## LD Reference Panel = 1KGphase3</span></span></code></pre>
<pre><code><span><span class="co">## Querying 1KG remote server.</span></span></code></pre>
<pre><code><span><span class="co">## Selecting 503 EUR individuals from 1kgphase3.</span></span></code></pre>
<pre><code><span><span class="co">## ========= echotabix::query =========</span></span></code></pre>
<pre><code><span><span class="co">## query_dat is already a GRanges object. Returning directly.</span></span></code></pre>
<pre><code><span><span class="co">## Explicit format: 'vcf'</span></span></code></pre>
<pre><code><span><span class="co">## Querying VCF tabix file.</span></span></code></pre>
<pre><code><span><span class="co">## Querying VCF file using: VariantAnnotation</span></span></code></pre>
<pre><code><span><span class="co">## Checking query chromosome style is correct.</span></span></code></pre>
<pre><code><span><span class="co">## Chromosome format: 1</span></span></code></pre>
<pre><code><span><span class="co">## Filtering query to 503 samples and returning ScanVcfParam object.</span></span></code></pre>
<pre><code><span><span class="co">## Retrieving data.</span></span></code></pre>
<pre><code><span><span class="co">## Time difference of 4 secs</span></span></code></pre>
<pre><code><span><span class="co">## Removing 10 / 99 non-overlapping SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## Saving VCF subset ==&gt; /tmp/RtmpVZRnW7/VCF/RtmpVZRnW7.chr19-373673-375039.ALL.chr19.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.bgz</span></span></code></pre>
<pre><code><span><span class="co">## Warning in .Call(.make_vcf_geno, filename, fixed, names(geno), as.list(geno), :</span></span>
<span><span class="co">## converting NULL pointer to R NULL</span></span></code></pre>
<pre><code><span><span class="co">## Time difference of 0.4 secs</span></span></code></pre>
<pre><code><span><span class="co">## Retrieved data with 10 rows across 503 samples.</span></span></code></pre>
<pre><code><span><span class="co">## echoLD::snpStats:: `MAF` column already present.</span></span></code></pre>
<pre><code><span><span class="co">## echoLD:snpStats:: Computing pairwise LD between 10 SNPs across 503 individuals (stats = R).</span></span></code></pre>
<pre><code><span><span class="co">## Time difference of 0 secs</span></span></code></pre>
<pre><code><span><span class="co">## 10 x 10 LD_matrix (sparse)</span></span></code></pre>
<pre><code><span><span class="co">## Converting obj to sparseMatrix.</span></span></code></pre>
<pre><code><span><span class="co">## Saving sparse LD matrix ==&gt; /tmp/RtmpVZRnW7/results/QTL/Kunkle_2019.microgliaQTL/ABCA7_ENSG00000160953/LD/ABCA7_ENSG00000160953.1KGphase3_LD.RDS</span></span></code></pre>
<pre><code><span><span class="co">## Removing 1 temp files.</span></span></code></pre>
<pre><code><span><span class="co">## + FILTER:: Filtering by LD features.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 3 ▶▶▶ Filter SNPs 🚰</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## FILTER:: Filtering by SNP features.</span></span></code></pre>
<pre><code><span><span class="co">## + FILTER:: Post-filtered data: 10 x 37</span></span></code></pre>
<pre><code><span><span class="co">## + Subsetting LD matrix and dat to common SNPs...</span></span></code></pre>
<pre><code><span><span class="co">## Removing unnamed rows/cols</span></span></code></pre>
<pre><code><span><span class="co">## Replacing NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## + LD_matrix = 10 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + dat = 10 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + 10 SNPs in common.</span></span></code></pre>
<pre><code><span><span class="co">## Converting obj to sparseMatrix.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 4 ▶▶▶ Fine-map 🔊</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## Gathering method sources.</span></span></code></pre>
<pre><code><span><span class="co">## Gathering method citations.</span></span></code></pre>
<pre><code><span><span class="co">## Preparing sample size column (N).</span></span></code></pre>
<pre><code><span><span class="co">## Using existing 'N' column.</span></span></code></pre>
<pre><code><span><span class="co">## Gathering method sources.</span></span></code></pre>
<pre><code><span><span class="co">## Gathering method citations.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="background-color: #FF55FF;">ABF</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All required columns present.</span></span></span>
<span><span class="co">## <span style="background-color: #FF55FF;">FINEMAP</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All required columns present.</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All optional columns present.</span></span></span>
<span><span class="co">## <span style="background-color: #FF55FF;">SUSIE</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All required columns present.</span></span></span>
<span><span class="co">## <span style="color: #55FFFF;">✅ All optional columns present.</span></span></span></code></pre>
<pre><code><span><span class="co">## ++ Fine-mapping using 3 tool(s): ABF, FINEMAP, SUSIE</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## +++ Multi-finemap:: ABF +++</span></span></code></pre>
<pre><code><span><span class="co">## Using all default values for finemap_args$ABF</span></span></code></pre>
<pre><code><span><span class="co">## Preparing sample size column (N).</span></span></code></pre>
<pre><code><span><span class="co">## Using existing 'N' column.</span></span></code></pre>
<pre><code><span><span class="co">## Running ABF.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Credible Set SNPs identified = 1</span></span></code></pre>
<pre><code><span><span class="co">## ++ Merging ABF results with multi-finemap data.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## +++ Multi-finemap:: FINEMAP +++</span></span></code></pre>
<pre><code><span><span class="co">## Using all default values for finemap_args$FINEMAP</span></span></code></pre>
<pre><code><span><span class="co">## Preparing sample size column (N).</span></span></code></pre>
<pre><code><span><span class="co">## Using existing 'N' column.</span></span></code></pre>
<pre><code><span><span class="co">## + Subsetting LD matrix and dat to common SNPs...</span></span></code></pre>
<pre><code><span><span class="co">## Removing unnamed rows/cols</span></span></code></pre>
<pre><code><span><span class="co">## Replacing NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## + LD_matrix = 10 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + dat = 10 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + 10 SNPs in common.</span></span></code></pre>
<pre><code><span><span class="co">## Converting obj to sparseMatrix.</span></span></code></pre>
<pre><code><span><span class="co">## Constructing master file.</span></span></code></pre>
<pre><code><span><span class="co">## Constructing data.z file.</span></span></code></pre>
<pre><code><span><span class="co">## Constructing data.ld file.</span></span></code></pre>
<pre><code><span><span class="co">## FINEMAP path: /github/home/.cache/R/echofinemap/FINEMAP/finemap_v1.4.1_x86_64/finemap_v1.4.1_x86_64</span></span></code></pre>
<pre><code><span><span class="co">## Inferred FINEMAP version: 1.4.1</span></span></code></pre>
<pre><code><span><span class="co">## Running FINEMAP.</span></span></code></pre>
<pre><code>## cd .../ABCA7_ENSG00000160953 &amp;&amp;
##     .../finemap_v1.4.1_x86_64
##    
##     --sss
##    
##     --in-files .../master
##    
##     --log
##    
##     --n-causal-snps 5
## 
## |--------------------------------------|
## | Welcome to FINEMAP v1.4.1            |
## |                                      |
## | (c) 2015-2022 University of Helsinki |
## |                                      |
## | Help :                               |
## | - ./finemap --help                   |
## | - www.finemap.me                     |
## | - www.christianbenner.com            |
## |                                      |
## | Contact :                            |
## | - finemap@christianbenner.com        |
## | - matti.pirinen@helsinki.fi          |
## |--------------------------------------|
## 
## --------
## SETTINGS
## --------
## - dataset            : all
## - corr-config        : 0.95
## - n-causal-snps      : 5
## - n-configs-top      : 50000
## - n-conv-sss         : 100
## - n-iter             : 100000
## - n-threads          : 1
## - prior-k0           : 0
## - prior-std          : 0.05 
## - prob-conv-sss-tol  : 0.001
## - prob-cred-set      : 0.95
## 
## ------------
## FINE-MAPPING (1/1)
## ------------
## - GWAS summary stats               : FINEMAP/data.z
## - SNP correlations                 : FINEMAP/data.ld
## - Causal SNP stats                 : FINEMAP/data.snp
## - Causal configurations            : FINEMAP/data.config
## - Credible sets                    : FINEMAP/data.cred
## - Log file                         : FINEMAP/data.log_sss
## 
- Reading summary statistics       : +
- Reading summary statistics       : -
- Reading summary stats            : done!   
- Reading SNP correlations         : 0.000%
- Reading SNP correlations         : done!   
- Reading input                    : done!   
## 
## 
- Estimating residual variance     : +
- Estimating residual variance     : -
- Updated prior SD of effect sizes : 0.05 0.108 0.234 0.505 
## 
## - Number of GWAS samples           : 90
## - Number of SNPs                   : 10
## - Prior-Pr(# of causal SNPs is k)  : 
##   (0 -&gt; 0)
##    1 -&gt; 0.595
##    2 -&gt; 0.297
##    3 -&gt; 0.0881
##    4 -&gt; 0.0171
##    5 -&gt; 0.00229
## 0 configurations evaluated (0.000/100%)
- 10 configurations evaluated (0.000/100%)
- 10 configurations evaluated (0.000/100%)
- 10 configurations evaluated (0.050/100%)
- 10 configurations evaluated (0.100/100%) : converged after 100 iterations
## 
- Computing causal SNP statistics  : 10.000%
- Computing causal SNP statistics  : done!   
## - Regional SNP heritability        : 0.922 (SD: 0.0298 ; 95% CI: [0.864,0.979])
## - Log10-BF of &gt;= one causal SNP    : 28.8
## - Post-expected # of causal SNPs   : 1
## - Post-Pr(# of causal SNPs is k)   : 
##   (0 -&gt; 0)
##    1 -&gt; 1
##    2 -&gt; 0
##    3 -&gt; 0
##    4 -&gt; 0
##    5 -&gt; 0
## - Computing credible sets          : 
- Writing causal configurations    : 10.000%
- Writing causal SNP statistics    : 10.000%
- Writing output                   : done!   
## - Run time                         : 0 hours, 0 minutes, 0 seconds</code></pre>
<pre><code><span><span class="co">## 1 data.cred* file(s) found in the same subfolder.</span></span></code></pre>
<pre><code><span><span class="co">## Selected file based on postPr_k: data.cred1</span></span></code></pre>
<pre><code><span><span class="co">## Importing conditional probabilities (.cred file).</span></span></code></pre>
<pre><code><span><span class="co">## No configurations were causal at PP&gt;=0.95.</span></span></code></pre>
<pre><code><span><span class="co">## Importing marginal probabilities (.snp file).</span></span></code></pre>
<pre><code><span><span class="co">## Importing configuration probabilities (.config file).</span></span></code></pre>
<pre><code><span><span class="co">## FINEMAP was unable to identify any credible sets at PP&gt;=0.95.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Credible Set SNPs identified = 0</span></span></code></pre>
<pre><code><span><span class="co">## ++ Merging FINEMAP results with multi-finemap data.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## +++ Multi-finemap:: SUSIE +++</span></span></code></pre>
<pre><code><span><span class="co">## Using all default values for finemap_args$SUSIE</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: Rfast</span></span></code></pre>
<pre><code><span><span class="co">## Preparing sample size column (N).</span></span></code></pre>
<pre><code><span><span class="co">## Using existing 'N' column.</span></span></code></pre>
<pre><code><span><span class="co">## + SUSIE:: sample_size=90</span></span></code></pre>
<pre><code><span><span class="co">## + Subsetting LD matrix and dat to common SNPs...</span></span></code></pre>
<pre><code><span><span class="co">## Removing unnamed rows/cols</span></span></code></pre>
<pre><code><span><span class="co">## Replacing NAs with 0</span></span></code></pre>
<pre><code><span><span class="co">## + LD_matrix = 10 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + dat = 10 SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## + 10 SNPs in common.</span></span></code></pre>
<pre><code><span><span class="co">## Converting obj to sparseMatrix.</span></span></code></pre>
<pre><code><span><span class="co">## + SUSIE:: Using `susie_rss()` from susieR v0.12.27</span></span></code></pre>
<pre><code><span><span class="co">## + SUSIE:: Extracting Credible Sets.</span></span></code></pre>
<pre><code><span><span class="co">## ++ Credible Set SNPs identified = 5</span></span></code></pre>
<pre><code><span><span class="co">## ++ Merging SUSIE results with multi-finemap data.</span></span></code></pre>
<pre><code><span><span class="co">## Identifying Consensus SNPs...</span></span></code></pre>
<pre><code><span><span class="co">## + support_thresh = 2</span></span></code></pre>
<pre><code><span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span></code></pre>
<pre><code><span><span class="co">## + 3 fine-mapping methods used.</span></span></code></pre>
<pre><code><span><span class="co">## + 6 Credible Set SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + 0 Consensus SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## Saving merged finemapping results ==&gt; /tmp/RtmpVZRnW7/results/QTL/Kunkle_2019.microgliaQTL/ABCA7_ENSG00000160953/Multi-finemap/1KGphase3_LD.Multi-finemap.tsv.gz</span></span></code></pre>
<pre><code><span><span class="co">## + Fine-mapping with 'ABF, FINEMAP, SUSIE' completed:</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 5 ▶▶▶ Plot 📈</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## +-------- Locus Plot:  ABCA7_ENSG00000160953 --------+</span></span></code></pre>
<pre><code><span><span class="co">## + support_thresh = 2</span></span></code></pre>
<pre><code><span><span class="co">## + Calculating mean Posterior Probability (mean.PP)...</span></span></code></pre>
<pre><code><span><span class="co">## + 3 fine-mapping methods used.</span></span></code></pre>
<pre><code><span><span class="co">## + 6 Credible Set SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + 0 Consensus SNPs identified.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in CS cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## + Filling NAs in PP cols with 0.</span></span></code></pre>
<pre><code><span><span class="co">## LD_matrix detected. Coloring SNPs by LD with lead SNP.</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: QTL full window track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: QTL track</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Merged fine-mapping track</span></span></code></pre>
<pre><code><span><span class="co">## Melting PP and CS from 4 fine-mapping methods.</span></span></code></pre>
<pre><code><span><span class="co">## + echoplot:: Constructing SNP labels.</span></span></code></pre>
<pre><code><span><span class="co">## Adding SNP group labels to locus plot.</span></span></code></pre>
<pre><code><span><span class="co">## ++ echoplot:: Adding Gene model track.</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: EnsDb.Hsapiens.v75</span></span></code></pre>
<pre><code><span><span class="co">## max_transcripts= 1 .</span></span></code></pre>
<pre><code><span><span class="co">## 1  transcripts from  1  genes returned.</span></span></code></pre>
<pre><code><span><span class="co">## Fetching data...OK</span></span>
<span><span class="co">## Parsing exons...OK</span></span>
<span><span class="co">## Defining introns...OK</span></span>
<span><span class="co">## Defining UTRs...OK</span></span>
<span><span class="co">## Defining CDS...OK</span></span>
<span><span class="co">## aggregating...</span></span>
<span><span class="co">## Done</span></span>
<span><span class="co">## Constructing graphics...</span></span>
<span><span class="co">## + Adding vertical lines to highlight SNP groups.</span></span>
<span><span class="co">## +&gt;+&gt;+&gt;+&gt;+ zoom =  1x  +&lt;+&lt;+&lt;+&lt;+</span></span>
<span><span class="co">## + echoplot:: Get window suffix...</span></span>
<span><span class="co">## + echoplot:: Removing QTL full window track @ zoom=1x</span></span>
<span><span class="co">## + Removing subplot margins...</span></span>
<span><span class="co">## + Reordering tracks...</span></span>
<span><span class="co">## + Ensuring last track shows genomic units.</span></span>
<span><span class="co">## + Aligning xlimits for each subplot...</span></span>
<span><span class="co">## + Checking track heights...</span></span>
<span><span class="co">## + echoplot:: Saving plot ==&gt; /tmp/RtmpVZRnW7/results/QTL/Kunkle_2019.microgliaQTL/ABCA7_ENSG00000160953/multiview.ABCA7_ENSG00000160953.1KGphase3.1x.png</span></span>
<span><span class="co">## Found more than one class "simpleUnit" in cache; using the first, from namespace 'hexbin'</span></span>
<span><span class="co">## Also defined by 'ggbio'</span></span>
<span><span class="co">## Recording all `finemap_locus` arguments.</span></span>
<span><span class="co">## Formatting locus results.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Step 6 ▶▶▶ Postprocess data 🎁</span> <span style="color: #00BBBB;">──────────────────────────────────────────────</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## Returning results as nested list.</span></span></code></pre>
<p><img src="QTLs_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">All loci done in: </span><span style="color: #FFFFFF;">0.27 min</span></span></span>
<span><span class="co"><span style="color: #55FFFF;">## </span></span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><div class="sourceCode" id="cb248"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre style="max-height: 200px;"><code><span><span class="co">## R Under development (unstable) (2023-01-11 r83598)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] snpStats_1.49.0   Matrix_1.5-3      survival_3.5-0    echolocatoR_2.0.3</span></span>
<span><span class="co">## [5] BiocStyle_2.27.0 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] ProtGenerics_1.31.0         fs_1.5.2                   </span></span>
<span><span class="co">##   [3] matrixStats_0.63.0          bitops_1.0-7               </span></span>
<span><span class="co">##   [5] EnsDb.Hsapiens.v75_2.99.0   httr_1.4.4                 </span></span>
<span><span class="co">##   [7] RColorBrewer_1.1-3          Rgraphviz_2.43.0           </span></span>
<span><span class="co">##   [9] tools_4.3.0                 backports_1.4.1            </span></span>
<span><span class="co">##  [11] utf8_1.2.2                  R6_2.5.1                   </span></span>
<span><span class="co">##  [13] DT_0.27                     lazyeval_0.2.2             </span></span>
<span><span class="co">##  [15] withr_2.5.0                 prettyunits_1.1.1          </span></span>
<span><span class="co">##  [17] GGally_2.1.2                gridExtra_2.3              </span></span>
<span><span class="co">##  [19] cli_3.6.0                   Biobase_2.59.0             </span></span>
<span><span class="co">##  [21] textshaping_0.3.6           labeling_0.4.2             </span></span>
<span><span class="co">##  [23] ggbio_1.47.0                sass_0.4.4                 </span></span>
<span><span class="co">##  [25] mvtnorm_1.1-3               readr_2.1.3                </span></span>
<span><span class="co">##  [27] proxy_0.4-27                pkgdown_2.0.7              </span></span>
<span><span class="co">##  [29] mixsqp_0.3-48               Rsamtools_2.15.1           </span></span>
<span><span class="co">##  [31] systemfonts_1.0.4           foreign_0.8-84             </span></span>
<span><span class="co">##  [33] R.utils_2.12.2              dichromat_2.0-0.1          </span></span>
<span><span class="co">##  [35] maps_3.4.1                  BSgenome_1.67.3            </span></span>
<span><span class="co">##  [37] readxl_1.4.1                susieR_0.12.27             </span></span>
<span><span class="co">##  [39] pals_1.7                    rstudioapi_0.14            </span></span>
<span><span class="co">##  [41] RSQLite_2.2.20              httpcode_0.3.0             </span></span>
<span><span class="co">##  [43] generics_0.1.3              BiocIO_1.9.1               </span></span>
<span><span class="co">##  [45] echoconda_0.99.9            dplyr_1.0.10               </span></span>
<span><span class="co">##  [47] zip_2.2.2                   interp_1.1-3               </span></span>
<span><span class="co">##  [49] fansi_1.0.3                 DescTools_0.99.47          </span></span>
<span><span class="co">##  [51] S4Vectors_0.37.3            catalogueR_1.0.1           </span></span>
<span><span class="co">##  [53] R.methodsS3_1.8.2           lifecycle_1.0.3            </span></span>
<span><span class="co">##  [55] yaml_2.3.6                  SummarizedExperiment_1.29.1</span></span>
<span><span class="co">##  [57] BiocFileCache_2.7.1         echoplot_0.99.6            </span></span>
<span><span class="co">##  [59] grid_4.3.0                  blob_1.2.3                 </span></span>
<span><span class="co">##  [61] crayon_1.5.2                dir.expiry_1.7.0           </span></span>
<span><span class="co">##  [63] lattice_0.20-45             GenomicFeatures_1.51.2     </span></span>
<span><span class="co">##  [65] mapproj_1.2.11              KEGGREST_1.39.0            </span></span>
<span><span class="co">##  [67] MungeSumstats_1.7.15        pillar_1.8.1               </span></span>
<span><span class="co">##  [69] knitr_1.41                  GenomicRanges_1.51.4       </span></span>
<span><span class="co">##  [71] rjson_0.2.21                osfr_0.2.9                 </span></span>
<span><span class="co">##  [73] boot_1.3-28.1               gld_2.6.6                  </span></span>
<span><span class="co">##  [75] codetools_0.2-18            glue_1.6.2                 </span></span>
<span><span class="co">##  [77] data.table_1.14.6           coloc_5.1.0.1              </span></span>
<span><span class="co">##  [79] vctrs_0.5.1                 png_0.1-8                  </span></span>
<span><span class="co">##  [81] XGR_1.1.8                   cellranger_1.1.0           </span></span>
<span><span class="co">##  [83] gtable_0.3.1                assertthat_0.2.1           </span></span>
<span><span class="co">##  [85] cachem_1.0.6                dnet_1.1.7                 </span></span>
<span><span class="co">##  [87] xfun_0.36                   openxlsx_4.2.5.1           </span></span>
<span><span class="co">##  [89] Rfast_2.0.6                 gargle_1.2.1               </span></span>
<span><span class="co">##  [91] ellipsis_0.3.2              nlme_3.1-161               </span></span>
<span><span class="co">##  [93] bit64_4.0.5                 progress_1.2.2             </span></span>
<span><span class="co">##  [95] filelock_1.0.2              googleAuthR_2.0.0          </span></span>
<span><span class="co">##  [97] GenomeInfoDb_1.35.12        rprojroot_2.0.3            </span></span>
<span><span class="co">##  [99] bslib_0.4.2                 irlba_2.3.5.1              </span></span>
<span><span class="co">## [101] rpart_4.1.19                colorspace_2.0-3           </span></span>
<span><span class="co">## [103] BiocGenerics_0.45.0         DBI_1.1.3                  </span></span>
<span><span class="co">## [105] Hmisc_4.7-2                 nnet_7.3-18                </span></span>
<span><span class="co">## [107] Exact_3.2                   tidyselect_1.2.0           </span></span>
<span><span class="co">## [109] bit_4.0.5                   compiler_4.3.0             </span></span>
<span><span class="co">## [111] curl_5.0.0                  graph_1.77.1               </span></span>
<span><span class="co">## [113] htmlTable_2.4.1             expm_0.999-7               </span></span>
<span><span class="co">## [115] basilisk.utils_1.11.1       xml2_1.3.3                 </span></span>
<span><span class="co">## [117] desc_1.4.2                  DelayedArray_0.25.0        </span></span>
<span><span class="co">## [119] bookdown_0.32               rtracklayer_1.59.1         </span></span>
<span><span class="co">## [121] checkmate_2.1.0             scales_1.2.1               </span></span>
<span><span class="co">## [123] hexbin_1.28.2               echoLD_0.99.9              </span></span>
<span><span class="co">## [125] RBGL_1.75.0                 RCircos_1.2.2              </span></span>
<span><span class="co">## [127] rappdirs_0.3.3              stringr_1.5.0              </span></span>
<span><span class="co">## [129] supraHex_1.37.0             digest_0.6.31              </span></span>
<span><span class="co">## [131] piggyback_0.1.4             rmarkdown_2.20             </span></span>
<span><span class="co">## [133] basilisk_1.11.2             XVector_0.39.0             </span></span>
<span><span class="co">## [135] htmltools_0.5.4             pkgconfig_2.0.3            </span></span>
<span><span class="co">## [137] jpeg_0.1-10                 base64enc_0.1-3            </span></span>
<span><span class="co">## [139] MatrixGenerics_1.11.0       echodata_0.99.16           </span></span>
<span><span class="co">## [141] highr_0.10                  ensembldb_2.23.1           </span></span>
<span><span class="co">## [143] dbplyr_2.3.0                fastmap_1.1.0              </span></span>
<span><span class="co">## [145] rlang_1.0.6                 htmlwidgets_1.6.1          </span></span>
<span><span class="co">## [147] farver_2.1.1                echofinemap_0.99.5         </span></span>
<span><span class="co">## [149] jquerylib_0.1.4             jsonlite_1.8.4             </span></span>
<span><span class="co">## [151] BiocParallel_1.33.9         R.oo_1.25.0                </span></span>
<span><span class="co">## [153] VariantAnnotation_1.45.0    RCurl_1.98-1.9             </span></span>
<span><span class="co">## [155] magrittr_2.0.3              Formula_1.2-4              </span></span>
<span><span class="co">## [157] GenomeInfoDbData_1.2.9      ggnetwork_0.5.10           </span></span>
<span><span class="co">## [159] patchwork_1.1.2             munsell_0.5.0              </span></span>
<span><span class="co">## [161] Rcpp_1.0.9                  ggnewscale_0.4.8           </span></span>
<span><span class="co">## [163] ape_5.6-2                   viridis_0.6.2              </span></span>
<span><span class="co">## [165] reticulate_1.27             RcppZiggurat_0.1.6         </span></span>
<span><span class="co">## [167] stringi_1.7.12              rootSolve_1.8.2.3          </span></span>
<span><span class="co">## [169] zlibbioc_1.45.0             MASS_7.3-58.1              </span></span>
<span><span class="co">## [171] plyr_1.8.8                  parallel_4.3.0             </span></span>
<span><span class="co">## [173] ggrepel_0.9.2               lmom_2.9                   </span></span>
<span><span class="co">## [175] deldir_1.0-6                echoannot_0.99.10          </span></span>
<span><span class="co">## [177] Biostrings_2.67.0           splines_4.3.0              </span></span>
<span><span class="co">## [179] hms_1.1.2                   igraph_1.3.5               </span></span>
<span><span class="co">## [181] reshape2_1.4.4              biomaRt_2.55.0             </span></span>
<span><span class="co">## [183] stats4_4.3.0                crul_1.3                   </span></span>
<span><span class="co">## [185] XML_3.99-0.13               evaluate_0.20              </span></span>
<span><span class="co">## [187] latticeExtra_0.6-30         biovizBase_1.47.0          </span></span>
<span><span class="co">## [189] BiocManager_1.30.19         tzdb_0.3.0                 </span></span>
<span><span class="co">## [191] tidyr_1.2.1                 purrr_1.0.1                </span></span>
<span><span class="co">## [193] reshape_0.8.9               ggplot2_3.4.0              </span></span>
<span><span class="co">## [195] echotabix_0.99.9            restfulr_0.0.15            </span></span>
<span><span class="co">## [197] AnnotationFilter_1.23.0     e1071_1.7-12               </span></span>
<span><span class="co">## [199] downloadR_0.99.6            viridisLite_0.4.1          </span></span>
<span><span class="co">## [201] class_7.3-20.1              ragg_1.2.5                 </span></span>
<span><span class="co">## [203] OrganismDbi_1.41.0          tibble_3.1.8               </span></span>
<span><span class="co">## [205] memoise_2.0.1               AnnotationDbi_1.61.0       </span></span>
<span><span class="co">## [207] GenomicAlignments_1.35.0    IRanges_2.33.0             </span></span>
<span><span class="co">## [209] cluster_2.1.4</span></span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
