<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Summarise • echolocatoR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Summarise">
<meta property="og:description" content="echolocatoR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">echolocatoR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/echolocatoR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/QTLs.html">QTLs</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="../articles/finemapping_portal.html">echolocatoR Fine-mapping Portal</a>
    </li>
    <li>
      <a href="../articles/plot_locus.html">Plot locus</a>
    </li>
    <li>
      <a href="../articles/summarise.html">Summarise</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RajLabMSSM/echolocatoR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Summarise</h1>
                        <h4 data-toc-skip class="author"><h4> ¶ Author:
<i>Brian M. Schilder</i> ¶ </h4></h4>
            
            <h4 data-toc-skip class="date"><h4> ¶ Updated:
<i>Nov-01-2022</i> ¶ </h4></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/RajLabMSSM/echolocatoR/blob/HEAD/vignettes/summarise.Rmd" class="external-link"><code>vignettes/summarise.Rmd</code></a></small>
      <div class="hidden name"><code>summarise.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RajLabMSSM/echolocatoR" class="external-link">echolocatoR</a></span><span class="op">)</span>  </span></code></pre></div>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">##   method from   </span></span>
<span><span class="co">##   +.gg   ggplot2</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">⠊⠉⠡⣀⣀⠊⠉⠡⣀⣀⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⢁⡐⠉⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠉⢂⡈⠑⣀⠉⢄⡈⠡⣀</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⡈⡐⢂⢁⠒⡈⡐⢂⢁⠒⡈⡐⢂⢁⠑⡈⡈⢄⢁⠡⠌⡈⠤⢁⠡⠌⡈⠤⢁⠡⠌⡈⡠⢁⢁⠊⡈⡐⢂</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="color: #55FFFF; font-weight: bold;">🦇  🦇  🦇 e c h o l o c a t o R 🦇  🦇  🦇</span> <span style="color: #00BBBB;">─────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="color: #55FFFF; font-weight: bold;">v2.0.2</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #55FFFF;">⠌⡈⡐⢂⢁⠒⡈⡐⢂⢁⠒⡈⡐⢂⢁⠑⡈⡈⢄⢁⠡⠌⡈⠤⢁⠡⠌⡈⠤⢁⠡⠌⡈⡠⢁⢁⠊⡈⡐⢂</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠌⢁⡐⠉⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠊⢂⡐⠑⣀⠉⢂⡈⠑⣀⠉⢄⡈⠡⣀</span>                                    </span></span>
<span><span class="co">## <span style="color: #55FFFF;">⠊⠉⠡⣀⣀⠊⠉⠡⣀⣀⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠⠊⠉⠢⣀⡠</span>                                    </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ If you use </span><span style="color: #00BBBB; font-weight: bold;">echolocatoR</span><span style="color: #00BBBB;"> or any of the </span><span style="color: #00BBBB; font-style: italic;">echoverse</span><span style="color: #00BBBB;"> subpackages, please cite:</span>      </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶ Brian M Schilder, Jack Humphrey, Towfique</span>                                </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     Raj (2021) echolocatoR: an automated</span>                                       </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     end-to-end statistical and functional</span>                                      </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     genomic fine-mapping pipeline,</span>                                             </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     Bioinformatics; btab658,</span>                                                   </span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://doi.org/10.1093/bioinformatics/btab658</span>                             </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ Please report any bugs/feature requests on GitHub:</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://github.com/RajLabMSSM/echolocatoR/issues</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">ⓞ Contributions are welcome!:</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     ▶</span></span></span>
<span><span class="co">## <span style="color: #FF55FF;">     https://github.com/RajLabMSSM/echolocatoR/pulls</span></span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
<div class="section level2">
<h2 id="download-data">Download data<a class="anchor" aria-label="anchor" href="#download-data"></a>
</h2>
</div>
<div class="section level2">
<h2 id="summarise">Summarise<a class="anchor" aria-label="anchor" href="#summarise"></a>
</h2>
<p>Using pre-merged data for vignette speed.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">merged_DT</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/get_Nalls2019_merged.html" class="external-link">get_Nalls2019_merged</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="get_snpgroup_counts">
<code>get_SNPgroup_counts()</code><a class="anchor" aria-label="anchor" href="#get_snpgroup_counts"></a>
</h3>
<p>Get the number of SNPs for each SNP group per locus.<br>
It also prints the mean number of SNPs for each SNP group across all
loci.<br><strong>NOTE</strong>: You will need to make sure to set
<code>merge_finemapping_results(minimum_support=1)</code> in the above
step to get accurate counts for all SNP groups.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_groups</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/get_SNPgroup_counts.html" class="external-link">get_SNPgroup_counts</a></span><span class="op">(</span>merged_DT <span class="op">=</span> <span class="va">merged_DT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## All loci (75) :</span></span></code></pre>
<pre><code><span><span class="co">##            Total.SNPs          nom.sig.GWAS              sig.GWAS </span></span>
<span><span class="co">##               4948.16                924.68                 82.36 </span></span>
<span><span class="co">##                    CS             Consensus          topConsensus </span></span>
<span><span class="co">##                  7.88                  2.69                  1.47 </span></span>
<span><span class="co">## topConsensus.leadGWAS </span></span>
<span><span class="co">##                  0.41</span></span></code></pre>
<pre><code><span><span class="co">## Loci with at least one Consensus SNP (69) :</span></span></code></pre>
<pre><code><span><span class="co">##            Total.SNPs          nom.sig.GWAS              sig.GWAS </span></span>
<span><span class="co">##               5019.07                911.41                 84.28 </span></span>
<span><span class="co">##                    CS             Consensus          topConsensus </span></span>
<span><span class="co">##                  7.77                  2.93                  1.59 </span></span>
<span><span class="co">## topConsensus.leadGWAS </span></span>
<span><span class="co">##                  0.45</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="get_cs_counts">
<code>get_CS_counts()</code><a class="anchor" aria-label="anchor" href="#get_cs_counts"></a>
</h3>
<p>County the number of tool-specific and UCS Credible Set SNPs per
locus.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">UCS_counts</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/get_CS_counts.html" class="external-link">get_CS_counts</a></span><span class="op">(</span>merged_DT <span class="op">=</span> <span class="va">merged_DT</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">UCS_counts</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="12%">
<col width="11%">
<col width="13%">
<col width="21%">
<col width="15%">
<col width="12%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="left">Locus</th>
<th align="right">ABF.CS_size</th>
<th align="right">SUSIE.CS_size</th>
<th align="right">POLYFUN_SUSIE.CS_size</th>
<th align="right">FINEMAP.CS_size</th>
<th align="right">mean.CS_size</th>
<th align="right">UCS.CS_size</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">GPNMB</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">13</td>
</tr>
<tr class="even">
<td align="left">MAP4K4</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">12</td>
</tr>
<tr class="odd">
<td align="left">MBNL2</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">12</td>
</tr>
<tr class="even">
<td align="left">TMEM163</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">12</td>
</tr>
<tr class="odd">
<td align="left">CRLS1</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">11</td>
</tr>
<tr class="even">
<td align="left">DNAH17</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">11</td>
</tr>
<tr class="odd">
<td align="left">GBF1</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">11</td>
</tr>
<tr class="even">
<td align="left">GCH1</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">2</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="right">11</td>
</tr>
<tr class="odd">
<td align="left">MIPOL1</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">11</td>
</tr>
<tr class="even">
<td align="left">FBRSL1</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">10</td>
</tr>
<tr class="odd">
<td align="left">FYN</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">10</td>
</tr>
<tr class="even">
<td align="left">GALC</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">10</td>
</tr>
<tr class="odd">
<td align="left">LCORL</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">10</td>
</tr>
<tr class="even">
<td align="left">LOC100131289</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">10</td>
</tr>
<tr class="odd">
<td align="left">MED12L</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">10</td>
</tr>
<tr class="even">
<td align="left">SLC2A13</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">10</td>
</tr>
<tr class="odd">
<td align="left">TRIM40</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">10</td>
</tr>
<tr class="even">
<td align="left">ATG14</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="left">CD19</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="even">
<td align="left">CHRNB1</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="left">FAM49B</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="even">
<td align="left">ITGA8</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="left">ITPKB</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="even">
<td align="left">MCCC1</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="left">RPS6KL1</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="even">
<td align="left">SH3GL2</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="left">STK39</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="even">
<td align="left">TMEM175</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="left">C5orf24</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">CLCN3</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">CRHR1</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">CTSB</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">DLG2</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">DYRK1A</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">IGSF9B</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">IP6K2</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">NOD2</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">NUCKS1</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">RETREG3</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">RNF141</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">SIPA1L2</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">SP1</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">SPPL2B</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">3</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">WNT3</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">FCGR2A</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">GS1-124K5.11</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="odd">
<td align="left">HIP1R</td>
<td align="right">1</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">1</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">KCNIP3</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="odd">
<td align="left">KPNA1</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">LINC00693</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="odd">
<td align="left">LRRK2</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">SETD1A</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="odd">
<td align="left">SNCA</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">SYT17</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="odd">
<td align="left">VAMP4</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">3</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">VPS13C</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">7</td>
</tr>
<tr class="odd">
<td align="left">ELOVL7</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="right">6</td>
</tr>
<tr class="even">
<td align="left">FAM171A2</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">FGF20</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">6</td>
</tr>
<tr class="even">
<td align="left">HLA-DRB5</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">4</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">INPP5F</td>
<td align="right">1</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">1</td>
<td align="right">6</td>
</tr>
<tr class="even">
<td align="left">KRTCAP2</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">MEX3C</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">6</td>
</tr>
<tr class="even">
<td align="left">RIMS1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">RIT2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">1</td>
<td align="right">6</td>
</tr>
<tr class="even">
<td align="left">RPS12</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">SCAF11</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">6</td>
</tr>
<tr class="even">
<td align="left">SPTSSB</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">UBAP2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">6</td>
</tr>
<tr class="even">
<td align="left">KCNS3</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">5</td>
</tr>
<tr class="odd">
<td align="left">CHD9</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">FAM47E-STBD1</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">PAM</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">3</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">SATB1</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">LMNB1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="plot">Plot<a class="anchor" aria-label="anchor" href="#plot"></a>
</h2>
<ul>
<li>The following functions each return a list containing both the
<code>...$plot</code> and the <code>...$data</code> used to make the
plot.<br>
</li>
<li>Where available, <code>snp_filter</code> allows user to use any
filtering argument (supplied as a string) to subset the data they want
to use in the plot/data.</li>
</ul>
<div class="section level3">
<h3 id="colocalization-results">Colocalization results<a class="anchor" aria-label="anchor" href="#colocalization-results"></a>
</h3>
<p>If you ran colocalization tests with <code>echolocatoR</code> (via
<code>catalogueR</code>) you can use those results to come up with a top
QTL nominated gene for each locus (potentially implicating that gene in
your phenotype).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coloc_res</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echodata/man/get_Nalls2019_coloc.html" class="external-link">get_Nalls2019_coloc</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="super-summary-plot">Super summary plot<a class="anchor" aria-label="anchor" href="#super-summary-plot"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">super_plot</span> <span class="op">&lt;-</span> <span class="fu">echoannot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/echoannot/man/super_summary_plot.html" class="external-link">super_summary_plot</a></span><span class="op">(</span>merged_DT <span class="op">=</span> <span class="va">merged_DT</span>, </span>
<span>                                            coloc_results <span class="op">=</span> <span class="va">coloc_res</span>,</span>
<span>                                            plot_missense <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## + SUMMARISE:: Nominating genes by top colocalized eQTL eGenes</span></span></code></pre>
<pre><code><span><span class="co">## Importing previously downloaded files: /github/home/.cache/R/echoannot/NOTT2019_epigenomic_peaks.rds</span></span></code></pre>
<pre><code><span><span class="co">## ++ NOTT2019:: 634,540 ranges retrieved.</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## 113 query SNP(s) detected with reference overlap.</span></span></code></pre>
<pre><code><span><span class="co">## ++ NOTT2019:: Getting regulatory regions data.</span></span></code></pre>
<pre><code><span><span class="co">## Importing Astrocyte enhancers ...</span></span></code></pre>
<pre><code><span><span class="co">## Importing Astrocyte promoters ...</span></span></code></pre>
<pre><code><span><span class="co">## Importing Neuronal enhancers ...</span></span></code></pre>
<pre><code><span><span class="co">## Importing Neuronal promoters ...</span></span></code></pre>
<pre><code><span><span class="co">## Importing Oligo enhancers ...</span></span></code></pre>
<pre><code><span><span class="co">## Importing Oligo promoters ...</span></span></code></pre>
<pre><code><span><span class="co">## Importing Microglia enhancers ...</span></span></code></pre>
<pre><code><span><span class="co">## Importing Microglia promoters ...</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span>
<span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## 48 query SNP(s) detected with reference overlap.</span></span></code></pre>
<pre><code><span><span class="co">## ++ NOTT2019:: Getting interaction anchors data.</span></span></code></pre>
<pre><code><span><span class="co">## Importing Microglia interactome ...</span></span></code></pre>
<pre><code><span><span class="co">## Importing Neuronal interactome ...</span></span></code></pre>
<pre><code><span><span class="co">## Importing Oligo interactome ...</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## 52 query SNP(s) detected with reference overlap.</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## 44 query SNP(s) detected with reference overlap.</span></span></code></pre>
<pre><code><span><span class="co">## CORCES2020:: Extracting overlapping cell-type-specific scATAC-seq peaks</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## 13 query SNP(s) detected with reference overlap.</span></span></code></pre>
<pre><code><span><span class="co">## CORCES2020:: Annotating peaks by cell-type-specific target genes</span></span></code></pre>
<pre><code><span><span class="co">## CORCES2020:: Extracting overlapping bulkATAC-seq peaks from brain tissue</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## 4 query SNP(s) detected with reference overlap.</span></span></code></pre>
<pre><code><span><span class="co">## CORCES2020:: Annotating peaks by bulk brain target genes</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## 70 query SNP(s) detected with reference overlap.</span></span></code></pre>
<pre><code><span><span class="co">## Converting dat to GRanges object.</span></span></code></pre>
<pre><code><span><span class="co">## 72 query SNP(s) detected with reference overlap.</span></span></code></pre>
<pre><code><span><span class="co">## + CORCES2020:: Found 142 hits with HiChIP_FitHiChIP coaccessibility loop anchors.</span></span></code></pre>
<pre><code><span><span class="co">## Found more than one class "simpleUnit" in cache; using the first, from namespace 'hexbin'</span></span></code></pre>
<pre><code><span><span class="co">## Also defined by 'ggbio'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 83 rows containing missing values (position_stack).</span></span></code></pre>
<p><img src="summarise_files/figure-html/super_summary_plot()-1.png" width="1440"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre style="max-height: 200px;"><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.5 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] echolocatoR_2.0.2 BiocStyle_2.25.0 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] utf8_1.2.2                  reticulate_1.26            </span></span>
<span><span class="co">##   [3] R.utils_2.12.1              tidyselect_1.2.0           </span></span>
<span><span class="co">##   [5] RSQLite_2.2.18              AnnotationDbi_1.59.1       </span></span>
<span><span class="co">##   [7] htmlwidgets_1.5.4           grid_4.2.1                 </span></span>
<span><span class="co">##   [9] BiocParallel_1.31.15        XGR_1.1.8                  </span></span>
<span><span class="co">##  [11] munsell_0.5.0               codetools_0.2-18           </span></span>
<span><span class="co">##  [13] ragg_1.2.4                  interp_1.1-3               </span></span>
<span><span class="co">##  [15] DT_0.26                     withr_2.5.0                </span></span>
<span><span class="co">##  [17] colorspace_2.0-3            OrganismDbi_1.39.1         </span></span>
<span><span class="co">##  [19] Biobase_2.57.3              filelock_1.0.2             </span></span>
<span><span class="co">##  [21] highr_0.9                   knitr_1.40                 </span></span>
<span><span class="co">##  [23] supraHex_1.35.0             rstudioapi_0.14            </span></span>
<span><span class="co">##  [25] stats4_4.2.1                DescTools_0.99.47          </span></span>
<span><span class="co">##  [27] labeling_0.4.2              MatrixGenerics_1.9.1       </span></span>
<span><span class="co">##  [29] GenomeInfoDbData_1.2.9      mixsqp_0.3-43              </span></span>
<span><span class="co">##  [31] farver_2.1.1                bit64_4.0.5                </span></span>
<span><span class="co">##  [33] echoconda_0.99.8            rprojroot_2.0.3            </span></span>
<span><span class="co">##  [35] basilisk_1.9.12             vctrs_0.5.0                </span></span>
<span><span class="co">##  [37] generics_0.1.3              xfun_0.34                  </span></span>
<span><span class="co">##  [39] biovizBase_1.45.0           BiocFileCache_2.5.2        </span></span>
<span><span class="co">##  [41] R6_2.5.1                    GenomeInfoDb_1.33.16       </span></span>
<span><span class="co">##  [43] AnnotationFilter_1.21.0     bitops_1.0-7               </span></span>
<span><span class="co">##  [45] cachem_1.0.6                reshape_0.8.9              </span></span>
<span><span class="co">##  [47] DelayedArray_0.23.2         assertthat_0.2.1           </span></span>
<span><span class="co">##  [49] BiocIO_1.7.1                scales_1.2.1               </span></span>
<span><span class="co">##  [51] nnet_7.3-18                 rootSolve_1.8.2.3          </span></span>
<span><span class="co">##  [53] gtable_0.3.1                lmom_2.9                   </span></span>
<span><span class="co">##  [55] ggbio_1.45.0                ensembldb_2.21.5           </span></span>
<span><span class="co">##  [57] rlang_1.0.6                 systemfonts_1.0.4          </span></span>
<span><span class="co">##  [59] splines_4.2.1               echodata_0.99.15           </span></span>
<span><span class="co">##  [61] lazyeval_0.2.2              rtracklayer_1.57.0         </span></span>
<span><span class="co">##  [63] dichromat_2.0-0.1           hexbin_1.28.2              </span></span>
<span><span class="co">##  [65] checkmate_2.1.0             BiocManager_1.30.19        </span></span>
<span><span class="co">##  [67] yaml_2.3.6                  reshape2_1.4.4             </span></span>
<span><span class="co">##  [69] snpStats_1.47.1             backports_1.4.1            </span></span>
<span><span class="co">##  [71] GenomicFeatures_1.49.8      ggnetwork_0.5.10           </span></span>
<span><span class="co">##  [73] Hmisc_4.7-1                 RBGL_1.73.0                </span></span>
<span><span class="co">##  [75] tools_4.2.1                 echoplot_0.99.5            </span></span>
<span><span class="co">##  [77] bookdown_0.29               ggplot2_3.3.6              </span></span>
<span><span class="co">##  [79] ellipsis_0.3.2              jquerylib_0.1.4            </span></span>
<span><span class="co">##  [81] catalogueR_1.0.0            RColorBrewer_1.1-3         </span></span>
<span><span class="co">##  [83] proxy_0.4-27                BiocGenerics_0.43.4        </span></span>
<span><span class="co">##  [85] coloc_5.1.0.1               Rcpp_1.0.9                 </span></span>
<span><span class="co">##  [87] plyr_1.8.7                  base64enc_0.1-3            </span></span>
<span><span class="co">##  [89] progress_1.2.2              zlibbioc_1.43.0            </span></span>
<span><span class="co">##  [91] purrr_0.3.5                 RCurl_1.98-1.9             </span></span>
<span><span class="co">##  [93] basilisk.utils_1.9.4        prettyunits_1.1.1          </span></span>
<span><span class="co">##  [95] rpart_4.1.16                deldir_1.0-6               </span></span>
<span><span class="co">##  [97] viridis_0.6.2               S4Vectors_0.35.4           </span></span>
<span><span class="co">##  [99] cluster_2.1.4               SummarizedExperiment_1.27.3</span></span>
<span><span class="co">## [101] ggrepel_0.9.1               fs_1.5.2                   </span></span>
<span><span class="co">## [103] crul_1.3                    magrittr_2.0.3             </span></span>
<span><span class="co">## [105] data.table_1.14.4           echotabix_0.99.8           </span></span>
<span><span class="co">## [107] dnet_1.1.7                  openxlsx_4.2.5.1           </span></span>
<span><span class="co">## [109] mvtnorm_1.1-3               ProtGenerics_1.29.1        </span></span>
<span><span class="co">## [111] matrixStats_0.62.0          hms_1.1.2                  </span></span>
<span><span class="co">## [113] patchwork_1.1.2             evaluate_0.17              </span></span>
<span><span class="co">## [115] XML_3.99-0.12               jpeg_0.1-9                 </span></span>
<span><span class="co">## [117] readxl_1.4.1                IRanges_2.31.2             </span></span>
<span><span class="co">## [119] gridExtra_2.3               compiler_4.2.1             </span></span>
<span><span class="co">## [121] biomaRt_2.53.3              tibble_3.1.8               </span></span>
<span><span class="co">## [123] crayon_1.5.2                R.oo_1.25.0                </span></span>
<span><span class="co">## [125] htmltools_0.5.3             echoannot_0.99.10          </span></span>
<span><span class="co">## [127] tzdb_0.3.0                  Formula_1.2-4              </span></span>
<span><span class="co">## [129] tidyr_1.2.1                 expm_0.999-6               </span></span>
<span><span class="co">## [131] Exact_3.2                   DBI_1.1.3                  </span></span>
<span><span class="co">## [133] dbplyr_2.2.1                MASS_7.3-58.1              </span></span>
<span><span class="co">## [135] rappdirs_0.3.3              boot_1.3-28                </span></span>
<span><span class="co">## [137] Matrix_1.5-1                readr_2.1.3                </span></span>
<span><span class="co">## [139] piggyback_0.1.4             cli_3.4.1                  </span></span>
<span><span class="co">## [141] R.methodsS3_1.8.2           echofinemap_0.99.4         </span></span>
<span><span class="co">## [143] parallel_4.2.1              igraph_1.3.5               </span></span>
<span><span class="co">## [145] GenomicRanges_1.49.1        pkgconfig_2.0.3            </span></span>
<span><span class="co">## [147] pkgdown_2.0.6.9000          GenomicAlignments_1.33.1   </span></span>
<span><span class="co">## [149] dir.expiry_1.5.1            RCircos_1.2.2              </span></span>
<span><span class="co">## [151] foreign_0.8-83              osfr_0.2.9                 </span></span>
<span><span class="co">## [153] xml2_1.3.3                  bslib_0.4.0                </span></span>
<span><span class="co">## [155] XVector_0.37.1              echoLD_0.99.8              </span></span>
<span><span class="co">## [157] stringr_1.4.1               VariantAnnotation_1.43.3   </span></span>
<span><span class="co">## [159] digest_0.6.30               graph_1.75.1               </span></span>
<span><span class="co">## [161] httpcode_0.3.0              Biostrings_2.65.6          </span></span>
<span><span class="co">## [163] rmarkdown_2.17              cellranger_1.1.0           </span></span>
<span><span class="co">## [165] htmlTable_2.4.1             gld_2.6.6                  </span></span>
<span><span class="co">## [167] restfulr_0.0.15             curl_4.3.3                 </span></span>
<span><span class="co">## [169] Rsamtools_2.13.4            rjson_0.2.21               </span></span>
<span><span class="co">## [171] lifecycle_1.0.3             nlme_3.1-160               </span></span>
<span><span class="co">## [173] jsonlite_1.8.3              desc_1.4.2                 </span></span>
<span><span class="co">## [175] viridisLite_0.4.1           BSgenome_1.65.4            </span></span>
<span><span class="co">## [177] fansi_1.0.3                 downloadR_0.99.5           </span></span>
<span><span class="co">## [179] pillar_1.8.1                susieR_0.12.27             </span></span>
<span><span class="co">## [181] lattice_0.20-45             GGally_2.1.2               </span></span>
<span><span class="co">## [183] survival_3.4-0              KEGGREST_1.37.3            </span></span>
<span><span class="co">## [185] fastmap_1.1.0               httr_1.4.4                 </span></span>
<span><span class="co">## [187] glue_1.6.2                  zip_2.2.2                  </span></span>
<span><span class="co">## [189] png_0.1-7                   bit_4.0.4                  </span></span>
<span><span class="co">## [191] Rgraphviz_2.41.2            class_7.3-20               </span></span>
<span><span class="co">## [193] stringi_1.7.8               sass_0.4.2                 </span></span>
<span><span class="co">## [195] blob_1.2.3                  textshaping_0.3.6          </span></span>
<span><span class="co">## [197] latticeExtra_0.6-30         memoise_2.0.1              </span></span>
<span><span class="co">## [199] dplyr_1.0.10                irlba_2.3.5.1              </span></span>
<span><span class="co">## [201] e1071_1.7-12                ape_5.6-2</span></span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
